{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# IRD Xenium Spatial Neighborhood Analysis\n",
    "\n",
    "This notebook performs downstream analysis of Xenium spatial transcriptomics data from the Multiple Myeloma IRD study. The study focuses on understanding the microenvironment changes in newly diagnosed multiple myeloma (NDMM) and after autologous stem cell transplant (ASCT) therapy. The analysis includes cell type composition, gene expression patterns, immune microenvironment characterization, and spatial neighborhood identification.\n",
    "\n",
    "Data preprocessing and cell type annotation were performed outside of this notebook.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "from scipy.stats import mannwhitneyu\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from itertools import combinations\n",
    "\n",
    "import matplotlib as mpl\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import custom helper functions\n",
    "import sys\n",
    "sys.path.append('/diskmnt/Users2/chouw/Projects/SenNet_bone/src/spatial/utils')\n",
    "import spatial_utils\n",
    "import plot_utils\n",
    "\n",
    "# If helper functions are updated, uncomment and run this line\n",
    "#import importlib\n",
    "#importlib.reload(plot_utils)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 1: Data Loading and Cell Type Composition\n",
    "\n",
    "Load the merged Xenium spatial transcriptomics dataset and analyze cell type composition across different collection timepoints (NBM, NDMM, PT).\n",
    "\n",
    "### Load merged Xenium data and revised cell type annotations\n",
    "\n",
    "The merged dataset contains cells from multiple samples across three collection timepoints:\n",
    "- **NBM**: Normal bone marrow\n",
    "- **NDMM**: Newly diagnosed multiple myeloma\n",
    "- **PT**: Post-treatment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_xenium_merge = sc.read_h5ad(\"/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/Xenium/analysis/merged.h5ad\")\n",
    "ird_xenium_merge.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The most updated cell type annotation is in the 'ct' column\n",
    "ird_xenium_merge.obs['ct'].unique().tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cell type composition analysis across timepoints\n",
    "\n",
    "Calculate the percentage of cell types per sample and compare across collection timepoints to identify disease-associated changes in cell type abundance.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_nbm_mm = ird_xenium_merge[ird_xenium_merge.obs['Collection'].isin(['NBM', 'NDMM'])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate percentage of each cell type in each sample\n",
    "ird_cell_info = ird_nbm_mm.obs.copy()\n",
    "ct_counts = ird_cell_info.groupby(['Sample', 'ct']).size().reset_index(name = 'ct_count')   ## Number of cells per cell type per sample\n",
    "sample_counts = ird_cell_info.groupby('Sample').size().reset_index(name='total_count')        ## Total number of cells per sample\n",
    "ct_counts = ct_counts.merge(sample_counts, on='Sample', how='left')\n",
    "ct_counts['pct'] = ct_counts['ct_count'] / ct_counts['total_count'] * 100\n",
    "ct_counts = ct_counts.merge(ird_cell_info[['Sample', 'Collection']].drop_duplicates(), on='Sample', how='left')\n",
    "ct_counts.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_ct = ird_xenium_merge.obs['ct'].unique().tolist()\n",
    "all_ct = [ct for ct in all_ct if ct not in ['Low Confidence']]\n",
    "high_abundance_ct = ['Erythroid', 'GMP', 'Late Myeloid', 'Neutrophil', 'PC']\n",
    "low_abundance_ct = [ct for ct in all_ct if ct not in high_abundance_ct]\n",
    "high_ct_counts = ct_counts[ct_counts['ct'].isin(high_abundance_ct)]\n",
    "low_ct_counts = ct_counts[ct_counts['ct'].isin(low_abundance_ct)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_utils.plot_multigroup_boxplot_with_significance(high_ct_counts, 'ct', 'pct', 'Collection', show_swarm = False,\n",
    "                                                    palette = 'Set2', figsize=(6, 4), xlabel='Cell type', ylabel='Percentage of cells', title=None, save_path=None) \n",
    "plt.xticks(rotation=45, ha='right')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_utils.plot_multigroup_boxplot_with_significance(low_ct_counts, 'ct', 'pct', 'Collection', show_swarm = False,\n",
    "                                                    palette = 'Set2', figsize=(12, 4), xlabel='Cell type', ylabel='Percentage of cells', title=None, save_path=None) \n",
    "plt.xticks(rotation=45, ha='right')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Observations\n",
    "- Immune Compartment\n",
    "    - Plasma cells take over the bone marrow in NDMM and revert to normal ranges after treatment. \n",
    "    - In compensation, erythroid, GMP, late myeloid, neutrophils proportions decreased in NDMM. In normal bone marrow, these cells account for 60-70% of the bone marrow.\n",
    "    - B cells and their progenitors decreased in proportion in NDMM, but were restored after stem cell transplant in PT.\n",
    "    - Macrophages increased in proportions in NDMM and were further increased in PT. Monocytes, on the other hand, decreased in proportions from normal bone marrow.\n",
    "    - CD4 T cells drastically decreased in proportions in NDMM and is further depleted in PT. CD8 T cells and NK cells show a higher proportion in NDMM and slightly decreased in PT.\n",
    "- Stromal Compartment\n",
    "    - Overall increased in proportions in NDMM and PT, including MSCs, endothelial eclls, and adipocytes\n",
    "\n",
    "These observations give rise to a few biological hypotheses, which will be tested in the following sections\n",
    "\n",
    "1. Plasma cells out-compete B cells and their progenitors for survival signals, leading to their proliferation and inhibits the normal B-to-PC transformation. This is restored post stem cell transplant.\n",
    "    - These signals may include APRIL/BAFF signaling from myeloid cells and CXCL signaling from stromal cells (MSCs).\n",
    "2. The post-transplant immune microenvironment are still dysfunctional, such as the elevated macrophage proportions and decreased CD4 T cells.\n",
    "    - These macrophages may still exhibit immunosuppressive capabilities that induces T cell exhaustion."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 2: APRIL/BAFF signaling between myeloid and B cells/plasma cells\n",
    "\n",
    "APRIL pathway: Myeloid cells expressing TNFSF13 (APRIL) and plasma cells expressing TNFRSF17 (BCMA), TNFRSF13B (TACI)\n",
    "\n",
    "BAFF pathway: Myeloid cells expressing TNFSF13B (BAFF) and plasma cells expressing TNFRSF13B (TACI), TNFRSF13C (BAFF-R)\n",
    "\n",
    "**Note**: Many of these genes are only present in the v6 custom panel."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Expression of APRIL/BAFF receptor/ligands across cell types and timepoints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_v6 = ird_xenium_merge[ird_xenium_merge.obs['Panel'] == 'BYGXJ6_hMulti', :].copy()\n",
    "ird_v6.obs['ct_timepoint'] = ird_v6.obs.apply(lambda x: f\"{x['ct']}_{x['Collection']}\", axis = 1)\n",
    "sc.pl.dotplot(ird_v6, var_names = ['TNFSF13', 'TNFSF13B', 'TNFRSF17', 'TNFRSF13B', 'TNFRSF13C'], groupby = 'ct_timepoint', standard_scale = 'var', dot_max = 0.5, cmap = 'Reds', swap_axes = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Observations:\n",
    "- APRIL expression is highest on monocytes, but also present on GMP/Late myeloid/neutrophil/pDCs. APRIL expression decrease in NDMM and increase in PT.\n",
    "- BCMA/TACI expression on plasma cells increase in NDMM and decrease in PT. They are also seen to be expressed in all other cell types in NDMM, most likely due to transcript leakage.\n",
    "- BAFF expression is also highest on monocytes, but the expression is not as high on GMP/Late myeloid/neutrophils. Interestingly there is some expression on HSPCs. BAFF seems to increase during PT.\n",
    "- BAFF-R is expressed in Mature B cells and at lower levels on early B cells.\n",
    "\n",
    "Hypothesis:\n",
    "**Plasma cells in NDMM hijacks the survival APRIL/BAFF signaling for early/mature B cells**\n",
    "\n",
    "Are plasma cells closer to APRIL+/BAFF+ cells in NDMM?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot mean expression of APRIL/BAFF receptor/ligands in key cell types across timepoints"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distance between (APRIL+) myeloid cells and (BCMA+) plasma cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rn_obj = sc.read_h5ad(\"/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/Xenium/analysis/radial_neighborhoods/Output/merged_RN.h5ad\")\n",
    "ird_v6 = rn_obj[rn_obj.obs['Panel'] == 'BYGXJ6_hMulti'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter out cells with unassigned radial neighborhood -> these are cells in low density regions\n",
    "ird_v6 = ird_v6[ird_v6.obs['rn'] != 'Unassigned']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "april_df = ird_v6[:, 'TNFSF13'].to_df()\n",
    "baff_df = ird_v6[:, 'TNFSF13B'].to_df()\n",
    "ird_v6.obs.loc[:, 'APRILpos'] = april_df['TNFSF13'] > 0\n",
    "ird_v6.obs.loc[:, 'APRIL_exp'] = april_df['TNFSF13'].values\n",
    "ird_v6.obs.loc[:, 'BAFFpos'] = baff_df['TNFSF13B'] > 0\n",
    "\n",
    "april_ligand_df = ird_v6[:, ['TNFRSF17', 'TNFRSF13B']].to_df()\n",
    "baff_ligand_df = ird_v6[:, ['TNFRSF13B', 'TNFRSF13C']].to_df()\n",
    "ird_v6.obs.loc[:, 'APRIL_receptor_pos'] = (april_ligand_df['TNFRSF17'] > 0) | (april_ligand_df['TNFRSF13B'] > 0)\n",
    "ird_v6.obs.loc[:, 'BAFF_receptor_pos'] = (baff_ligand_df['TNFRSF13B'] > 0) | (baff_ligand_df['TNFRSF13C'] > 0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_cells_info = ird_v6.obs.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ird_april_myeloid = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'].isin(['Granulo.', 'Mc/Mp', 'cDC', 'pDC'])) & (ird_xenium_merge.obs['APRILpos'] == True)]\n",
    "#ird_april_pc = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'] == 'PC') & (ird_xenium_merge.obs['APRIL_receptor_pos'] == True)]\n",
    "#ird_april_b = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'] == 'B') & (ird_xenium_merge.obs['APRIL_receptor_pos'] == True)]\n",
    "\n",
    "#ird_baff_myeloid = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'].isin(['Granulo.', 'Mc/Mp', 'cDC', 'pDC'])) & (ird_xenium_merge.obs['BAFFpos'] == True)]\n",
    "#ird_baff_pc = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'] == 'PC') & (ird_xenium_merge.obs['BAFF_receptor_pos'] == True)]\n",
    "#ird_baff_b = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'] == 'B') & (ird_xenium_merge.obs['BAFF_receptor_pos'] == True)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Nearest distance between all myeloid cells to all PC/B cells\n",
    "\n",
    "Since the detection of genes by Xenium probes can be sparse, we first consider all myeloid cells to be the source of APRIL expression and plasma cell/B cell as the source of APRIL receptor (BCMA/TACI) expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myeloid_pos = ird_cells_info[ird_cells_info['ct'].isin(['GMP', 'Late Myeloid', 'Neutrophil', 'Monocyte', 'Macrophage','cDC', 'pDC'])][['Sample', 'x_centroid', 'y_centroid']]\n",
    "plasma_pos = ird_cells_info[ird_cells_info['ct'] == 'PC'][['Sample', 'x_centroid', 'y_centroid']]\n",
    "b_pos = ird_cells_info[ird_cells_info['ct'] == 'Mature B'][['Sample', 'x_centroid', 'y_centroid']]\n",
    "earlyb_pos = ird_cells_info[ird_cells_info['ct'] == 'Early B'][['Sample', 'x_centroid', 'y_centroid']]\n",
    "mye_pc_dist = spatial_utils.nearest_dist_between_two_celltypes(myeloid_pos, plasma_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "mye_b_dist = spatial_utils.nearest_dist_between_two_celltypes(myeloid_pos, b_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "mye_earlyb_dist = spatial_utils.nearest_dist_between_two_celltypes(myeloid_pos, earlyb_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mye_pc_dist['ref_ct'] = 'Plasma cell'\n",
    "mye_b_dist['ref_ct'] = 'Mature B cell'\n",
    "mye_earlyb_dist['ref_ct'] = 'Early B cell'\n",
    "mye_dist = pd.concat([mye_pc_dist, mye_b_dist, mye_earlyb_dist], axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import gaussian_kde\n",
    "def get_most_likely_dist(sample_ref_df, dist_col, dist_eval):\n",
    "    kde = gaussian_kde(sample_ref_df[dist_col])\n",
    "    kde_pdf = kde.pdf(dist_eval)\n",
    "    most_likely_dist = dist_eval[np.argmax(kde_pdf)]\n",
    "    return most_likely_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = []\n",
    "for sample in mye_dist['Sample'].unique():\n",
    "    sample_df = mye_dist[mye_pc_dist['Sample'] == sample]\n",
    "    for ref_ct in sample_df['ref_ct'].unique():\n",
    "        sample_ref_df = sample_df[sample_df['ref_ct'] == ref_ct]\n",
    "        if len(sample_ref_df) < 10:\n",
    "            continue\n",
    "\n",
    "        res.append({\n",
    "            'Sample': sample,\n",
    "            'Reference Cell Type': ref_ct,\n",
    "            'Most Likely Distance': get_most_likely_dist(sample_ref_df, 'nearest_dist_to_df2', dist_eval),\n",
    "            'Average Distance': sample_ref_df['nearest_dist_to_df2'].mean(),\n",
    "            'Median Distance': sample_ref_df['nearest_dist_to_df2'].median()\n",
    "        })\n",
    "\n",
    "most_likely_dists_df = pd.DataFrame(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add metadata information\n",
    "metadata = ird_cells_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "most_likely_dists_df['Collection'] = most_likely_dists_df['Sample'].map(metadata['Collection'])\n",
    "most_likely_dists_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "timecols = {\"NBM\": \"#0C7515\", \"NDMM\": \"#E619B9\", \"PT\": \"#CF99C3\"} \n",
    "_ = plot_utils.plot_multigroup_boxplot_with_significance(most_likely_dists_df, 'Reference Cell Type', 'Most Likely Distance', 'Collection', show_swarm = True,\n",
    "                                                    palette = timecols, figsize=(8, 6), xlabel='Cell type', ylabel='Distance from myeloid cells', title=None, save_path=None) \n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "#plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_myeloid_nearestDist_to_PC_B_cells_boxplot_v11062025.pdf', dpi = 300, transparent = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Randomly shuffle cell types to test for significant spatial association"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Randomly shuffle cell type labels within each sample to create a null distribution\n",
    "num_shuffles = 10\n",
    "dist_eval = np.arange(0, 501, 1)\n",
    "res = []\n",
    "for sample in ird_cells_info.loc[:, 'Sample'].unique().tolist():\n",
    "    sample_df = ird_cells_info[ird_cells_info['Sample'] == sample]\n",
    "    \n",
    "    # Observed data\n",
    "    myeloid_pos = sample_df[sample_df['ct'].isin(['GMP', 'Late Myeloid', 'Neutrophil', 'Monocyte', 'Macrophage','cDC', 'pDC'])][['Sample', 'x_centroid', 'y_centroid']]\n",
    "    plasma_pos = sample_df[sample_df['ct'] == 'Early B'][['Sample', 'x_centroid', 'y_centroid']]\n",
    "    mye_pc_dist = spatial_utils.nearest_dist_between_two_celltypes(myeloid_pos, plasma_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "    observed_dist = get_most_likely_dist(mye_pc_dist, 'nearest_dist_to_df2', dist_eval)\n",
    "    \n",
    "    # Shuffled data\n",
    "    shuffled_dists = []\n",
    "    for i in range(num_shuffles):\n",
    "        sample_df.loc[:, 'shuffled_ct'] = sample_df['ct'].sample(frac=1, random_state = i).values\n",
    "        myeloid_pos_shuffled = sample_df[sample_df['shuffled_ct'].isin(['GMP', 'Late Myeloid', 'Neutrophil', 'Monocyte', 'Macrophage','cDC', 'pDC'])][['Sample', 'x_centroid', 'y_centroid']]\n",
    "        plasma_pos_shuffled = sample_df[sample_df['shuffled_ct'] == 'Early B'][['Sample', 'x_centroid', 'y_centroid']]\n",
    "        dist_shuffled = spatial_utils.nearest_dist_between_two_celltypes(myeloid_pos_shuffled, plasma_pos_shuffled, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "        shuffled_dists.append(get_most_likely_dist(dist_shuffled, 'nearest_dist_to_df2', dist_eval))\n",
    "    mean_shuffled_dist = np.mean(shuffled_dists)\n",
    "    \n",
    "    sample_res = {\n",
    "        'Sample': sample,\n",
    "        'Observed Distance': observed_dist,\n",
    "        'Shuffled Distances': mean_shuffled_dist\n",
    "    }\n",
    "    res.append(sample_res)\n",
    "\n",
    "shuffled_dists_df = pd.DataFrame(res)\n",
    "shuffled_dists_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shuffled_dists_df_melt = shuffled_dists_df.melt(id_vars = ['Sample'], value_vars = ['Observed Distance', 'Shuffled Distances'], var_name = 'Distance Type', value_name = 'Distance')\n",
    "sns.boxplot(shuffled_dists_df_melt, x = 'Distance Type', y = 'Distance', hue = 'Distance Type')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()\n",
    "#plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_myeloid_nearestDist_to_PC_B_cells_boxplot_v11062025.pdf', dpi = 300, transparent = True)\n",
    "#plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = ird_cells_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "shuffled_dists_df_melt['Collection'] = shuffled_dists_df_melt['Sample'].map(metadata['Collection'])\n",
    "sns.boxplot(shuffled_dists_df_melt, x = 'Collection', y = 'Distance', hue = 'Distance Type')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_sample = 'S13-47674A1U1'\n",
    "sample_df = ird_cells_info[ird_cells_info['Sample'] == test_sample]\n",
    "myeloid_pos = sample_df[sample_df['ct'].isin(['GMP', 'Late Myeloid', 'Neutrophil', 'Monocyte', 'Macrophage','cDC', 'pDC'])][['Sample', 'x_centroid', 'y_centroid']]\n",
    "plasma_pos = sample_df[sample_df['ct'] == 'Early B'][['Sample', 'x_centroid', 'y_centroid']]\n",
    "mye_pc_dist = spatial_utils.nearest_dist_between_two_celltypes(myeloid_pos, plasma_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "mye_pc_dist.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_df.loc[:, 'shuffled_ct'] = sample_df['ct'].sample(frac=1, random_state = i).values\n",
    "myeloid_pos_shuffled = sample_df[sample_df['shuffled_ct'].isin(['GMP', 'Late Myeloid', 'Neutrophil', 'Monocyte', 'Macrophage','cDC', 'pDC'])][['Sample', 'x_centroid', 'y_centroid']]\n",
    "plasma_pos_shuffled = sample_df[sample_df['shuffled_ct'] == 'Early B'][['Sample', 'x_centroid', 'y_centroid']]\n",
    "dist_shuffled = spatial_utils.nearest_dist_between_two_celltypes(myeloid_pos_shuffled, plasma_pos_shuffled, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_most_likely_dist(mye_pc_dist, 'nearest_dist_to_df2', dist_eval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_most_likely_dist(dist_shuffled, 'nearest_dist_to_df2', dist_eval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(mye_pc_dist, x = 'nearest_dist_to_df2', binwidth = 1, stat = 'density', common_norm = False, alpha = .5)\n",
    "sns.histplot(dist_shuffled, x = 'nearest_dist_to_df2', binwidth = 1, stat = 'density', common_norm = False, alpha = .5)\n",
    "plt.xlim(0, 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shuffled_dists_df['Collection'] = shuffled_dists_df['Sample'].map(metadata['Collection'])\n",
    "shuffled_dists_df[shuffled_dists_df['Outlier'] == True]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Nearest distance between APRIL+ myeloid and APRIL_ligand+ PC/B cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_april_myeloid = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'].isin(['Granulo.', 'Mc/Mp', 'cDC', 'pDC'])) & (ird_xenium_merge.obs['APRILpos'] == True)]\n",
    "ird_april_pc = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'] == 'PC') & (ird_xenium_merge.obs['APRIL_receptor_pos'] == True)]\n",
    "ird_april_b = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'] == 'B') & (ird_xenium_merge.obs['APRIL_receptor_pos'] == True)]\n",
    "\n",
    "ird_baff_myeloid = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'].isin(['Granulo.', 'Mc/Mp', 'cDC', 'pDC'])) & (ird_xenium_merge.obs['BAFFpos'] == True)]\n",
    "ird_baff_pc = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'] == 'PC') & (ird_xenium_merge.obs['BAFF_receptor_pos'] == True)]\n",
    "ird_baff_b = ird_xenium_merge.obs[(ird_xenium_merge.obs['annot'] == 'B') & (ird_xenium_merge.obs['BAFF_receptor_pos'] == True)]\n",
    "\n",
    "myeloid_april_pos = ird_april_myeloid[['x_centroid', 'y_centroid', 'Sample']]\n",
    "plasma_april_pos = ird_april_pc[['x_centroid', 'y_centroid', 'Sample']]\n",
    "bcell_april_pos = ird_april_b[['x_centroid', 'y_centroid', 'Sample']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "april_pc_dist = nearest_dist_between_two_celltypes(myeloid_april_pos, plasma_april_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "april_b_dist = nearest_dist_between_two_celltypes(myeloid_april_pos, bcell_april_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "april_pc_dist['ref_ct'] = 'Plasma cell'\n",
    "april_b_dist['ref_ct'] = 'B cell'\n",
    "april_dists = pd.concat([april_pc_dist, april_b_dist], axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "april_dists.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = ird_cell_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "april_dists['Collection'] = april_dists['Sample'].map(metadata['Collection'])\n",
    "april_dists['Collection'] = april_dists['Collection'].astype('str')\n",
    "april_dists['Condition'] = april_dists['Collection'] + '-' + april_dists['ref_ct']\n",
    "april_dists.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(april_dists[april_dists['Collection']=='NDMM'], x = 'nearest_dist_to_df2', hue = 'Condition', binwidth = 5, stat = 'density', common_norm = False)\n",
    "plt.xlim(0, 200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(april_dists[april_dists['Collection']=='NBM'], x = 'nearest_dist_to_df2', hue = 'Condition', binwidth = 5, stat = 'density', common_norm = False)\n",
    "plt.xlim(0, 200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(april_dists[april_dists['Collection']=='PT'], x = 'nearest_dist_to_df2', hue = 'Condition', binwidth = 5, stat = 'density', common_norm = False)\n",
    "plt.xlim(0, 200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Calculate most likely distance using KDE, on a per sample basis\n",
    "from scipy.stats import gaussian_kde\n",
    "\n",
    "res = []\n",
    "for sample in april_dists['Sample'].unique():\n",
    "    sample_df = april_dists[april_dists['Sample'] == sample]\n",
    "    for ref_ct in sample_df['ref_ct'].unique():\n",
    "        sample_ref_df = sample_df[sample_df['ref_ct'] == ref_ct]\n",
    "        if len(sample_ref_df) < 10:\n",
    "            continue\n",
    "        # Estimate distance using KDE\n",
    "        kde = gaussian_kde(sample_ref_df['nearest_dist_to_df2'])\n",
    "        dist_eval = np.arange(0, 501, 1)\n",
    "        kde_pdf = kde.pdf(dist_eval)\n",
    "        most_likely_dist = dist_eval[np.argmax(kde_pdf)]\n",
    "        res.append({\n",
    "            'Sample': sample,\n",
    "            'Reference Cell Type': ref_ct,\n",
    "            'Most Likely Distance': most_likely_dist,\n",
    "            'Average Distance': sample_ref_df['nearest_dist_to_df2'].mean(),\n",
    "            'Median Distance': sample_ref_df['nearest_dist_to_df2'].median()\n",
    "        })\n",
    "\n",
    "most_likely_dists_df = pd.DataFrame(res)\n",
    "most_likely_dists_df['Collection'] = most_likely_dists_df['Sample'].map(metadata['Collection'])\n",
    "most_likely_dists_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot by comparing distance between cell types within the same timepoint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_name = \"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_APRILpairs_boxplot_v10242025.pdf\"\n",
    "with PdfPages(pdf_name) as pdf:\n",
    "    fig1 = sns.boxplot(most_likely_dists_df, x = 'Collection', y = 'Most Likely Distance', hue = 'Reference Cell Type')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Collection', y = 'Most Likely Distance', hue = 'Reference Cell Type', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from APRIL+ myeloid cells to BCMA/TACI+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()\n",
    "\n",
    "    fig2 = sns.boxplot(most_likely_dists_df, x = 'Collection', y = 'Average Distance', hue = 'Reference Cell Type')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Collection', y = 'Average Distance', hue = 'Reference Cell Type', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from APRIL+ myeloid cells to BCMA/TACI+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()\n",
    "\n",
    "    fig3 = sns.boxplot(most_likely_dists_df, x = 'Collection', y = 'Median Distance', hue = 'Reference Cell Type')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Collection', y = 'Median Distance', hue = 'Reference Cell Type', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from APRIL+ myeloid cells to BCMA/TACI+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stats test between B cells and plasma cell distances in each condition\n",
    "stats_res = []\n",
    "dist_metrics = ['Most Likely Distance', 'Average Distance', 'Median Distance']\n",
    "for condition in most_likely_dists_df['Collection'].unique():\n",
    "    condition_df = most_likely_dists_df[most_likely_dists_df['Collection'] == condition]\n",
    "\n",
    "    for ct1, ct2 in combinations(condition_df['Reference Cell Type'].unique(), 2):\n",
    "        for metric in dist_metrics:\n",
    "            ct1_dists = condition_df[condition_df['Reference Cell Type'] == ct1][metric].values.tolist()\n",
    "            ct2_dists = condition_df[condition_df['Reference Cell Type'] == ct2][metric].values.tolist()\n",
    "            if len(ct1_dists) > 0 and len(ct2_dists) > 0:\n",
    "                u_stat, p_val = mannwhitneyu(ct1_dists, ct2_dists, alternative='two-sided')\n",
    "                stats_res.append({\n",
    "                    'Condition': condition,\n",
    "                    'Cell Type 1': ct1,\n",
    "                    'Cell Type 2': ct2,\n",
    "                    'Distance Metric': metric,\n",
    "                    'U_statistic': u_stat,\n",
    "                    'p_value': p_val\n",
    "                })\n",
    "stats_res_df = pd.DataFrame(stats_res)\n",
    "stats_res_df['p_adj'] = multipletests(stats_res_df['p_value'], method='fdr_bh')[1]\n",
    "stats_res_df.to_csv(\"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_APRILpairs_stats_v10242025.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot by comparing distance between timepoint within the same cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_name = \"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_APRILpairs_boxplot_byTimepoint_v10242025.pdf\"\n",
    "with PdfPages(pdf_name) as pdf:\n",
    "    fig = sns.boxplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Most Likely Distance', hue = 'Collection')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Most Likely Distance', hue = 'Collection', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from APRIL+ myeloid cells to BCMA/TACI+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()\n",
    "\n",
    "    fig = sns.boxplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Average Distance', hue = 'Collection')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Average Distance', hue = 'Collection', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from APRIL+ myeloid cells to BCMA/TACI+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()\n",
    "\n",
    "    fig = sns.boxplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Median Distance', hue = 'Collection')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Median Distance', hue = 'Collection', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from APRIL+ myeloid cells to BCMA/TACI+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stats test between timepoints in each reference cell type\n",
    "from scipy.stats import mannwhitneyu\n",
    "from itertools import combinations\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "stats_res = []\n",
    "dist_metrics = ['Most Likely Distance', 'Average Distance', 'Median Distance']\n",
    "for ct in most_likely_dists_df['Reference Cell Type'].unique():\n",
    "    conditions = most_likely_dists_df['Collection'].unique()\n",
    "    ct_df = most_likely_dists_df[most_likely_dists_df['Reference Cell Type'] == ct]\n",
    "\n",
    "    for cond1, cond2 in combinations(conditions, 2):\n",
    "        for metric in dist_metrics:\n",
    "            cond1_dists = ct_df[ct_df['Collection'] == cond1][metric].values.tolist()\n",
    "            cond2_dists = ct_df[ct_df['Collection'] == cond2][metric].values.tolist()\n",
    "            if len(cond1_dists) > 0 and len(cond2_dists) > 0:\n",
    "                u_stat, p_val = mannwhitneyu(cond1_dists, cond2_dists, alternative='two-sided')\n",
    "                stats_res.append({'Condition 1': cond1, 'Condition 2': cond2, 'Reference Cell Type': ct, 'Distance Metric': metric, 'U statistic': u_stat, 'p-value': p_val})\n",
    "stats_df = pd.DataFrame(stats_res)\n",
    "stats_df['p_adj'] = multipletests(stats_df['p-value'], method='fdr_bh')[1]\n",
    "stats_df.to_csv(\"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_APRILpairs_byTimepoints_stats_v10242025.csv\")\n",
    "stats_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Nearest distance between BAFF+ myeloid vs BAFF_ligand+ PC/B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myeloid_baff_pos = ird_baff_myeloid[['x_centroid', 'y_centroid', 'Sample']]\n",
    "plasma_baff_pos = ird_baff_pc[['x_centroid', 'y_centroid', 'Sample']]\n",
    "bcell_baff_pos = ird_baff_b[['x_centroid', 'y_centroid', 'Sample']]\n",
    "baff_pc_dist = nearest_dist_between_two_celltypes(myeloid_baff_pos, plasma_baff_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "baff_b_dist = nearest_dist_between_two_celltypes(myeloid_baff_pos, bcell_baff_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baff_pc_dist['ref_ct'] = 'Plasma cell'\n",
    "baff_b_dist['ref_ct'] = 'B cell'\n",
    "baff_dists = pd.concat([baff_pc_dist, baff_b_dist], axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = ird_cell_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "baff_dists['Collection'] = baff_dists['Sample'].map(metadata['Collection'])\n",
    "baff_dists['Collection'] = baff_dists['Collection'].astype('str')\n",
    "baff_dists['Condition'] = baff_dists['Collection'] + '-' + baff_dists['ref_ct']\n",
    "baff_dists.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(baff_dists[baff_dists['Collection']=='NDMM'], x = 'nearest_dist_to_df2', hue = 'Condition', binwidth = 5, stat = 'density', common_norm = False)\n",
    "plt.xlim(0, 200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(baff_dists[baff_dists['Collection']=='NBM'], x = 'nearest_dist_to_df2', hue = 'Condition', binwidth = 5, stat = 'density', common_norm = False)\n",
    "plt.xlim(0, 200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(baff_dists[baff_dists['Collection']=='PT'], x = 'nearest_dist_to_df2', hue = 'Condition', binwidth = 5, stat = 'density', common_norm = False)\n",
    "plt.xlim(0, 200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Calculate most likely distance using KDE, on a per sample basis\n",
    "from scipy.stats import gaussian_kde\n",
    "\n",
    "res = []\n",
    "for sample in baff_dists['Sample'].unique():\n",
    "    sample_df = baff_dists[baff_dists['Sample'] == sample]\n",
    "    for ref_ct in sample_df['ref_ct'].unique():\n",
    "        sample_ref_df = sample_df[sample_df['ref_ct'] == ref_ct]\n",
    "        if len(sample_ref_df) < 10:\n",
    "            continue\n",
    "        kde = gaussian_kde(sample_ref_df['nearest_dist_to_df2'])\n",
    "        dist_eval = np.arange(0, 501, 1)\n",
    "        kde_pdf = kde.pdf(dist_eval)\n",
    "        most_likely_dist = dist_eval[np.argmax(kde_pdf)]\n",
    "        res.append({\n",
    "            'Sample': sample,\n",
    "            'Reference Cell Type': ref_ct,\n",
    "            'Most Likely Distance': most_likely_dist,\n",
    "            'Average Distance': sample_ref_df['nearest_dist_to_df2'].mean(),\n",
    "            'Median Distance': sample_ref_df['nearest_dist_to_df2'].median()\n",
    "        })\n",
    "\n",
    "most_likely_dists_df = pd.DataFrame(res)\n",
    "most_likely_dists_df['Collection'] = most_likely_dists_df['Sample'].map(metadata['Collection'])\n",
    "most_likely_dists_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_name = \"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_BAFFpairs_boxplot_byCelltype_v10242025.pdf\"\n",
    "with PdfPages(pdf_name) as pdf:\n",
    "    fig1 = sns.boxplot(most_likely_dists_df, x = 'Collection', y = 'Most Likely Distance', hue = 'Reference Cell Type')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Collection', y = 'Most Likely Distance', hue = 'Reference Cell Type', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from BAFF+ myeloid cells to TACI/BAFF-R+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()\n",
    "\n",
    "    fig2 = sns.boxplot(most_likely_dists_df, x = 'Collection', y = 'Average Distance', hue = 'Reference Cell Type')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Collection', y = 'Average Distance', hue = 'Reference Cell Type', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from BAFF+ myeloid cells to TACI/BAFF-R+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.ylim(0, 2000)\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()\n",
    "\n",
    "    fig3 = sns.boxplot(most_likely_dists_df, x = 'Collection', y = 'Median Distance', hue = 'Reference Cell Type')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Collection', y = 'Median Distance', hue = 'Reference Cell Type', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from BAFF+ myeloid cells to TACI/BAFF-R+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.ylim(0, 2000)\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stats_res = []\n",
    "dist_metrics = ['Most Likely Distance', 'Average Distance', 'Median Distance']\n",
    "for condition in most_likely_dists_df['Collection'].unique():\n",
    "    condition_df = most_likely_dists_df[most_likely_dists_df['Collection'] == condition]\n",
    "\n",
    "    for ct1, ct2 in combinations(condition_df['Reference Cell Type'].unique(), 2):\n",
    "        for metric in dist_metrics:\n",
    "            ct1_dists = condition_df[condition_df['Reference Cell Type'] == ct1][metric].values.tolist()\n",
    "            ct2_dists = condition_df[condition_df['Reference Cell Type'] == ct2][metric].values.tolist()\n",
    "            if len(ct1_dists) > 0 and len(ct2_dists) > 0:\n",
    "                u_stat, p_val = mannwhitneyu(ct1_dists, ct2_dists, alternative='two-sided')\n",
    "                stats_res.append({\n",
    "                    'Condition': condition,\n",
    "                    'Cell Type 1': ct1,\n",
    "                    'Cell Type 2': ct2,\n",
    "                    'Distance Metric': metric,\n",
    "                    'U_statistic': u_stat,\n",
    "                    'p_value': p_val\n",
    "                })\n",
    "stats_res_df = pd.DataFrame(stats_res)\n",
    "stats_res_df['p_adj'] = multipletests(stats_res_df['p_value'], method='fdr_bh')[1]\n",
    "stats_res_df.to_csv(\"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_BAFFpairs_byCelltype_stats_v10242025.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_name = \"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_BAFFpairs_boxplot_byTimepoint_v10242025.pdf\"\n",
    "with PdfPages(pdf_name) as pdf:\n",
    "    fig = sns.boxplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Most Likely Distance', hue = 'Collection')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Most Likely Distance', hue = 'Collection', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from BAFF+ myeloid cells to TACI/BAFF-R+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()\n",
    "\n",
    "    fig = sns.boxplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Average Distance', hue = 'Collection')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Average Distance', hue = 'Collection', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from BAFF+ myeloid cells to TACI/BAFF-R+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.ylim(0, 2000)\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()\n",
    "\n",
    "    fig = sns.boxplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Median Distance', hue = 'Collection')\n",
    "    sns.swarmplot(most_likely_dists_df, x = 'Reference Cell Type', y = 'Median Distance', hue = 'Collection', dodge = True, color = 'k', alpha = .5)\n",
    "    plt.title('Distance from BAFF+ myeloid cells to TACI/BAFF-R+ B/Plasma cells')\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "    plt.ylim(0, 2000)\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig(dpi = 300, transparent = True)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stats test between timepoints in each reference cell type\n",
    "from scipy.stats import mannwhitneyu\n",
    "from itertools import combinations\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "stats_res = []\n",
    "dist_metrics = ['Most Likely Distance', 'Average Distance', 'Median Distance']\n",
    "for ct in most_likely_dists_df['Reference Cell Type'].unique():\n",
    "    conditions = most_likely_dists_df['Collection'].unique()\n",
    "    ct_df = most_likely_dists_df[most_likely_dists_df['Reference Cell Type'] == ct]\n",
    "\n",
    "    for cond1, cond2 in combinations(conditions, 2):\n",
    "        for metric in dist_metrics:\n",
    "            cond1_dists = ct_df[ct_df['Collection'] == cond1][metric].values.tolist()\n",
    "            cond2_dists = ct_df[ct_df['Collection'] == cond2][metric].values.tolist()\n",
    "            if len(cond1_dists) > 0 and len(cond2_dists) > 0:\n",
    "                u_stat, p_val = mannwhitneyu(cond1_dists, cond2_dists, alternative='two-sided')\n",
    "                stats_res.append({'Condition 1': cond1, 'Condition 2': cond2, 'Reference Cell Type': ct, 'Distance Metric': metric, 'U statistic': u_stat, 'p-value': p_val})\n",
    "stats_df = pd.DataFrame(stats_res)\n",
    "stats_df['p_adj'] = multipletests(stats_df['p-value'], method='fdr_bh')[1]\n",
    "stats_df.to_csv(\"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_BAFFpairs_byTimepoint_stats_v10242025.csv\")\n",
    "stats_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count number of BAFF+ myeloid cells within 200um of BAFF_ligand+ plasma cells, per sample\n",
    "proximal_threshold = 200\n",
    "proximal_counts = res1[res1['nearest_dist_to_df2'] <= proximal_threshold].groupby(['Sample']).size()\n",
    "proximal_counts = proximal_counts[proximal_counts > 0].to_frame('Counts')\n",
    "# Normalize by total number of myeloid cells per sample\n",
    "total_myeloid_counts = myeloid_info['Sample'].value_counts()\n",
    "proximal_counts = proximal_counts.join(total_myeloid_counts.rename('total_myeloid'))\n",
    "proximal_counts['frac_proximal'] = proximal_counts['Counts'] / proximal_counts['total_myeloid']\n",
    "proximal_counts = proximal_counts.join(metadata, how='left')\n",
    "sns.boxplot(proximal_counts, x = 'Collection', y = 'frac_proximal', showfliers = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### APRIL receptor expression vs distance in PC/B cells vs APRIL+ cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### APRIL expression in myeloid cells with distance from plasma cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "april_pos = ird_v6[(ird_v6.obs['APRILpos'] == True) & (ird_v6.obs['ct'].isin(['GMP', 'Late Myeloid', 'Neutrophil', 'Monocyte', 'Macrophage','cDC', 'pDC'])), :]\n",
    "#april_neg_pos = ird_v6[ird_v6.obs['APRILpos'] == False, :]\n",
    "ird_pc = ird_v6[ird_v6.obs['ct'] == 'PC', :]\n",
    "ird_b = ird_v6[(ird_v6.obs['ct'] == 'Mature B'), :]\n",
    "april_dist_to_PC = spatial_utils.nearest_dist_between_two_celltypes(april_pos.obs.copy(), ird_pc.obs.copy(), sample_col='Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "april_dist_to_B = spatial_utils.nearest_dist_between_two_celltypes(april_pos.obs.copy(), ird_b.obs.copy(), sample_col='Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "#pc_dist_to_aprilneg = nearest_dist_between_two_celltypes(ird_pc.obs.copy(), april_neg_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "#b_dist_to_aprilneg = nearest_dist_between_two_celltypes(ird_b.obs.copy(), april_neg_pos, sample_col='Sample', x_col='x_centroid', y_col='y_centroid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = [0, 20, 50, np.inf]\n",
    "bin_labels = ['0-20', '20-50', '>50']\n",
    "april_pos.obs['dist_to_nearest_PC'] = april_dist_to_PC['nearest_dist_to_df2'].values\n",
    "april_pos.obs['dist_to_nearest_B'] = april_dist_to_B['nearest_dist_to_df2'].values\n",
    "#ird_pc.obs['dist_to_nearest_APRILneg'] = pc_dist_to_aprilneg['nearest_dist_to_df2'].values\n",
    "#ird_b.obs['dist_to_nearest_APRILneg'] = b_dist_to_aprilneg['nearest_dist_to_df2'].values\n",
    "\n",
    "april_pos.obs['dist_bin_to_nearest_PC'] = pd.cut(april_pos.obs['dist_to_nearest_PC'], \n",
    "                                                    bins = bins, \n",
    "                                                    labels = bin_labels)\n",
    "april_pos.obs['dist_bin_to_nearest_B'] = pd.cut(april_pos.obs['dist_to_nearest_B'], \n",
    "                                                    bins=bins, \n",
    "                                                    labels = bin_labels)\n",
    "#ird_pc.obs['dist_bin_to_nearest_APRILneg'] = pd.cut(ird_pc.obs['dist_to_nearest_APRILneg'], \n",
    "#                                                    bins=[0, 50, 100, 150, 200, np.inf], \n",
    "#                                                    labels = ['0-50', '50-100', '100-150', '150-200', '>200'])\n",
    "#ird_b.obs['dist_bin_to_nearest_APRILneg'] = pd.cut(ird_b.obs['dist_to_nearest_APRILneg'], \n",
    "#                                                    bins=[0, 50, 100, 150, 200, np.inf], \n",
    "#                                                    labels = ['0-50', '50-100', '100-150', '150-200', '>200'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each sample, calculate the mean expression of APRIL in each distance bin\n",
    "april_pos_pc_avg = april_pos.obs.groupby(['Sample', 'dist_bin_to_nearest_PC'])['APRIL_exp'].mean().reset_index()\n",
    "april_pos_b_avg = april_pos.obs.groupby(['Sample', 'dist_bin_to_nearest_B'])['APRIL_exp'].mean().reset_index()\n",
    "metadata = ird_cells_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "april_pos_pc_avg['Collection'] = april_pos_pc_avg['Sample'].map(metadata['Collection'])\n",
    "april_pos_b_avg['Collection'] = april_pos_b_avg['Sample'].map(metadata['Collection'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(april_pos_pc_avg, x = 'Collection', y = 'APRIL_exp', hue = 'dist_bin_to_nearest_PC')\n",
    "sns.swarmplot(april_pos_pc_avg, x = 'Collection', y = 'APRIL_exp', hue = 'dist_bin_to_nearest_PC', dodge = True, color = 'k', alpha = .5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(april_pos_b_avg, x = 'Collection', y = 'APRIL_exp', hue = 'dist_bin_to_nearest_B')\n",
    "sns.swarmplot(april_pos_b_avg, x = 'Collection', y = 'APRIL_exp', hue = 'dist_bin_to_nearest_B', dodge = True, color = 'k', alpha = .5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### BCMA expression in plasma cells with distance from APRIL+ cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter to only include cells that fall within assigned distance bins\n",
    "# Group cells by both Collection and distance bin, plot expression boxplot\n",
    "ird_pc_df = ird_pc.obs.copy()#[ird_pc.obs['dist_bin_to_nearest_APRILpos'].cat.codes != -1].copy()\n",
    "ird_b_df = ird_b.obs.copy()#[ird_b.obs['dist_bin_to_nearest_APRILpos'].cat.codes != -1].copy()\n",
    "ird_pc_aprilR_expr = ird_pc[:, ['TNFRSF17', 'TNFRSF13B']].to_df()\n",
    "ird_b_aprilR_expr = ird_b[:, ['TNFRSF17', 'TNFRSF13B']].to_df()\n",
    "ird_pc_df['BCMA_expr'] = ird_pc_aprilR_expr['TNFRSF17'].values\n",
    "ird_pc_df['TACI_expr'] = ird_pc_aprilR_expr['TNFRSF13B'].values\n",
    "ird_b_df['BCMA_expr'] = ird_b_aprilR_expr['TNFRSF17'].values\n",
    "ird_b_df['TACI_expr'] = ird_b_aprilR_expr['TNFRSF13B'].values\n",
    "#sns.violinplot(ird_pc_df, x = 'Collection', y = 'BCMA_expr', hue = 'dist_bin_to_nearest_APRILpos', inner = 'quartile')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_pc_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate number of cells in each distance bin per sample\n",
    "metadata = ird_cells_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "pc_bin_counts = ird_pc_df.groupby(['Sample', 'dist_bin_to_nearest_APRILpos']).size().unstack(fill_value=0).reset_index()\n",
    "pc_bin_counts['Collection'] = pc_bin_counts['Sample'].map(metadata['Collection'])\n",
    "pc_bin_counts = pc_bin_counts.melt(id_vars=['Sample', 'Collection'], var_name='Distance Bin', value_name='Cell Count')\n",
    "pc_bin_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(pc_bin_counts, x = 'Collection', y = 'Cell Count', hue = 'Distance Bin')  # %? cell  ### Only look at low PC% in NDM??\n",
    "plt.ylim(0, 800)\n",
    "plt.title('Number of plasma cells in each distance bin to nearest APRIL+ cell')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate average BCMA & TACI expression within each distance bin, for each sample\n",
    "ird_pc_df = ird_pc_df[ird_pc_df['dist_bin_to_nearest_APRILpos'].cat.codes != -1].copy()\n",
    "pc_bcma_avg = ird_pc_df.groupby(['Sample', 'dist_bin_to_nearest_APRILpos'])['BCMA_expr'].mean().reset_index()\n",
    "pc_taci_avg = ird_pc_df.groupby(['Sample', 'dist_bin_to_nearest_APRILpos'])['TACI_expr'].mean().reset_index()\n",
    "\n",
    "metadata = ird_cells_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "pc_bcma_avg['Collection'] = pc_bcma_avg['Sample'].map(metadata['Collection'])\n",
    "pc_taci_avg['Collection'] = pc_taci_avg['Sample'].map(metadata['Collection'])\n",
    "pc_bcma_avg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (8, 5))\n",
    "sns.boxplot(pc_bcma_avg[pc_bcma_avg['dist_bin_to_nearest_APRILpos']!='>200'], x = 'Collection', y = 'BCMA_expr', hue = 'dist_bin_to_nearest_APRILpos', showfliers = False)\n",
    "sns.swarmplot(pc_bcma_avg[pc_bcma_avg['dist_bin_to_nearest_APRILpos']!='>200'], x = 'Collection', y = 'BCMA_expr', hue = 'dist_bin_to_nearest_APRILpos', dodge = True, color = 'k', alpha = .5)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.title('Average BCMA expression in Plasma cells by distance to nearest APRIL+ cell')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (8, 5))\n",
    "sns.boxplot(pc_taci_avg[pc_taci_avg['dist_bin_to_nearest_APRILpos']!='>200'], x = 'Collection', y = 'TACI_expr', hue = 'dist_bin_to_nearest_APRILpos', showfliers = False)\n",
    "sns.swarmplot(pc_taci_avg[pc_taci_avg['dist_bin_to_nearest_APRILpos']!='>200'], x = 'Collection', y = 'TACI_expr', hue = 'dist_bin_to_nearest_APRILpos', dodge = True, color = 'k', alpha = .5)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.title('Average TACI expression in Plasma cells by distance to nearest APRIL+ cell')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b_bin_counts = ird_b_df.groupby(['Sample', 'dist_bin_to_nearest_APRILpos']).size().unstack(fill_value=0).reset_index()\n",
    "b_bin_counts['Collection'] = b_bin_counts['Sample'].map(metadata['Collection'])\n",
    "b_bin_counts = b_bin_counts.melt(id_vars=['Sample', 'Collection'], var_name='Distance Bin', value_name='Cell Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(b_bin_counts, x = 'Collection', y = 'Cell Count', hue = 'Distance Bin')\n",
    "plt.ylim(0, 1200)\n",
    "plt.title('Number of B cells in each distance bin to nearest APRIL+ cell')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_b_df = ird_b_df[ird_b_df['dist_bin_to_nearest_APRILpos'].cat.codes != -1].copy()\n",
    "b_bcma_avg = ird_b_df.groupby(['Sample', 'dist_bin_to_nearest_APRILpos'])['BCMA_expr'].mean().reset_index()\n",
    "b_taci_avg = ird_b_df.groupby(['Sample', 'dist_bin_to_nearest_APRILpos'])['TACI_expr'].mean().reset_index()\n",
    "\n",
    "metadata = ird_cells_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "b_bcma_avg['Collection'] = b_bcma_avg['Sample'].map(metadata['Collection'])\n",
    "b_taci_avg['Collection'] = b_taci_avg['Sample'].map(metadata['Collection'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (8, 5))\n",
    "sns.boxplot(b_bcma_avg, x = 'Collection', y = 'BCMA_expr', hue = 'dist_bin_to_nearest_APRILpos', showfliers = False)\n",
    "sns.swarmplot(b_bcma_avg, x = 'Collection', y = 'BCMA_expr', hue = 'dist_bin_to_nearest_APRILpos', dodge = True, color = 'k', alpha = .5)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.title('Average BCMA expression in B cells by distance to nearest APRIL+ cell')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (8, 5))\n",
    "sns.boxplot(b_taci_avg, x = 'Collection', y = 'TACI_expr', hue = 'dist_bin_to_nearest_APRILpos', showfliers = False)\n",
    "sns.swarmplot(b_taci_avg, x = 'Collection', y = 'TACI_expr', hue = 'dist_bin_to_nearest_APRILpos', dodge = True, color = 'k', alpha = .5)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.title('Average TACI expression in B cells by distance to nearest APRIL+ cell')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploring a ligand-receptor score based on ligand expression, receptor expression, and distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial import KDTree\n",
    "def compute_pairwise_LR_scores(adata, ligand_gene, receptor_gene,\n",
    "                               distance_threshold=200, dist_lambda = 50,\n",
    "                               expr_layer = None, celltype_col='annot', x_col='x_centroid', y_col='y_centroid'):\n",
    "    \"\"\"\n",
    "    DOCS (This is per sample)\n",
    "    \"\"\"\n",
    "    # 1) Get coordinates and expression\n",
    "    coords = adata.obs[[x_col, y_col]].copy().values\n",
    "    if expr_layer is None:\n",
    "        L_exp = adata[:, ligand_gene].X.toarray().flatten()\n",
    "        R_exp = adata[:, receptor_gene].X.toarray().flatten()\n",
    "    else:\n",
    "        L_exp = adata[:, ligand_gene].layers[expr_layer].toarray().flatten()\n",
    "        R_exp = adata[:, receptor_gene].layers[expr_layer].toarray().flatten()\n",
    "\n",
    "    # Find cells expressing ligand and receptor\n",
    "    ligand_idx = np.where(L_exp > 0)[0]\n",
    "    receptor_idx = np.where(R_exp > 0)[0]\n",
    "\n",
    "    # If no ligand or receptor expressing cells, return empty dataframe\n",
    "    if len(ligand_idx) == 0 or len(receptor_idx) == 0:\n",
    "        print(f\"No ligand or receptor expressing cells for {ligand_gene}-{receptor_gene} in this sample.\")\n",
    "        return pd.DataFrame(), pd.DataFrame()\n",
    "\n",
    "    # 2) Compute ligand receptor pairs within distance threshold\n",
    "    # Build KDTree for receptor cells\n",
    "    tree = KDTree(coords[receptor_idx])\n",
    "    # For all ligand points, find all points within distance threshold\n",
    "    results = tree.query_ball_point(coords[ligand_idx], r=distance_threshold)\n",
    "\n",
    "    # 3) Compute scores for each ligand-receptor pair\n",
    "    score_results = []\n",
    "    for li_idx, neighbors in enumerate(results):\n",
    "        ligand_cell_idx = ligand_idx[li_idx]\n",
    "        if len(neighbors) == 0:\n",
    "            continue\n",
    "        # Vectorized distance calculation\n",
    "        neighbor_global_idx = receptor_idx[neighbors]\n",
    "        dists = np.linalg.norm(coords[ligand_cell_idx] - coords[neighbor_global_idx], axis=1)\n",
    "        Li = L_exp[ligand_cell_idx]\n",
    "        Rj = R_exp[neighbor_global_idx]\n",
    "        K = np.exp(-dists / dist_lambda)\n",
    "        score_ij = Li * Rj * K\n",
    "        # Store results\n",
    "        for r_idx, dist, L_exp_val, R_exp_val, score in zip(neighbor_global_idx, dists, \n",
    "                                                        [Li]*len(neighbor_global_idx), \n",
    "                                                        Rj, score_ij):\n",
    "            score_results.append({\n",
    "                'Ligand Cell Index': ligand_cell_idx,\n",
    "                'Receptor Cell Index': r_idx,\n",
    "                'Ligand Cell Type': adata.obs[celltype_col].iloc[ligand_cell_idx],\n",
    "                'Receptor Cell Type': adata.obs[celltype_col].iloc[r_idx],\n",
    "                'Distance': dist,\n",
    "                'Ligand Expression': L_exp_val,\n",
    "                'Receptor Expression': R_exp_val,\n",
    "                'LR Score': score\n",
    "            })\n",
    "\n",
    "    pairwise_scores_df = pd.DataFrame(score_results)\n",
    "    \n",
    "    # aggregation per celltype pair\n",
    "    agg = pairwise_scores_df.groupby(['Ligand Cell Type','Receptor Cell Type']).agg(\n",
    "        score_sum=('LR Score','sum'),\n",
    "        score_mean=('LR Score','mean'),\n",
    "        pair_count=('LR Score','count')\n",
    "    ).reset_index()\n",
    "\n",
    "    # normalize by possible pairs (N_i * N_j)\n",
    "    counts = adata.obs[celltype_col].value_counts().to_dict()\n",
    "    agg['N_ligandCt'] = agg['Ligand Cell Type'].map(counts)\n",
    "    agg['N_ReceptorCt'] = agg['Receptor Cell Type'].map(counts)\n",
    "    agg['norm_score'] = agg['score_sum'] / (agg['N_ligandCt'] * agg['N_ReceptorCt'])\n",
    "\n",
    "    return pairwise_scores_df, agg\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_LR_scores(adata, ligand_gene, receptor_gene,\n",
    "                      distance_threshold=200, dist_lambda = 50,\n",
    "                      expr_layer = None, celltype_col='annot', sample_col='Sample',\n",
    "                      x_col='x_centroid', y_col='y_centroid'):\n",
    "    \"\"\"\n",
    "    DOCS\n",
    "    \"\"\"\n",
    "    all_scores = []\n",
    "    all_agg = []\n",
    "    for sample in adata.obs[sample_col].unique():\n",
    "        print(f\"Processing sample: {sample}\")\n",
    "        sample_adata = adata[adata.obs[sample_col] == sample]\n",
    "        sample_scores, agg = compute_pairwise_LR_scores(sample_adata, ligand_gene, receptor_gene,\n",
    "                                                   distance_threshold, dist_lambda,\n",
    "                                                   expr_layer, celltype_col, x_col, y_col)\n",
    "        sample_scores['Sample'] = sample\n",
    "        agg['Sample'] = sample\n",
    "        all_agg.append(agg)\n",
    "        all_scores.append(sample_scores)\n",
    "\n",
    "    all_scores_df = pd.concat(all_scores, axis=0)\n",
    "    all_agg_df = pd.concat(all_agg, axis=0)\n",
    "    \n",
    "    return all_scores_df, all_agg_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compute all ligand-receptor cell pairs and their scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out cells with unassigned radial neighborhoods\n",
    "ird_xenium_merge_filtered = ird_xenium_merge[ird_xenium_merge.obs['radial_neighborhood'] != 'Unassigned', :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ligand_gene = 'TNFSF13'  # APRIL\n",
    "receptor_gene = 'TNFRSF17'  # BCMA\n",
    "april_bcma_scores_allsamples, agg_allsamples = compute_LR_scores(ird_xenium_merge_filtered, ligand_gene, receptor_gene,\n",
    "                                                                 distance_threshold=200, dist_lambda = 100,\n",
    "                                                                 expr_layer = None, celltype_col='annot', sample_col = 'Sample', x_col='x_centroid', y_col='y_centroid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "april_taci_scores_allsamples, agg_at_allsamples = compute_LR_scores(ird_xenium_merge, 'TNFSF13', 'TNFRSF13B',\n",
    "                                                                 distance_threshold=200, dist_lambda = 100,\n",
    "                                                                 expr_layer = None, celltype_col='annot', sample_col = 'Sample', x_col='x_centroid', y_col='y_centroid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "april_bcma_scores_allsamples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### LR score per receptor cell - average or sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In april_bcma_scores_allsamples, for each unique Receptor Cell Index, average (or sum) the LR Score across all Ligand Cell Indexs\n",
    "receptor_avg_scores = april_bcma_scores_allsamples.groupby(['Receptor Cell Index', 'Sample']).agg(\n",
    "    avg_LR_score=('LR Score', 'mean'),\n",
    "    sum_LR_score=('LR Score', 'sum'),\n",
    "    Receptor_Cell_Type=('Receptor Cell Type', 'first')\n",
    ").reset_index()\n",
    "metadata = ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "receptor_avg_scores['Collection'] = receptor_avg_scores['Sample'].map(metadata['Collection'])\n",
    "receptor_avg_scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "receptor_avg_scores_per_sample = receptor_avg_scores.groupby(['Sample', 'Receptor_Cell_Type']).agg(\n",
    "    mean_sum_LR_score=('sum_LR_score', 'mean'),\n",
    "    mean_avg_LR_score=('avg_LR_score', 'mean'),\n",
    "    Collection = ('Collection', 'first')\n",
    ")\n",
    "plt.figure(figsize=(8,4))\n",
    "sns.boxplot(receptor_avg_scores_per_sample, x = 'Receptor_Cell_Type', y = 'mean_sum_LR_score', hue = 'Collection', showfliers = False)\n",
    "plt.title('Average of Sum LR Score per BCMA+ cell')\n",
    "plt.xlabel('Receptor Cell Type')\n",
    "plt.ylabel('Average Sum LR Score')\n",
    "plt.xticks(rotation=45)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8,4))\n",
    "sns.boxplot(receptor_avg_scores_per_sample, x = 'Receptor_Cell_Type', y = 'mean_avg_LR_score', hue = 'Collection', showfliers = False)\n",
    "plt.title('Average of Mean LR Score per BCMA+ cell')\n",
    "plt.xlabel('Receptor Cell Type')\n",
    "plt.ylabel('Average Mean LR Score')\n",
    "plt.xticks(rotation=45)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Number of (& density of) ligand+ cells around PCs in a certain distance bin (#cells/area of bin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Density of ligand+ cells around each receptor cell type within distance bins per sample\n",
    "binwidth = 20\n",
    "bin_limit = 200\n",
    "bins = np.arange(0, bin_limit+1, binwidth)\n",
    "april_bcma_scores_allsamples['dist_bin'] = pd.cut(april_bcma_scores_allsamples['Distance'], bins=bins, labels = bins[0:-1])\n",
    "april_bcma_scores_allsamples['Receptor Cell Index'] = april_bcma_scores_allsamples['Receptor Cell Index'].astype(int)\n",
    "density_results = april_bcma_scores_allsamples.groupby(['Sample', 'Receptor Cell Index', 'dist_bin']).size().reset_index(name='count')\n",
    "density_results = density_results[density_results['count'] != 0]\n",
    "# Normalize by area of annulus\n",
    "density_results['area'] = np.pi * ((density_results['dist_bin'].astype(float) + binwidth)**2 - (density_results['dist_bin'].astype(float))**2)\n",
    "density_results['density'] = density_results['count'] / density_results['area']\n",
    "# Add cell type information\n",
    "density_results['Sample_Index'] = density_results['Sample'] + '_' + density_results['Receptor Cell Index'].astype(str)\n",
    "celltype_map = april_bcma_scores_allsamples[['Receptor Cell Index', 'Sample', 'Receptor Cell Type']].drop_duplicates()\n",
    "celltype_map['Sample_Index'] = celltype_map['Sample'] + '_' + celltype_map['Receptor Cell Index'].astype(int).astype(str)\n",
    "density_results = density_results.merge(celltype_map[['Sample_Index', 'Receptor Cell Type']], on='Sample_Index', how='left')\n",
    "density_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Average count of ligand+ cells per distance bin to PC cells per sample\n",
    "avg_count = density_results[density_results['Receptor Cell Type']=='PC'].groupby(['Sample', 'dist_bin']).agg(\n",
    "    mean_count = ('count', 'mean')\n",
    ").reset_index()\n",
    "\n",
    "metadata = ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "avg_count['Collection'] = avg_count['Sample'].map(metadata['Collection'])\n",
    "sns.boxplot(avg_count, x = 'dist_bin', y = 'mean_count', hue = 'Collection')\n",
    "plt.title('Average count of APRIL+ ligand cells around BCMA+ PC cells per distance bin')\n",
    "plt.xlabel('Distance bin (um)')\n",
    "plt.ylabel('Average count of ligand+ cells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_cells_info = ird_xenium_merge.obs.copy()\n",
    "metadata = ird_cells_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_cells_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a column in ird_cells_info called APRILstat_ct that indicates whether the cells are APRIL+, APRIL_receptor+, both, or neither, and concatenate the cell type\n",
    "apr = ird_cells_info['APRILpos']\n",
    "rcp = ird_cells_info['APRIL_receptor_pos']\n",
    "\n",
    "labels = np.select(\n",
    "    condlist=[apr & rcp, apr & ~rcp, ~apr & rcp],\n",
    "    choicelist=['both', 'APRIL+', 'APRILrec+'],\n",
    "    default='neither'\n",
    ")\n",
    "\n",
    "ird_cells_info['APRILstat_ct'] = pd.Series(labels, index=ird_cells_info.index) + '_' + ird_cells_info['annot'].astype(str)\n",
    "\n",
    "# optional: check\n",
    "ird_cells_info['APRILstat_ct'].value_counts().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sn219_april_annot = ird_cells_info[ird_cells_info['Sample'] == 'SN219R1-Ma1Fd2-1U1'][['Original_Barcode', 'APRILstat_ct']]\n",
    "sn219_april_annot.columns = ['cell_id', 'group']\n",
    "sn219_april_annot.to_csv('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_SN219_APRIL_annotation_v10082025.csv', index=False)\n",
    "s13_april_annot = ird_cells_info[ird_cells_info['Sample'] == 'S13-47674A1U1'][['Original_Barcode', 'APRILstat_ct']]\n",
    "s13_april_annot.columns = ['cell_id', 'group']\n",
    "s13_april_annot.to_csv('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_S13-47674_APRIL_annotation_v10082025.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s18_april_annot = ird_cells_info[ird_cells_info['Sample'] == 'S18-36373A1U1'][['Original_Barcode', 'APRILstat_ct']]\n",
    "s18_april_annot.columns = ['cell_id', 'group']\n",
    "s18_april_annot.to_csv('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_S18-36373A1U1_APRIL_annotation_v10082025.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#avg_count[avg_count['Sample']=='S13-47674A1U1']\n",
    "avg_count[avg_count['Sample']=='S18-36373A1U1']\n",
    "#avg_count[avg_count['Sample']=='SN219R1-Ma1Fd2-1U1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Average count of ligand+ cells per distance bin to B cells per sample\n",
    "avg_count = density_results[density_results['Receptor Cell Type']=='B'].groupby(['Sample', 'dist_bin']).agg(\n",
    "    mean_count = ('count', 'mean')\n",
    ").reset_index()\n",
    "\n",
    "avg_count['Collection'] = avg_count['Sample'].map(metadata['Collection'])\n",
    "sns.boxplot(avg_count, x = 'dist_bin', y = 'mean_count', hue = 'Collection')\n",
    "plt.title('Average count of APRIL+ ligand cells around BCMA+ B cells per distance bin')\n",
    "plt.xlabel('Distance bin (um)')\n",
    "plt.ylabel('Average count of ligand+ cells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Check % of APRIL+ cells per sample\n",
    "april_df = ird_xenium_merge[:, 'TNFSF13'].to_df()\n",
    "baff_df = ird_xenium_merge[:, 'TNFSF13B'].to_df()\n",
    "ird_xenium_merge.obs.loc[:, 'APRILpos'] = april_df['TNFSF13'] > 0\n",
    "ird_xenium_merge.obs.loc[:, 'BAFFpos'] = baff_df['TNFSF13B'] > 0\n",
    "\n",
    "april_ligand_df = ird_xenium_merge[:, ['TNFRSF17', 'TNFRSF13B']].to_df()\n",
    "baff_ligand_df = ird_xenium_merge[:, ['TNFRSF13B', 'TNFRSF13C']].to_df()\n",
    "ird_xenium_merge.obs.loc[:, 'APRIL_receptor_pos'] = (april_ligand_df['TNFRSF17'] > 0) | (april_ligand_df['TNFRSF13B'] > 0)\n",
    "ird_xenium_merge.obs.loc[:, 'BAFF_receptor_pos'] = (baff_ligand_df['TNFRSF13B'] > 0) | (baff_ligand_df['TNFRSF13C'] > 0)\n",
    "\n",
    "april_pos_frac = ird_xenium_merge.obs.groupby('Sample').agg(\n",
    "    total_cells = ('APRILpos', 'size'),\n",
    "    april_pos_cells = ('APRILpos', 'sum'),\n",
    "    april_receptor_pos_cells = ('APRIL_receptor_pos', 'sum')\n",
    ").reset_index()\n",
    "april_pos_frac = april_pos_frac[(april_pos_frac['april_pos_cells'] > 0) & (april_pos_frac['april_receptor_pos_cells'] > 0)]\n",
    "april_pos_frac['frac_april_pos'] = april_pos_frac['april_pos_cells'] / april_pos_frac['total_cells']\n",
    "april_pos_frac['frac_april_receptor_pos'] = april_pos_frac['april_receptor_pos_cells'] / april_pos_frac['total_cells']\n",
    "april_pos_frac['Collection'] = april_pos_frac['Sample'].map(metadata['Collection'])\n",
    "april_pos_frac.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(april_pos_frac, x = 'Collection', y = 'frac_april_pos')\n",
    "sns.swarmplot(april_pos_frac, x = 'Collection', y = 'frac_april_pos', color = 'k', alpha = .5)\n",
    "plt.title('Fraction of APRIL+ cells per sample')\n",
    "plt.xlabel('Timepoint')\n",
    "plt.ylabel('Fraction of APRIL+ cells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Average density per sample and distance bin for PCs\n",
    "avg_density = density_results[density_results['Receptor Cell Type'] == 'PC'].groupby(['Sample', 'dist_bin']).agg(\n",
    "    mean_density = ('density', 'mean')\n",
    ").reset_index()\n",
    "# Pivot results for plotting - sample as columns, dist_bin as index\n",
    "density_mat = avg_density.pivot(index='dist_bin', columns='Sample', values='mean_density')\n",
    "# Reorder sample columns by their timepoint\n",
    "metadata = ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates()#.set_index('Sample')\n",
    "sample_order = metadata.sort_values(['Collection', 'Sample'])['Sample'].tolist()  # Collection is ordered categorical already\n",
    "sample_order = [s for s in sample_order if s in density_mat.columns]  # Keep only samples present in the matrix\n",
    "density_mat = density_mat.reindex(columns=sample_order)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### For APRIL+ cells and BCMA+ PCs, the average expression of APRIL and BCMA with distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_summary = april_bcma_scores_allsamples[april_bcma_scores_allsamples['Receptor Cell Type'] == 'PC'].groupby(['Sample', 'dist_bin']).agg(\n",
    "    mean_L = ('Ligand Expression', 'mean'),\n",
    "    mean_R = ('Receptor Expression', 'mean')\n",
    ").reset_index()\n",
    "lig_mat = expr_summary.pivot(index = 'dist_bin', columns = 'Sample', values = 'mean_L').fillna(0)\n",
    "rec_mat = expr_summary.pivot(index = 'dist_bin', columns = 'Sample', values = 'mean_R').fillna(0)\n",
    "# Reorder sample columns by their timepoint\n",
    "metadata = ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates()#.set_index('Sample')\n",
    "sample_order = metadata.sort_values(['Collection', 'Sample'])['Sample'].tolist()  # Collection is ordered categorical already\n",
    "sample_order = [s for s in sample_order if s in lig_mat.columns]  # Keep only samples present in the matrix\n",
    "lig_mat = lig_mat.reindex(columns=sample_order)\n",
    "sample_order = [s for s in sample_order if s in rec_mat.columns]  # Keep only samples present in the matrix\n",
    "rec_mat = rec_mat.reindex(columns=sample_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "expr_summary['Collection'] = expr_summary['Sample'].map(metadata['Collection'])\n",
    "sns.boxplot(expr_summary, x = 'Collection', y = 'mean_L', hue = 'dist_bin')\n",
    "plt.legend(title = 'Distance Bins', bbox_to_anchor=(1, 1), loc='upper left')\n",
    "plt.title('Mean APRIL expression in APRIL+ cells around BCMA+ PC per distance bin')\n",
    "plt.xlabel('Timepoint')\n",
    "plt.ylabel('Mean APRIL expression')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(expr_summary, x = 'Collection', y = 'mean_R', hue = 'dist_bin')\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "plt.legend(title = 'Distance Bins', bbox_to_anchor=(1, 1), loc='upper left')\n",
    "plt.title('Mean BCMA expression in BCMA+ PCs around APRIL+ cells per distance bin')\n",
    "plt.xlabel('Timepoint')\n",
    "plt.ylabel('Mean BCMA expression')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### For APRIL+ cells and BCMA+ B cells, the average expression of APRIL and BCMA with distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_summary_B = april_bcma_scores_allsamples[april_bcma_scores_allsamples['Receptor Cell Type'] == 'B'].groupby(['Sample', 'dist_bin']).agg(\n",
    "    mean_L = ('Ligand Expression', 'mean'),\n",
    "    mean_R = ('Receptor Expression', 'mean')\n",
    ").reset_index()\n",
    "lig_mat_B = expr_summary_B.pivot(index = 'dist_bin', columns = 'Sample', values = 'mean_L').fillna(0)\n",
    "rec_mat_B = expr_summary_B.pivot(index = 'dist_bin', columns = 'Sample', values = 'mean_R').fillna(0)\n",
    "# Reorder sample columns by their timepoint\n",
    "metadata = ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates()#.set_index('Sample')\n",
    "sample_order = metadata.sort_values(['Collection', 'Sample'])['Sample'].tolist()  # Collection is ordered categorical already\n",
    "sample_order = [s for s in sample_order if s in lig_mat_B.columns]  # Keep only samples present in the matrix\n",
    "lig_mat_B = lig_mat_B.reindex(columns=sample_order)\n",
    "sample_order = [s for s in sample_order if s in rec_mat_B.columns]  # Keep only samples present in the matrix\n",
    "rec_mat_B = rec_mat_B.reindex(columns=sample_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "expr_summary_B['Collection'] = expr_summary_B['Sample'].map(metadata['Collection'])\n",
    "sns.boxplot(expr_summary_B, x = 'Collection', y = 'mean_L', hue = 'dist_bin')\n",
    "plt.legend(title = 'Distance Bins', bbox_to_anchor=(1, 1), loc='upper left')\n",
    "plt.title('Mean APRIL expression in APRIL+ cells around BCMA+ B cells per distance bin')\n",
    "plt.xlabel('Timepoint')\n",
    "plt.ylabel('Mean APRIL expression')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(expr_summary_B, x = 'Collection', y = 'mean_R', hue = 'dist_bin')\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc='upper left')\n",
    "plt.legend(title = 'Distance Bins', bbox_to_anchor=(1, 1), loc='upper left')\n",
    "plt.title('Mean BCMA expression in BCMA+ B cells around APRIL+ cells per distance bin')\n",
    "plt.xlabel('Timepoint')\n",
    "plt.ylabel('Mean BCMA expression')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create concentric radial heatmap plots\n",
    "theta_vals = np.linspace(0, 2*np.pi, len(lig_mat.columns), endpoint=False)\n",
    "\n",
    "fig, ax = plt.subplots(subplot_kw=dict(projection='polar'), figsize=(7,7))\n",
    "\n",
    "# Parameters\n",
    "n_dist = lig_mat.shape[0]\n",
    "n_samples = lig_mat.shape[1]\n",
    "theta_edges = np.linspace(0, 2*np.pi, n_samples + 1)\n",
    "\n",
    "# Define separate radial scales for inner (ligand) and outer (receptor)\n",
    "r_max = 100#bin_limit  # Maximum distance for receptor ring\n",
    "\n",
    "r_edges_lig = np.linspace(0, r_max, n_dist + 1)\n",
    "mesh_L = ax.pcolormesh(theta_edges, r_edges_lig, lig_mat.values,\n",
    "                       cmap='Blues', shading='auto')#, vmin=0, vmax=1)\n",
    "\n",
    "# Color code the collections\n",
    "collections = metadata.loc[lig_mat.columns, 'Collection']\n",
    "cond_groups = collections.groupby(collections).indices\n",
    "\n",
    "for cond, idx in cond_groups.items():\n",
    "    start = theta_edges[min(idx)]\n",
    "    end = theta_edges[max(idx)+1]\n",
    "    ax.plot([start, start], [0, r_max], color='red', lw=3)\n",
    "    ax.plot([end, end], [0, r_max], color='red', lw=3)\n",
    "plt.title('APRIL expression of APRIL+ cells around BCMA+ Plasma cells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(subplot_kw=dict(projection='polar'), figsize=(7,7))\n",
    "r_edges_lig = np.linspace(0, r_max, n_dist + 1)\n",
    "mesh_L = ax.pcolormesh(theta_edges, r_edges_lig, rec_mat.values,\n",
    "                       cmap='Blues', shading='auto')#, vmin=0, vmax=1)\n",
    "\n",
    "for cond, idx in cond_groups.items():\n",
    "    start = theta_edges[min(idx)]\n",
    "    end = theta_edges[max(idx)+1]\n",
    "    ax.plot([start, start], [0, r_max], color='red', lw=3)\n",
    "    ax.plot([end, end], [0, r_max], color='red', lw=3)\n",
    "plt.title('BCMA expression of BCMA+ Plasma cells around APRIL+ cells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(subplot_kw=dict(projection='polar'), figsize=(7,7))\n",
    "# Define separate radial scales for inner (ligand) and outer (receptor)\n",
    "r_inner_max = 100   # inner ring radius\n",
    "r_outer_min = 110   # where outer ring starts\n",
    "r_outer_max = 200   # outer ring limit\n",
    "\n",
    "# Ligand ring: map distance bins [0, max]  [0, r_inner_max]\n",
    "r_edges_lig = np.linspace(0, r_inner_max, n_dist + 1)\n",
    "mesh_L = ax.pcolormesh(theta_edges, r_edges_lig, lig_mat.values,\n",
    "                       cmap='Blues', shading='auto')#, vmin=0, vmax=1)\n",
    "# Receptor ring: map distance bins [0, max]  [r_outer_min, r_outer_max]\n",
    "r_edges_rec = np.linspace(r_outer_min, r_outer_max, n_dist + 1)\n",
    "mesh_R = ax.pcolormesh(theta_edges, r_edges_rec, rec_mat.values,\n",
    "                       cmap='Oranges', shading='auto')#, vmin=0, vmax=1\n",
    "\n",
    "for cond, idx in cond_groups.items():\n",
    "    start = theta_edges[min(idx)]\n",
    "    end = theta_edges[max(idx)+1]\n",
    "    ax.plot([start, start], [0, r_outer_max], color='red', lw=3)\n",
    "    ax.plot([end, end], [0, r_outer_max], color='red', lw=3)\n",
    "plt.title('APRIL (inner ring) and BCMA (outer ring) expression around Plasma cells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "theta_vals = np.linspace(0, 2*np.pi, len(lig_mat_B.columns), endpoint=False)\n",
    "\n",
    "fig, ax = plt.subplots(subplot_kw=dict(projection='polar'), figsize=(7,7))\n",
    "\n",
    "# Parameters\n",
    "n_dist = lig_mat_B.shape[0]\n",
    "n_samples = lig_mat_B.shape[1]\n",
    "theta_edges = np.linspace(0, 2*np.pi, n_samples + 1)\n",
    "\n",
    "# Define separate radial scales for inner (ligand) and outer (receptor)\n",
    "r_inner_max = 100   # inner ring radius\n",
    "r_outer_min = 110   # where outer ring starts\n",
    "r_outer_max = 200   # outer ring limit\n",
    "\n",
    "# Ligand ring: map distance bins [0, max]  [0, r_inner_max]\n",
    "r_edges_lig = np.linspace(0, r_inner_max, n_dist + 1)\n",
    "mesh_L = ax.pcolormesh(theta_edges, r_edges_lig, lig_mat_B.values,\n",
    "                       cmap='Blues', shading='auto')#, vmin=0, vmax=1)\n",
    "# Receptor ring: map distance bins [0, max]  [r_outer_min, r_outer_max]\n",
    "r_edges_rec = np.linspace(r_outer_min, r_outer_max, n_dist + 1)\n",
    "mesh_R = ax.pcolormesh(theta_edges, r_edges_rec, rec_mat_B.values,\n",
    "                       cmap='Oranges', shading='auto')#, vmin=0, vmax=1\n",
    "\n",
    "# Color code the collections\n",
    "collections = metadata.loc[lig_mat_B.columns, 'Collection']\n",
    "cond_groups = collections.groupby(collections).indices\n",
    "for cond, idx in cond_groups.items():\n",
    "    start = theta_edges[min(idx)]\n",
    "    end = theta_edges[max(idx)+1]\n",
    "    ax.plot([start, start], [0, r_outer_max], color='red', lw=3)\n",
    "    ax.plot([end, end], [0, r_outer_max], color='red', lw=3)\n",
    "plt.title('APRIL (inner ring) and BCMA (outer ring) expression around B cells')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Percent ct pairs are interacting\n",
    "\n",
    "#Are there more APRIL+ cells around PC than B in NDMM?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Summarize average expression and score per sample, per distance bin\n",
    "bins = np.arange(0, 200, 20)\n",
    "april_bcma_scores_allsamples['dist_bin'] = pd.cut(april_bcma_scores_allsamples['Distance'], bins=bins, labels = bins[0:-1])\n",
    "expr_summary = april_bcma_scores_allsamples.groupby(['Sample', 'dist_bin']).agg(\n",
    "    mean_L = ('Ligand Expression', 'mean'),\n",
    "    mean_R = ('Receptor Expression', 'mean'),\n",
    "    mean_score = ('LR Score', 'mean'),\n",
    "    sum_score = ('LR Score', 'sum'),\n",
    "    count_pairs = ('LR Score', 'size')\n",
    ").reset_index()\n",
    "expr_summary.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lig_mat = expr_summary.pivot(index = 'dist_bin', columns = 'Sample', values = 'mean_L').fillna(0)\n",
    "rec_mat = expr_summary.pivot(index = 'dist_bin', columns = 'Sample', values = 'mean_R').fillna(0)\n",
    "score_mat = expr_summary.pivot(index = 'dist_bin', columns = 'Sample', values = 'sum_score').fillna(0)\n",
    "count_mat = expr_summary.pivot(index = 'dist_bin', columns = 'Sample', values = 'count_pairs').fillna(0)\n",
    "lig_mat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reorder sample columns by their timepoint\n",
    "metadata = ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates()#.set_index('Sample')\n",
    "sample_order = metadata.sort_values(['Collection', 'Sample'])['Sample'].tolist()  # Collection is ordered categorical already\n",
    "sample_order = [s for s in sample_order if s in lig_mat.columns]  # Keep only samples present in the matrix\n",
    "lig_mat = lig_mat.reindex(columns=sample_order)\n",
    "rec_mat = rec_mat.reindex(columns=sample_order)\n",
    "score_mat = score_mat.reindex(columns=sample_order)\n",
    "count_mat = count_mat.reindex(columns=sample_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create concentric radial heatmap plots\n",
    "#r_vals = lig_mat.index.astype(float)\n",
    "theta_vals = np.linspace(0, 2*np.pi, len(lig_mat.columns), endpoint=False)\n",
    "\n",
    "fig, ax = plt.subplots(subplot_kw=dict(projection='polar'), figsize=(7,7))\n",
    "\n",
    "# Parameters\n",
    "n_dist = lig_mat.shape[0]\n",
    "n_samples = lig_mat.shape[1]\n",
    "theta_edges = np.linspace(0, 2*np.pi, n_samples + 1)\n",
    "\n",
    "# Define separate radial scales for inner (ligand) and outer (receptor)\n",
    "r_inner_max = 100   # inner ring radius\n",
    "r_outer_min = 110   # where outer ring starts\n",
    "r_outer_max = 200   # outer ring limit\n",
    "\n",
    "# Ligand ring: map distance bins [0, max]  [0, r_inner_max]\n",
    "r_edges_lig = np.linspace(0, r_inner_max, n_dist + 1)\n",
    "mesh_L = ax.pcolormesh(theta_edges, r_edges_lig, lig_mat.values,\n",
    "                       cmap='Blues', shading='auto')#, vmin=0, vmax=1)\n",
    "\n",
    "# Color code the collections\n",
    "collections = metadata.set_index('Sample').loc[lig_mat.columns, 'Collection']\n",
    "cond_groups = collections.groupby(collections).indices\n",
    "\n",
    "for cond, idx in cond_groups.items():\n",
    "    start = theta_edges[min(idx)]\n",
    "    end = theta_edges[max(idx)+1]\n",
    "    ax.plot([start, start], [0, r_inner_max], color='red', lw=3)\n",
    "    ax.plot([end, end], [0, r_inner_max], color='red', lw=3)\n",
    "plt.title('APRIL')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(subplot_kw=dict(projection='polar'), figsize=(7,7))\n",
    "mesh_L = ax.pcolormesh(theta_edges, r_edges_lig, rec_mat.values,\n",
    "                       cmap='Blues', shading='auto')#, vmin=0, vmax=1)\n",
    "\n",
    "# Color code the collections\n",
    "collections = metadata.set_index('Sample').loc[lig_mat.columns, 'Collection']\n",
    "cond_groups = collections.groupby(collections).indices\n",
    "\n",
    "for cond, idx in cond_groups.items():\n",
    "    start = theta_edges[min(idx)]\n",
    "    end = theta_edges[max(idx)+1]\n",
    "    ax.plot([start, start], [0, r_inner_max], color='red', lw=3)\n",
    "    ax.plot([end, end], [0, r_inner_max], color='red', lw=3)\n",
    "plt.title('BCMA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(subplot_kw=dict(projection='polar'), figsize=(7,7))\n",
    "mesh_L = ax.pcolormesh(theta_edges, r_edges_lig, score_mat.values,\n",
    "                       cmap='Blues', shading='auto')#, vmin=0, vmax=1)\n",
    "\n",
    "# Color code the collections\n",
    "collections = metadata.set_index('Sample').loc[lig_mat.columns, 'Collection']\n",
    "cond_groups = collections.groupby(collections).indices\n",
    "\n",
    "for cond, idx in cond_groups.items():\n",
    "    start = theta_edges[min(idx)]\n",
    "    end = theta_edges[max(idx)+1]\n",
    "    ax.plot([start, start], [0, r_inner_max], color='red', lw=3)\n",
    "    ax.plot([end, end], [0, r_inner_max], color='red', lw=3)\n",
    "plt.title('LR score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(subplot_kw=dict(projection='polar'), figsize=(7,7))\n",
    "mesh_L = ax.pcolormesh(theta_edges, r_edges_lig, count_mat.values,\n",
    "                       cmap='Blues', shading='auto')#, vmin=0, vmax=1)\n",
    "\n",
    "# Color code the collections\n",
    "collections = metadata.set_index('Sample').loc[lig_mat.columns, 'Collection']\n",
    "cond_groups = collections.groupby(collections).indices\n",
    "\n",
    "for cond, idx in cond_groups.items():\n",
    "    start = theta_edges[min(idx)]\n",
    "    end = theta_edges[max(idx)+1]\n",
    "    ax.plot([start, start], [0, r_inner_max], color='red', lw=3)\n",
    "    ax.plot([end, end], [0, r_inner_max], color='red', lw=3)\n",
    "plt.title('LR pair count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create concentric radial heatmap plots\n",
    "#r_vals = lig_mat.index.astype(float)\n",
    "theta_vals = np.linspace(0, 2*np.pi, len(lig_mat.columns), endpoint=False)\n",
    "\n",
    "fig, ax = plt.subplots(subplot_kw=dict(projection='polar'), figsize=(7,7))\n",
    "\n",
    "# Parameters\n",
    "n_dist = lig_mat.shape[0]\n",
    "n_samples = lig_mat.shape[1]\n",
    "theta_edges = np.linspace(0, 2*np.pi, n_samples + 1)\n",
    "\n",
    "# Define separate radial scales for inner (ligand) and outer (receptor)\n",
    "r_inner_max = 100   # inner ring radius\n",
    "r_outer_min = 110   # where outer ring starts\n",
    "r_outer_max = 200   # outer ring limit\n",
    "\n",
    "# Ligand ring: map distance bins [0, max]  [0, r_inner_max]\n",
    "r_edges_lig = np.linspace(0, r_inner_max, n_dist + 1)\n",
    "mesh_L = ax.pcolormesh(theta_edges, r_edges_lig, lig_mat.values,\n",
    "                       cmap='Blues', shading='auto')#, vmin=0, vmax=1)\n",
    "\n",
    "r_edges_rec = np.linspace(r_outer_min, r_outer_max, n_dist + 1)\n",
    "mesh_R = ax.pcolormesh(theta_edges, r_edges_rec, rec_mat.values,\n",
    "                       cmap='Oranges', shading='auto')#, vmin=0, vmax=1\n",
    "\n",
    "# Color code the collections\n",
    "collections = metadata.set_index('Sample').loc[lig_mat.columns, 'Collection']\n",
    "cond_groups = collections.groupby(collections).indices\n",
    "\n",
    "for cond, idx in cond_groups.items():\n",
    "    start = theta_edges[min(idx)]\n",
    "    end = theta_edges[max(idx)+1]\n",
    "    ax.plot([start, start], [0, r_outer_max], color='red', lw=3)\n",
    "    ax.plot([end, end], [0, r_outer_max], color='red', lw=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 3: Myeloid Immunosuppression Analysis\n",
    "\n",
    "Myeloid cells have been shown to be a source of a immunosuppression tumor microenvironment, such as myeloid-derived suppressive cells (MDSCs) and tumor-associated macrophages (TAMs). Due to the limited gene panel and the sparsity of transcript detection, we may not be able to identify these populations using unsupervised clustering of their Xenium transcript profiles. Alternatively, this section analyzes the expression of immune checkpoint molecules and immunosuppressive factors in myeloid populations (especially macrophages) across disease stages to identify changes in immunosuppressive capacity. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genes associated with myeloid immunosuppression\n",
    "\n",
    "Myeloid cells are assessed for immunosuppressive phenotype based on expression of the following markers:\n",
    "\n",
    "**Immune checkpoint and immunosuppressive molecules:**\n",
    "CD274 (PD-L1)(on all panel), SPP1, MARCO, IL1R2, IL1RL1, TGFB1 (on v5, v6 panel)\n",
    "\n",
    "**Transcriptional regulators:**\n",
    "STAT3, STAT1, NFKB1, NFKBIA\n",
    "\n",
    "Cells expressing 1 of these markers are considered to have an immunosuppressive MDSC-like phenotype. We also test for spatial proximity to exhausted T cells, which are hypothesized to be induced by immunosuppressive myeloid cells. These T cells are identified by the following markers:\n",
    "\n",
    "**T cell exhaustion markers**:\n",
    "CTLA4, LAG3, HAVCR2 (TIM-3), TIGIT\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_xenium_merge.obs['ct_timepoint'] = ird_xenium_merge.obs.apply(lambda x: f\"{x['ct']}_{x['Collection']}\", axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PDL1 (CD274) is expressed at higher levels in macrophages and megakaryocytes. PDL1+ macrophages could represent a possible mMDSC phenotype. Megakaryocytes maybe somewhat novel players expressing PDL1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_xenium_merge, var_names = 'CD274', groupby = 'ct_timepoint', swap_axes = True, dot_max = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_myeloid = ird_xenium_merge[ird_xenium_merge.obs['ct'].isin(['GMP', 'Late Myeloid', 'Neutrophil', 'Ba/Eo/Ma/', 'Monocyte', 'Macrophage', 'pDC', 'cDC']), :].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "timecols = {\"NBM\": \"#0C7515\", \"NDMM\": \"#E619B9\", \"PT\": \"#CF99C3\"} "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SPP1 expression is higher in macrophages in NDMM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_macro = ird_myeloid[ird_myeloid.obs['ct'].isin(['Macrophage']), :].copy()\n",
    "sc.pl.dotplot(ird_macro, var_names = 'SPP1', groupby = 'Collection', swap_axes = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_macro.obs['SPP1_exp'] = ird_macro[:, 'SPP1'].X.toarray()\n",
    "ird_macro.obs['SPP1_pos'] = ird_macro[:, 'SPP1'].X.toarray() > 0\n",
    "macro_spp1_exp = ird_macro.obs.groupby('Sample')['SPP1_exp'].mean().reset_index()\n",
    "macro_spp1_exp = macro_spp1_exp.merge(ird_macro.obs.groupby('Sample')['SPP1_pos'].mean().reset_index(), on='Sample', how='left')\n",
    "macro_spp1_exp['Collection'] = macro_spp1_exp['Sample'].map(ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')['Collection'])\n",
    "plot_utils.plot_comparison_with_significance(macro_spp1_exp, 'Collection', 'SPP1_pos', order=['NBM', 'NDMM', 'PT'], \n",
    "                                             palette=timecols, xlabel = 'Timepoint', ylabel = 'Fraction of Macrophages with SPP1 Expression')\n",
    "                                             #save_path = '/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_macrophage_SPP1fraction.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PDL1 (CD274) expression in macrophages is higher in NDMM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_macro, var_names = 'CD274', groupby = 'Collection', swap_axes = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_macro.obs['PDL1_exp'] = ird_macro[:, 'CD274'].X.toarray()\n",
    "ird_macro.obs['PDL1_pos'] = ird_macro[:, 'CD274'].X.toarray() > 0\n",
    "macro_pdl1_exp = ird_macro.obs.groupby('Sample')['PDL1_exp'].mean().reset_index()\n",
    "macro_pdl1_exp = macro_pdl1_exp.merge(ird_macro.obs.groupby('Sample')['PDL1_pos'].mean().reset_index(), on='Sample', how='left')\n",
    "macro_pdl1_exp['Collection'] = macro_pdl1_exp['Sample'].map(ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')['Collection'])\n",
    "plot_utils.plot_comparison_with_significance(macro_pdl1_exp, 'Collection', 'PDL1_pos', order=['NBM', 'NDMM', 'PT'], \n",
    "                                             palette=timecols, xlabel = 'Timepoint', ylabel = 'Fraction of Macrophages with PDL1 Expression')\n",
    "                                             #save_path = '/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_macrophage_PDL1fraction.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_macro, var_names = 'MARCO', groupby = 'Collection', swap_axes = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_macro.obs['MARCO_exp'] = ird_macro[:, 'MARCO'].X.toarray()\n",
    "ird_macro.obs['MARCO_pos'] = ird_macro[:, 'MARCO'].X.toarray() > 0\n",
    "macro_marco_exp = ird_macro.obs.groupby('Sample')['MARCO_exp'].mean().reset_index()\n",
    "macro_marco_exp = macro_marco_exp.merge(ird_macro.obs.groupby('Sample')['MARCO_pos'].mean().reset_index(), on='Sample', how='left')\n",
    "macro_marco_exp['Collection'] = macro_marco_exp['Sample'].map(ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')['Collection'])\n",
    "plot_utils.plot_comparison_with_significance(macro_marco_exp, 'Collection', 'MARCO_exp', order=['NBM', 'NDMM', 'PT'], palette=timecols)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Immunosuppression gene set scores\n",
    "\n",
    "Both SPP1+ macrophages and PDL1+ macrophages have been reported to exhibit immunosuppressive activities.\n",
    "\n",
    "Due to their sparse expression and the lack of relevant MDSC/SPP1+ macrophage markers, we use a geneset score approach to identify macrophages that are likely to exhibit immunosuppressive capacities.\n",
    "\n",
    "We include both SPP1 and CD274 (PDL1) as part of this gene set. While MARCO expression is lower in macrophages in NDMM, it's reported to be a marker of SPP1+ macrophages. As a result, we still included MARCO for the geneset score calculation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.score_genes(ird_macro, gene_list = ['SPP1', 'CD274', 'MARCO'], score_name = 'ImmSupp_score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(ird_macro.obs['ImmSupp_score'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validate the expression of other markers that are reported to be associated with SPP1+ macrophages and MDSC activities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_macro.obs['ImmSupp_score_high'] = ['high' if x > 0 else 'low' for x in ird_macro.obs['ImmSupp_score']]\n",
    "dotplot = sc.pl.dotplot(ird_macro, var_names = ['SPP1', 'CD274', 'MARCO'], groupby = 'ImmSupp_score_high', return_fig = True)\n",
    "#dotplot.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_macrophage_ImmSuppScore_SPP1CD274MARCO.pdf', dpi = 300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dotplot = sc.pl.dotplot(ird_macro, var_names = ['TGFB1', 'MMP9', 'STAT1'], groupby = 'ImmSupp_score_high', return_fig = True)\n",
    "dotplot.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_macrophage_ImmSuppScore_TGFB1MMP9STAT1.pdf', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dotplot = sc.pl.dotplot(ird_macro, var_names = ['STAT3', 'NFKB1', 'NFKBIA'], groupby = 'ImmSupp_score_high', return_fig = True)\n",
    "dotplot.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_macrophage_ImmSuppScore_STAT3NFKB1NFKBIA.pdf', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dotplot = sc.pl.dotplot(ird_macro, var_names = ['TREM2', 'STAT5A', 'IFI44L', 'ISG15'], groupby = 'ImmSupp_score_high', return_fig = True)\n",
    "dotplot.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_macrophage_ImmSuppScore_TREM2STAT5AIFI44LISG15.pdf', dpi = 300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validate if macrophages with high immunosuppression scores are enriched in NDMM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dotplot = sc.pl.dotplot(ird_macro, var_names = 'ImmSupp_score', groupby = 'Collection', return_fig = True)\n",
    "dotplot.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_macrophage_ImmSuppScore_timepoint.pdf', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Percent macrophage with high SPP1 score per sample, grouped by Collection\n",
    "ird_macro.obs['ImmSupp_score_pos'] = ird_macro.obs['ImmSupp_score'] > 0.\n",
    "ird_macro_immSupp_summary = ird_macro.obs.groupby(['Sample'])['ImmSupp_score_pos'].mean().reset_index()\n",
    "ird_macro_immSupp_summary = ird_macro_immSupp_summary.merge(ird_macro.obs[['Sample', 'Collection']].drop_duplicates(), on='Sample', how='left')\n",
    "plot_utils.plot_comparison_with_significance(ird_macro_immSupp_summary, 'Collection', 'ImmSupp_score_pos', order=['NBM', 'NDMM', 'PT'], palette=timecols, \n",
    "                                       xlabel='Timepoint', ylabel='Fraction of Macrophages with High Immunosuppression Score', title=None)\n",
    "                                       #save_path='/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_macrophage_ImmSuppScoreFrac.pdf', figsize=(6, 5))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Validate if macrophages with high immunosuppression scores have higher expression of other SPP1/MDSC markers in NDMM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_macro_immSupp = ird_macro[ird_macro.obs['ImmSupp_score_high'] == 'high', :].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_macro_immSupp, var_names = ['TGFB1', 'MMP9', 'STAT1'], groupby = 'Collection', swap_axes = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_macro_immSupp, var_names = ['STAT3', 'NFKB1', 'NFKBIA'], groupby = 'Collection', swap_axes = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_macro_immSupp, var_names = ['TREM2', 'STAT5A', 'IFI44L', 'ISG15'], groupby = 'Collection', swap_axes = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spatial localization of immunosuppressive macrophages\n",
    "\n",
    "Are the macrophages enriched in certain radial neighborhoods? Radial neighborhoods were defined by Julia, which clusters neighbor composition within 50 microns of each cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rn_obj = sc.read_h5ad('/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/Xenium/analysis/radial_neighborhoods/Output/merged_RN.h5ad')\n",
    "rn_obj.obs.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count number of cells in each radial neighborhood\n",
    "rn_counts = rn_obj.obs['rn'].value_counts().reset_index()\n",
    "rn_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add radial neighborhood information to the macrophage data\n",
    "ird_macro.obs['rn'] = rn_obj.obs.loc[ird_macro.obs.index, 'rn'].values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the fraction of macrophages that are immunosuppressive in each radial neighborhood. They are more enriched in:\n",
    "- nbhd 8: CD8 T/NK cell neighborhood\n",
    "- nbhd 9: stromal neighborhood\n",
    "- nbhd 12 (few cells per sample): osteoblast neighborhood"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the number of macrophages in each radial neighborhood, grouped by ImmSupp_score_pos\n",
    "macro_rn_counts = (\n",
    "    ird_macro.obs\n",
    "    .groupby(['rn', 'ImmSupp_score_pos'])\n",
    "    .size()\n",
    "    .reset_index(name='macrophage_count')\n",
    ")\n",
    "macro_rn_counts = macro_rn_counts.merge(rn_counts, on='rn', how='left')\n",
    "macro_rn_counts['macrophage_frac'] = macro_rn_counts['macrophage_count'] / macro_rn_counts['count']\n",
    "display(macro_rn_counts)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.barplot(macro_rn_counts, x = 'macrophage_frac', y = 'rn', hue = 'ImmSupp_score_pos')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the fraction of macrophages that are immunnosuppressive per radial neighborhood on a sample level, and compare across collection timepoints."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "macro_per_sample_rn = ird_macro.obs.groupby(['Sample', 'rn']).agg(\n",
    "    macrophage_count = ('rn', 'count'),\n",
    "    immSupp_count = ('ImmSupp_score_pos', 'sum')\n",
    ").reset_index()\n",
    "macro_per_sample_rn['macrophage_frac'] = macro_per_sample_rn['macrophage_count'] / macro_per_sample_rn['macrophage_count'].sum()\n",
    "macro_per_sample_rn['immSupp_frac'] = macro_per_sample_rn['immSupp_count'] / macro_per_sample_rn['macrophage_count']\n",
    "macro_per_sample_rn['immSupp_frac'] = macro_per_sample_rn['immSupp_frac'].fillna(0)\n",
    "macro_per_sample_rn['Collection'] = macro_per_sample_rn['Sample'].map(ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')['Collection'])\n",
    "macro_per_sample_rn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ = plot_utils.plot_multigroup_boxplot_with_significance(macro_per_sample_rn, 'rn', 'immSupp_frac', 'Collection', figsize = (15, 6), show_outliers = False,\n",
    "                                                     xlabel = 'Radial neighborhood', ylabel = 'Fraction of macrophages that are immunosuppressive',\n",
    "                                                     palette = timecols, save_path = '/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_frac_ImmSupp_macrophage_rn_timepoint.pdf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualize spatial location of immunosuppressive macrophages\n",
    "\n",
    "Scatterplot of cell centroids of macrophages, grouped by their immunosuppressive gene expression, and plasma cells (tumor)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_sample = 'S13-35096A1U1'\n",
    "ird_macro_test_meta = ird_macro.obs.loc[ird_macro.obs['Sample'] == test_sample, :].copy()\n",
    "ird_pc_test_meta = ird_xenium_merge.obs.loc[(ird_xenium_merge.obs['Sample'] == test_sample) & (ird_xenium_merge.obs['ct'] == 'PC'), :].copy()\n",
    "ird_test_meta = pd.concat([ird_macro_test_meta, ird_pc_test_meta], axis = 0)\n",
    "ird_test_meta.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defind subtypes: PCs remain PCs, macrophages are split into ImmSupp and Non-ImmSupp\n",
    "ird_test_meta['ct_subtype'] = ['PC' if x == 'PC' else 'ImmSupp' if x == 'Macrophage' and y == 'high' else 'Non-ImmSupp' for x, y in zip(ird_test_meta['ct'], ird_test_meta['ImmSupp_score_high'])]\n",
    "ird_test_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize = (15, 15))\n",
    "sns.scatterplot(ird_test_meta, x = 'x_centroid', y = 'y_centroid', hue = 'ct_subtype', palette = 'muted', s = 2, linewidth = 0)\n",
    "# Maintain aspect ratio\n",
    "plt.gca().set_aspect('equal', 'box')\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', markerscale = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Archived"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene expression dotplot of immunosuppression genes only in the myeloid populations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_v6 = ird_xenium_merge[ird_xenium_merge.obs['Panel'] == 'BYGXJ6_hMulti'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_v6.obs['ct_timepoint'] = ird_v6.obs.apply(lambda x: f\"{x['ct']}_{x['Collection']}\", axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_myeloid = ird_v6[ird_v6.obs['ct'].isin(['GMP', 'Late Myeloid', 'Neutrophil', 'Ba/Eo/Ma/', 'Monocyte', 'Macrophage', 'pDC', 'cDC']), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_myeloid, var_names = ['CD274', 'IL1RL1'], groupby = 'ct_timepoint', swap_axes = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_myeloid, var_names = ['IL1R2'], groupby = 'ct_timepoint', swap_axes = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_myeloid, var_names = ['TGFB1', 'STAT3', 'STAT5A', 'S100A12'], groupby = 'ct_timepoint', swap_axes = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PDL1 expression and immunosuppressive score (old version) in all myeloid cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify immune suppressor genes in Mc/Mp, Granulo.\n",
    "genes = ['CD274', 'TGFB1', 'CTLA4', 'HAVCR2', 'LAG3']#, 'STAT3']\n",
    "\n",
    "myeloid_supp_df = ird_myeloid[:, genes].X.toarray()\n",
    "myeloid_supp_df_binary = myeloid_supp_df > 0\n",
    "ird_myeloid.obs['Immune_supp'] = np.any(myeloid_supp_df_binary, axis=1)\n",
    "ird_myeloid.obs['PDL1_pos'] = ird_myeloid[:, 'CD274'].X.toarray() > 0\n",
    "\n",
    "# Percent myeloid per sample expressing any of the one immuno suppresor genes\n",
    "myeloid_supp_pct = ird_myeloid.obs.groupby('Sample')['Immune_supp'].mean().reset_index()\n",
    "myeloid_supp_pct = myeloid_supp_pct.merge(ird_myeloid.obs.groupby('Sample')['PDL1_pos'].mean().reset_index(), on='Sample', how='left')\n",
    "myeloid_supp_pct = myeloid_supp_pct.merge(ird_myeloid.obs.groupby('Sample')['Immune_supp'].mean().reset_index(), on='Sample', how='left')\n",
    "myeloid_supp_pct = myeloid_supp_pct.merge(ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates(), on='Sample', how='left')\n",
    "myeloid_supp_pct.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, stats = plot_utils.plot_comparison_with_significance(\n",
    "        myeloid_supp_pct, \n",
    "        x_col='Collection', \n",
    "        y_col='PDL1_pos',\n",
    "        order=['NBM', 'NDMM', 'PT'],\n",
    "        ylabel='Fraction of PDL1+ myeloid cells'\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look at paired data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify paired samples in NDMM and PT, they will have the same UPN (already in meta)\n",
    "ird_meta = ird_xenium_merge.obs.copy()\n",
    "ird_meta[['UPN', 'Collection']].drop_duplicates().value_counts('UPN').reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look at paired samples in NDMM and PT, they will have the same UPN (already in meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "paired_UPNs = ['WU030', 'WU025', 'WU107', 'WU068', 'WU066', 'WU050', 'WU007', 'WU043']\n",
    "upn_sample_meta = ird_meta[['Sample', 'UPN', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "myeloid_supp_pct['UPN'] = myeloid_supp_pct['Sample'].map(upn_sample_meta['UPN'])\n",
    "myeloid_supp_paired_pct = myeloid_supp_pct[myeloid_supp_pct['UPN'].isin(paired_UPNs)].copy()\n",
    "myeloid_supp_paired_pct['Collection'] = pd.Categorical(myeloid_supp_paired_pct['Collection'], categories=['NDMM', 'PT'], ordered=True)\n",
    "# Some samples are from same UPN, need to average the percentages per collection\n",
    "myeloid_supp_paired_pct = myeloid_supp_paired_pct.groupby(['UPN', 'Collection']).agg({'PDL1_pos': 'mean'}).reset_index().dropna()\n",
    "sns.boxplot(myeloid_supp_paired_pct, x='Collection', y='PDL1_pos', palette='Set2')\n",
    "\n",
    "# Plot line connecting paired samples\n",
    "for upn in paired_UPNs:\n",
    "    data_upn = myeloid_supp_paired_pct[myeloid_supp_paired_pct['UPN'] == upn]\n",
    "    plt.plot(data_upn['Collection'], data_upn['PDL1_pos'], color='gray', linestyle='--', marker='o')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "paired_UPNs = ['WU030', 'WU025', 'WU107', 'WU068', 'WU066', 'WU050', 'WU007', 'WU043']\n",
    "upn_sample_meta = ird_meta[['Sample', 'UPN', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "myeloid_supp_pct['UPN'] = myeloid_supp_pct['Sample'].map(upn_sample_meta['UPN'])\n",
    "myeloid_supp_paired_pct = myeloid_supp_pct[myeloid_supp_pct['UPN'].isin(paired_UPNs)].copy()\n",
    "myeloid_supp_paired_pct['Collection'] = pd.Categorical(myeloid_supp_paired_pct['Collection'], categories=['NDMM', 'PT'], ordered=True)\n",
    "# Some samples are from same UPN, need to average the percentages per collection\n",
    "myeloid_supp_paired_pct = myeloid_supp_paired_pct.groupby(['UPN', 'Collection']).agg({'Immune_supp': 'mean'}).reset_index().dropna()\n",
    "sns.boxplot(myeloid_supp_paired_pct, x='Collection', y='Immune_supp', palette='Set2')\n",
    "\n",
    "# Plot line connecting paired samples\n",
    "for upn in paired_UPNs:\n",
    "    data_upn = myeloid_supp_paired_pct[myeloid_supp_paired_pct['UPN'] == upn]\n",
    "    plt.plot(data_upn['Collection'], data_upn['Immune_supp'], color='gray', linestyle='--', marker='o')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate p-value using Wilcoxon signed-rank test\n",
    "from scipy.stats import wilcoxon\n",
    "# Extract percentage values for NDMM and PT for each paired UPN, make sure they are in the same order\n",
    "ndmm_pct_df = myeloid_supp_paired_pct[myeloid_supp_paired_pct['Collection'] == 'NDMM'].sort_values('UPN')\n",
    "pt_pct_df = myeloid_supp_paired_pct[myeloid_supp_paired_pct['Collection'] == 'PT'].sort_values('UPN')\n",
    "\n",
    "# Perform Wilcoxon signed-rank test\n",
    "wilcoxon(ndmm_pct_df['PDL1_pos'].values, pt_pct_df['PDL1_pos'].values).pvalue\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Distance between immunosuppressive myeloid cells (e.g. PDL1+) and (exhaustive) T cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check if CD274+ myeloid cells are around PDCD1+ T cells - answer seems to be \"no\", maybe due to sparse transcript detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exh_genes = ['HAVCR2', 'PDCD1', 'LAG3', 'CTLA4', 'TIGIT']\n",
    "ird_tcells = ird_xenium_merge[ird_xenium_merge.obs['annot'] == 'T', :]\n",
    "exh_df = ird_tcells[:, exh_genes].X.toarray()\n",
    "exh_df_binary = exh_df > 0\n",
    "ird_tcells.obs['Exh_pos'] = np.any(exh_df_binary, axis=1)\n",
    "ird_tcells.obs['PDCD1_pos'] = ird_tcells[:, 'PDCD1'].X.toarray() > 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check if CD274+ Mc/Mp are around PDCD1+ T cells\n",
    "PDL1_pos_mcmp = ird_myeloid.obs.loc[ird_myeloid.obs['Immune_supp'] == True, ['Sample', 'x_centroid', 'y_centroid']].copy()\n",
    "PDCD1_tcell_pos = ird_tcells.obs.loc[ird_tcells.obs['Exh_pos'] == True, ['Sample', 'x_centroid', 'y_centroid']].copy()\n",
    "\n",
    "res = nearest_dist_between_two_celltypes(PDL1_pos_mcmp, PDCD1_tcell_pos, 'Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(res, x='nearest_dist_to_df2', binwidth = 10)\n",
    "plt.xlim(0, 200)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look at the number of MDSCs within 20 microns of exh. T cells in each sample, normalized by the number of cells within 20 microns of each exh. T cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial import KDTree\n",
    "all_cell_pos = ird_xenium_merge.obs[['Sample', 'x_centroid', 'y_centroid']].copy()\n",
    "PDL1_pos_mcmp = ird_myeloid.obs.loc[ird_myeloid.obs['PDL1_pos'] == True, ['Sample', 'x_centroid', 'y_centroid']].copy()\n",
    "PDCD1_tcell_pos = ird_tcells.obs.loc[ird_tcells.obs['PDCD1_pos'] == True, ['Sample', 'x_centroid', 'y_centroid']].copy()\n",
    "all_cell_pos = all_cell_pos.drop(PDCD1_tcell_pos.index)\n",
    "\n",
    "res = []\n",
    "distance_threshold = 20\n",
    "for sample in PDL1_pos_mcmp['Sample'].unique():\n",
    "    # Extract positions of MDSCs, Exh. T cells, and all cells in the sample\n",
    "    mdsc_pos = PDL1_pos_mcmp[PDL1_pos_mcmp['Sample'] == sample]\n",
    "    tcell_pos = PDCD1_tcell_pos[PDCD1_tcell_pos['Sample'] == sample]\n",
    "    cell_pos = all_cell_pos[all_cell_pos['Sample'] == sample]\n",
    "\n",
    "    # Build KDTree for T cells and MDSCs for distance query\n",
    "    t_tree = KDTree(tcell_pos[['x_centroid', 'y_centroid']])\n",
    "    mdsc_tree = KDTree(mdsc_pos[['x_centroid', 'y_centroid']])\n",
    "    cell_tree = KDTree(cell_pos[['x_centroid', 'y_centroid']])\n",
    "        \n",
    "    # For Metric 1: MDSCs within 20 microns of T cells / All cells within 20 microns of T cells\n",
    "    ind_mdsc_near_t = mdsc_tree.query_ball_point(tcell_pos[['x_centroid', 'y_centroid']], r=distance_threshold)\n",
    "    ind_cell_near_t = cell_tree.query_ball_point(tcell_pos[['x_centroid', 'y_centroid']], r=distance_threshold)\n",
    "    \n",
    "    total_mdscs_around_tcells = sum(len(neighbors) for neighbors in ind_mdsc_near_t)\n",
    "    total_cells_around_tcells = sum(len(neighbors) for neighbors in ind_cell_near_t)\n",
    "    metric1 = total_mdscs_around_tcells / total_cells_around_tcells if total_cells_around_tcells > 0 else 0\n",
    "    \n",
    "    # For Metric 2: T cells with at least 1 MDSC neighbor / Total T cells\n",
    "    ind_t_near_mdsc = mdsc_tree.query_ball_point(tcell_pos[['x_centroid', 'y_centroid']], r=distance_threshold)\n",
    "    total_tcells_with_mdsc = sum(1 for neighbors in ind_t_near_mdsc if len(neighbors) > 0)\n",
    "    total_tcells = len(tcell_pos)\n",
    "    metric2 = total_tcells_with_mdsc / total_tcells if total_tcells > 0 else 0\n",
    "    \n",
    "    res.append({\n",
    "        'Sample': sample,\n",
    "        'frac_mdsc_around_tcells': metric1,\n",
    "        'frac_tcells_with_mdsc': metric2,\n",
    "        'total_mdscs_around_tcells': total_mdscs_around_tcells,\n",
    "        'total_cells_around_tcells': total_cells_around_tcells,\n",
    "        'total_tcells_with_mdsc': total_tcells_with_mdsc,\n",
    "        'total_tcells': total_tcells\n",
    "        })\n",
    "res = pd.DataFrame(res)\n",
    "res['Collection'] = res['Sample'].map(ird_meta[['Sample', 'Collection']].drop_duplicates().set_index('Sample')['Collection'])\n",
    "# Normalize by total number of myeloid cells per sample\n",
    "total_myeloid = ird_myeloid.obs.groupby('Sample').size().reset_index(name='total_myeloid')\n",
    "res = res.merge(total_myeloid, on='Sample', how='left')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(res, x='Collection', y='frac_mdsc_around_tcells', palette='Set2')\n",
    "sns.swarmplot(res, x='Collection', y='frac_mdsc_around_tcells', color='black', size=8, alpha = .5)\n",
    "plt.xlabel('Timepoint')\n",
    "plt.ylabel('Fraction of exh. T cells neighbors being MDSCs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.boxplot(res, x='Collection', y='frac_tcells_with_mdsc', palette='Set2')\n",
    "sns.swarmplot(res, x='Collection', y='frac_tcells_with_mdsc', color='black', size=8, alpha = .5)\n",
    "plt.xlabel('Timepoint')\n",
    "plt.ylabel('Fraction of exh. T cells having MDSC neighbors')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 4: Stromal Niche Dysfunction - MSC Analysis\n",
    "\n",
    "Characterize mesenchymal stromal cells (MSCs) and their functional changes across disease stages. MSCs are critical components of the bone marrow niche that support hematopoiesis and regulate plasma cell survival through factors like CXCL12. We test if CXCL12 expression in MSCs are significantly altered in disease and treatment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_msc = ird_xenium_merge[ird_xenium_merge.obs['ct'] == 'MSC'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_msc_v6 = ird_msc[ird_msc.obs['Panel'] == 'BYGXJ6_hMulti'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(ird_msc_v6, groupby ='Collection', method = 'wilcoxon')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups(ird_msc_v6, n_genes = 25, show = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_xenium_merge, var_names = ['CD74', 'FTH1'], groupby = 'annot_timepoint', swap_axes = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plasma cell contamination issue\n",
    "\n",
    "Initial differential expression analysis reveals MSCs in NDMM samples are heavily confounded by plasma cell gene expression. This contamination likely arises from:\n",
    "1. **Transcript leakage**: High plasma cell abundance leads transcript leakage to neighboring cells\n",
    "2. **Segmentation errors**: Cells near plasma cells may incorrectly capture plasma transcripts\n",
    "\n",
    "**Solution**: Apply stringent filtering to remove MSCs expressing canonical plasma cell markers before analyzing MSC-specific gene expression changes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stringent filtering - remove MSCs with any expression of plasma cell genes SDC1, SLAMF7, MZB1, TNFRSF17\n",
    "msc_PCexp = ird_msc[:, ['SDC1', 'SLAMF7', 'MZB1', 'TNFRSF17', 'B2M', 'TENT5C', 'CD74']].X.toarray().sum(axis=1)\n",
    "ird_msc.obs['PC_exp'] = msc_PCexp > 1e-6\n",
    "ird_msc_PC_filtered = ird_msc[ird_msc.obs['PC_exp'] == False].copy()\n",
    "len(ird_msc_PC_filtered)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CXCL12/CXCR4 axis analysis\n",
    "\n",
    "The **CXCL12-CXCR4 axis** is critical for plasma cell homing and retention in the bone marrow niche:\n",
    "\n",
    "- **CXCL12** (SDF-1): Chemokine produced by MSCs and other stromal cells that attracts CXCR4+ cells\n",
    "- **CXCR4**: Chemokine receptor expressed on plasma cells and B cell precursors\n",
    "\n",
    "**Analysis approach:**\n",
    "1. Quantify CXCL12 expression in filtered MSCs across disease stages\n",
    "2. Quantify CXCR4 expression in B-lineage cells (Early B, Mature B, Plasma cells)\n",
    "\n",
    "\n",
    "**Interpretation notes:**\n",
    "- Decreased MSC CXCL12 may indicate stromal dysfunction and reduced niche support\n",
    "- Altered CXCR4 levels in plasma cells may affect their retention and survival in the niche\n",
    "- Cellchat analysis using Xenium data (done separately) indicate that CXCL12-CXCR4 signaling axis is affected\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(ird_msc_PC_filtered, groupby ='Collection', method = 'wilcoxon')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups(ird_msc_PC_filtered, n_genes = 25, show = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_msc_PC_filtered_v6 = ird_msc_PC_filtered[ird_msc_PC_filtered.obs['Panel'] == 'BYGXJ6_hMulti'].copy()\n",
    "sc.tl.rank_genes_groups(ird_msc_PC_filtered_v6, groupby ='Collection', method = 'wilcoxon')\n",
    "sc.pl.rank_genes_groups(ird_msc_PC_filtered_v6, n_genes = 25, show = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_msc_PC_filtered, var_names = ['CXCL12'], groupby = 'Collection')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_msc_PC_filtered.obs['CXCL12_exp'] = ird_msc_PC_filtered[:, 'CXCL12'].X.toarray().flatten()\n",
    "cxcl12_exp = ird_msc_PC_filtered.obs.groupby('Sample')['CXCL12_exp'].mean().reset_index()\n",
    "cxcl12_exp['Collection'] = cxcl12_exp['Sample'].map(ird_msc_PC_filtered.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')['Collection'])\n",
    "\n",
    "timecols = {\"NBM\": \"#0C7515\", \"NDMM\": \"#E619B9\", \"PT\": \"#CF99C3\"} \n",
    "plot_utils.plot_comparison_with_significance(cxcl12_exp, 'Collection', 'CXCL12_exp', order=['NBM', 'NDMM', 'PT'], palette=timecols, \n",
    "                                       xlabel='Timepoint', ylabel='Mean CXCL12 Expression in MSCs', title=None,\n",
    "                                       save_path='/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_Xenium_merge_CXCL12_expression_boxplot_v111125.pdf', figsize=(6, 5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Boxplot of mean CXCR4 expression in PC, Mature B, and Early B cells per sample, grouped by Collection\n",
    "ird_B_lineage = ird_xenium_merge[ird_xenium_merge.obs['ct'].isin(['PC', 'Mature B', 'Early B'])].copy()\n",
    "ird_B_lineage.obs['CXCR4_exp'] = ird_B_lineage[:, 'CXCR4'].X.toarray().flatten()\n",
    "cxcr4_exp = ird_B_lineage.obs.groupby(['Sample', 'ct'])['CXCR4_exp'].mean().reset_index()\n",
    "cxcr4_exp['Collection'] = cxcr4_exp['Sample'].map(ird_B_lineage.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')['Collection'])\n",
    "\n",
    "timecols = {\"NBM\": \"#0C7515\", \"NDMM\": \"#E619B9\", \"PT\": \"#CF99C3\"} \n",
    "order = cxcr4_exp['ct'].unique().tolist()          # set your desired x order\n",
    "hue_order = sorted(cxcr4_exp['Collection'].unique())  # set your desired hue order\n",
    "\n",
    "plot_utils.plot_multigroup_boxplot_with_significance(cxcr4_exp, x_col='ct', y_col='CXCR4_exp', hue_col='Collection', \n",
    "                                                     order=order, hue_order=hue_order, show_swarm=True, \n",
    "                                                     palette=timecols, figsize=(6, 5), xlabel='Timepoint', ylabel='Mean CXCR4 expression', title=None, save_path=None)\n",
    "plt.legend(bbox_to_anchor=(1.0, 1), loc='upper left')\n",
    "plt.ylim(-0.1, 1.1)\n",
    "plt.tight_layout()\n",
    "plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_Xenium_merge_CXCR4_expression_boxplot_v111125.pdf', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_xenium_merge, var_names = ['CXCL12'], groupby = 'ct_timepoint', swap_axes = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_xenium_merge, var_names = ['CXCR4'], groupby = 'annot_timepoint', swap_axes = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(ird_msc_PC_filtered, var_names = ['DKK1'], groupby = 'Collection', dot_max = .02)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distance between MSCs and B/Plasma cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "msc_df = ird_msc.obs[['Sample', 'x_centroid', 'y_centroid']].copy()\n",
    "b_df = ird_xenium_merge.obs[ird_xenium_merge.obs['ct'] == 'Mature B'][['Sample', 'x_centroid', 'y_centroid']].copy()\n",
    "earlyB_df = ird_xenium_merge.obs[ird_xenium_merge.obs['ct'] == 'Early B'][['Sample', 'x_centroid', 'y_centroid']].copy()\n",
    "PC_df = ird_xenium_merge.obs[ird_xenium_merge.obs['ct'] == 'PC'][['Sample', 'x_centroid', 'y_centroid']].copy()\n",
    "msc_B_dist = spatial_utils.nearest_dist_between_two_celltypes(msc_df, b_df, sample_col = 'Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "msc_earlyB_dist = spatial_utils.nearest_dist_between_two_celltypes(msc_df, earlyB_df, sample_col = 'Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "msc_PC_dist = spatial_utils.nearest_dist_between_two_celltypes(msc_df, PC_df, sample_col = 'Sample', x_col='x_centroid', y_col='y_centroid')\n",
    "msc_B_dist['ref_ct'] = 'B cell'\n",
    "msc_earlyB_dist['ref_ct'] = 'Early B'\n",
    "msc_PC_dist['ref_ct'] = 'Plasma cell'\n",
    "msc_dists = pd.concat([msc_B_dist, msc_earlyB_dist, msc_PC_dist], axis = 0)\n",
    "msc_dists.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import gaussian_kde\n",
    "\n",
    "res = []\n",
    "for sample in msc_dists['Sample'].unique():\n",
    "    sample_df = msc_dists[msc_dists['Sample'] == sample]\n",
    "    for ref_ct in sample_df['ref_ct'].unique():\n",
    "        sample_ref_df = sample_df[sample_df['ref_ct'] == ref_ct]\n",
    "        if len(sample_ref_df) < 10:\n",
    "            continue\n",
    "        # Estimate distance using KDE\n",
    "        kde = gaussian_kde(sample_ref_df['nearest_dist_to_df2'])\n",
    "        dist_eval = np.arange(0, 501, 1)\n",
    "        kde_pdf = kde.pdf(dist_eval)\n",
    "        most_likely_dist = dist_eval[np.argmax(kde_pdf)]\n",
    "        res.append({\n",
    "            'Sample': sample,\n",
    "            'Reference Cell Type': ref_ct,\n",
    "            'Most Likely Distance': most_likely_dist,\n",
    "            'Average Distance': sample_ref_df['nearest_dist_to_df2'].mean(),\n",
    "            'Median Distance': sample_ref_df['nearest_dist_to_df2'].median()\n",
    "        })\n",
    "\n",
    "most_likely_dists_df = pd.DataFrame(res)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "most_likely_dists_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "collection_meta = ird_xenium_merge.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "most_likely_dists_df['Collection'] = most_likely_dists_df['Sample'].map(collection_meta['Collection'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_utils.plot_multigroup_boxplot_with_significance(most_likely_dists_df, x_col = 'Reference Cell Type', y_col = 'Most Likely Distance', hue_col = 'Collection', \n",
    "                                                    order=None, hue_order=None, show_swarm=True, palette=timecols, figsize=(6, 5), xlabel='Timepoint', ylabel=r'Distance to MSCs ($\\mu$m)', \n",
    "                                                    title=None, save_path='/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_MSC_PCB_distance.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 5: Spatial Neighbor Cell Type Composition\n",
    "\n",
    "Analyze the cellular neighborhood (nearest neighbors) composition around each cell type using spatial proximity. This approach identifies which cell types preferentially localize near each other, revealing potential cellular interactions and microenvironmental niches.\n",
    "\n",
    "### Nearest neighbor identification: Delaunay + KDTree hybrid approach\n",
    "\n",
    "The neighbor detection combines two complementary methods:\n",
    "1. **Delaunay triangulation**: Identifies neighbors based on spatial tessellation\n",
    "2. **Distance threshold (KDTree)**: Filters neighbors by maximum distance, removing distant connections from the triangulation\n",
    "\n",
    "Only cells that satisfy **both** criteria (Delaunay neighbors within distance threshold) are retained as true spatial neighbors. This hybrid approach avoids spurious long-range connections while maintaining biologically relevant local neighborhoods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial import Delaunay, KDTree\n",
    "def get_proximal_neighbors(df, x_col='x_centroid', y_col='y_centroid', distance_threshold=50):\n",
    "    \"\"\"\n",
    "    Identify proximal spatial neighbors using Delaunay triangulation intersected with distance threshold.\n",
    "    \n",
    "    This hybrid approach combines:\n",
    "    1. Delaunay triangulation to find neighbors\n",
    "    2. KDTree radius search to enforce distance constraints\n",
    "    \n",
    "    Only cells that are BOTH Delaunay neighbors AND within the distance threshold are retained,\n",
    "    ensuring biologically relevant local neighborhoods without spurious long-range connections.\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pd.DataFrame\n",
    "        DataFrame containing cell coordinates with an index that will be preserved\n",
    "    x_col : str\n",
    "        Column name for x-coordinates (default: 'x_centroid')\n",
    "    y_col : str  \n",
    "        Column name for y-coordinates (default: 'y_centroid')\n",
    "    distance_threshold : float\n",
    "        Maximum distance (in same units as coordinates) for considering cells as neighbors.\n",
    "        Typical values: 20-50 microns for Xenium data\n",
    "\n",
    "    Returns:\n",
    "    --------\n",
    "    dict\n",
    "        Dictionary mapping each cell index to a list of neighbor cell indices.\n",
    "        Keys and values are original DataFrame indices.\n",
    "        \n",
    "    Example:\n",
    "    --------\n",
    "    # Get neighbors within 25 microns for a sample\n",
    "    sample_data = cell_info[cell_info['Sample'] == 'Sample1']\n",
    "    neighbors = get_proximal_neighbors(sample_data, distance_threshold=25)\n",
    "    \n",
    "    # Access neighbors of first cell\n",
    "    first_cell_idx = sample_data.index[0]\n",
    "    first_cell_neighbors = neighbors[first_cell_idx]\n",
    "    \"\"\"\n",
    "    if not isinstance(df, pd.DataFrame):\n",
    "        raise ValueError(\"Input 'df' must be a pandas DataFrame.\")\n",
    "    if x_col not in df.columns:\n",
    "        raise ValueError(f\"Column '{x_col}' not found in DataFrame.\")\n",
    "    if y_col not in df.columns:\n",
    "        raise ValueError(f\"Column '{y_col}' not found in DataFrame.\")\n",
    "\n",
    "    points = df[[x_col, y_col]].values\n",
    "    original_indices_map = df.index # To map points array indices back to original df indices\n",
    "\n",
    "    \n",
    "    if len(points) < 3:\n",
    "        # Delaunay triangulation requires at least 3 points to form a simplex (triangle in 2D).\n",
    "        print(\"Not enough points to perform Delaunay triangulation. Need at least 3 points.\")\n",
    "        return {idx: [] for idx in df.index} # Return mapping to original DataFrame indices\n",
    "        \n",
    "    tri = Delaunay(points)\n",
    "\n",
    "    # tri.vertex_neighbor_vertices is a CSR-like structure:\n",
    "    # indptr[i] to indptr[i+1] gives the slice in 'indices' for the i-th point in the 'points' array\n",
    "    # indices[indptr[i]:indptr[i+1]] gives the actual indices (referring to the 'points' array)\n",
    "    # of the neighbors for the i-th point.\n",
    "    indptr, delaunay_indices = tri.vertex_neighbor_vertices\n",
    "    kdtree = KDTree(points)\n",
    "\n",
    "    proximal_neighbors = {}\n",
    "    for i in range(len(points)):\n",
    "        current_original_idx = original_indices_map[i]\n",
    "        \n",
    "        # a. Get Delaunay neighbors (indices relative to 'points' array)\n",
    "        delaunay_neighbors_for_i = delaunay_indices[indptr[i]:indptr[i+1]]\n",
    "        delaunay_set = set(delaunay_neighbors_for_i)\n",
    "        \n",
    "        # b. Get K-D tree neighbors (indices relative to 'points' array)\n",
    "        # query_ball_point returns a list of indices of points within the radius\n",
    "        kdtree_neighbors_for_i = kdtree.query_ball_point(points[i], r=distance_threshold)\n",
    "        # Remove the point itself from its k-d tree neighbors\n",
    "        kdtree_set = set(kdtree_neighbors_for_i) - {i} \n",
    "        \n",
    "        # c. Intersect the two sets\n",
    "        intersected_neighbor_indices_in_points_array = list(delaunay_set.intersection(kdtree_set))\n",
    "        \n",
    "        # d. Map intersected indices back to original DataFrame indices\n",
    "        final_neighbor_original_indices = [original_indices_map[n_idx] for n_idx in intersected_neighbor_indices_in_points_array]\n",
    "        \n",
    "        proximal_neighbors[current_original_idx] = final_neighbor_original_indices\n",
    "        \n",
    "    return proximal_neighbors\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_xenium_merge = sc.read_h5ad(\"/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/Xenium/analysis/merged_filtered.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_xenium_merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_cell_info = ird_xenium_merge.obs.copy()\n",
    "ird_cell_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_cell_info['annot'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_cell_info['Collection'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "def neighboring_celltype_counts(cell_info, ct_col_name, ct_of_interest, neighbors_dict):\n",
    "    \"\"\"\n",
    "    Count the cell type composition of neighbors around a specific cell type.\n",
    "    \n",
    "    For all cells of a given type, aggregates their neighbors' cell types to understand\n",
    "    the composition around that cell type.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    cell_info : pd.DataFrame\n",
    "        DataFrame with cell metadata including cell type annotations\n",
    "    ct_col_name : str\n",
    "        Column name containing cell type annotations (e.g., 'annot', 'group')\n",
    "    ct_of_interest : str\n",
    "        Specific cell type to analyze (e.g., 'PC', 'MSC', 'T')\n",
    "    neighbors_dict : dict\n",
    "        Output from get_proximal_neighbors() function mapping cell indices to neighbor indices\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    dict\n",
    "        Dictionary mapping each cell type to the count of how many times it appears\n",
    "        as a neighbor of the cell type of interest. All cell types in the dataset\n",
    "        are included (with 0 for types that never appear as neighbors).\n",
    "        \n",
    "    Example output:\n",
    "    ---------------\n",
    "    {'PC': 1250, 'MSC': 450, 'T': 890, 'B': 120, ...}\n",
    "    # Interpretation: Plasma cells (PC) have 1250 plasma cell neighbors total,\n",
    "    # 450 MSC neighbors, 890 T cell neighbors, etc.\n",
    "    \n",
    "    Example usage:\n",
    "    --------------\n",
    "    # Count what cell types are around plasma cells in a sample\n",
    "    neighbors = get_proximal_neighbors(sample_data)\n",
    "    pc_neighborhood = neighboring_celltype_counts(\n",
    "        sample_data, \n",
    "        ct_col_name='annot',\n",
    "        ct_of_interest='PC',\n",
    "        neighbors_dict=neighbors\n",
    "    )\n",
    "    \"\"\"\n",
    "    ct_index = cell_info[cell_info[ct_col_name] == ct_of_interest].index.values\n",
    "    ct_neighbor_indices = [neighbors_dict[i] for i in ct_index]\n",
    "    ct_neighbor_indices_flat = [item for sublist in ct_neighbor_indices for item in sublist]\n",
    "    ct_neighbors = [cell_info[ct_col_name][i] for i in ct_neighbor_indices_flat]\n",
    "    ct_neighbors_counter = Counter(ct_neighbors)\n",
    "    \n",
    "    # Get all unique cell types\n",
    "    all_cell_types = np.unique(cell_info[ct_col_name])\n",
    "    # Create dictionary with zeros for all cell types, then update with actual counts\n",
    "    ct_neighbors_counts = {ct: 0 for ct in all_cell_types}\n",
    "    ct_neighbors_counts.update(ct_neighbors_counter)\n",
    "\n",
    "    return ct_neighbors_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = ird_cell_info[ird_cell_info['Sample'] == 'IRD_S18-30740A1U3']\n",
    "all_nbs = get_proximal_neighbors(test, x_col='x_centroid', y_col='y_centroid', distance_threshold=25)\n",
    "nb_ct = neighboring_celltype_counts(test, 'annot', 'PC', all_nbs)\n",
    "nb_ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_neighboring_celltype_analysis(df, ct_col, sample_col, distance_threshold=25, x_col='x_centroid', y_col='y_centroid'):\n",
    "    \"\"\"\n",
    "    Calculates neighbor composition across all cell types and samples. Calls neighboring_celltype_counts().\n",
    "\n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : pd.DataFrame\n",
    "        DataFrame with cell metadata including coordinates, cell types, and sample IDs\n",
    "    ct_col : str\n",
    "        Column name for cell type annotations\n",
    "    sample_col : str\n",
    "        Column name for sample identifiers\n",
    "    distance_threshold : float\n",
    "        Maximum distance (microns) for neighbor consideration (default: 25)\n",
    "    x_col : str\n",
    "        Column name for x-coordinates (default: 'x_centroid')\n",
    "    y_col : str\n",
    "        Column name for y-coordinates (default: 'y_centroid')\n",
    "        \n",
    "    Returns:\n",
    "        pd.DataFrame: DataFrame with columns for cell type, neighboring cell type, sample ID, and counts of neighboring cell types.\n",
    "    \"\"\"\n",
    "\n",
    "    # Initialize an empty DataFrame to store results\n",
    "    all_nb_df = pd.DataFrame()\n",
    "\n",
    "    # Loop through each sample\n",
    "    for sample in df[sample_col].unique().tolist():\n",
    "        sample_df = df[df[sample_col] == sample]\n",
    "        # Get proximal neighbor dict for all cells\n",
    "        all_neighbors = get_proximal_neighbors(sample_df, x_col=x_col, y_col=y_col, distance_threshold=distance_threshold)\n",
    "\n",
    "        for ct in sample_df[ct_col].unique().tolist():       \n",
    "            # Count neighboring cell types of all cell types\n",
    "            nb_counts = neighboring_celltype_counts(sample_df, ct_col, ct, all_neighbors)\n",
    "            nb_df = pd.DataFrame.from_dict(nb_counts, orient = \"index\", columns = [\"counts\"])\n",
    "            nb_df['Cell Type'] = ct\n",
    "            nb_df['Neighboring Cell Type'] = nb_df.index\n",
    "            nb_df['Sample'] = sample\n",
    "            nb_df.reset_index(drop = True, inplace = True)\n",
    "            all_nb_df = pd.concat([all_nb_df, nb_df], axis = 0)\n",
    "            \n",
    "        print(f\"Finished processing sample {sample}\")\n",
    "    \n",
    "    return all_nb_df\n",
    "\n",
    "\n",
    "all_nb_df = run_neighboring_celltype_analysis(ird_cell_info, 'annot', 'Sample', distance_threshold = 25, x_col='x_centroid', y_col='y_centroid')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scaled neighboring cell type composition heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_nb_df.to_csv(\"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_Xenium_merge_neighboringCelltype_counts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find sample ID whose collection is NBM\n",
    "nbm_samples = ird_cell_info[ird_cell_info['Collection'] == 'NBM']['Sample'].unique().tolist()\n",
    "ndmm_samples = ird_cell_info[ird_cell_info['Collection'] == 'NDMM']['Sample'].unique().tolist()\n",
    "pt_samples = ird_cell_info[ird_cell_info['Collection'] == 'PT']['Sample'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate z score for each neighboring cell type across cell types within each sample\n",
    "all_nb_df['pct'] = all_nb_df['counts'] / all_nb_df.groupby(['Sample', 'Cell Type'])['counts'].transform('sum') * 100\n",
    "all_nb_df['zscore_nct'] = all_nb_df.groupby(['Sample', 'Neighboring Cell Type'])['pct'].transform(lambda x: (x - x.mean()) / x.std())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nbm_nb_df = all_nb_df[all_nb_df['Sample'].isin(nbm_samples)]\n",
    "ndmm_nb_df = all_nb_df[all_nb_df['Sample'].isin(ndmm_samples)]\n",
    "pt_nb_df = all_nb_df[all_nb_df['Sample'].isin(pt_samples)]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate mean and std of pct for each cell type and neighboring cell type, within each sample, grouped by disease\n",
    "summary_nbm_df = nbm_nb_df.groupby(['Cell Type', 'Neighboring Cell Type']).agg(\n",
    "    mean_zscore_nct=('zscore_nct', 'mean')\n",
    ").reset_index()\n",
    "summary_nbm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_nbm_df['mean_pct_scaled'] = summary_nbm_df.groupby('Neighboring Cell Type')['mean_zscore_nct'].transform(\n",
    "    lambda x: (x - x.mean()) / x.std() if x.std() > 0 else 0\n",
    ")\n",
    "# Draw heatmap of scaled mean pct\n",
    "plt.figure(figsize = (12, 8))\n",
    "heatmap_data = summary_nbm_df.pivot(index = 'Cell Type', columns = 'Neighboring Cell Type', values = 'mean_zscore_nct')\n",
    "sns.heatmap(heatmap_data, cmap = 'vlag', center = 0, cbar_kws={'label': 'Mean Scaled Neighboring Cell Type Composition'})\n",
    "plt.title(\"NBM samples\")\n",
    "plt.tight_layout()\n",
    "#plt.savefig(\"/diskmnt/Users2/chouw/Projects/BM_spatial/NBM_MM_Xenium_merge_WC_neighboringCelltype_composition_healthy_samples_heatmap_v09192025.pdf\", dpi = 300, transparent = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_ndmm_df = ndmm_nb_df.groupby(['Cell Type', 'Neighboring Cell Type']).agg(\n",
    "    mean_zscore_nct=('zscore_nct', 'mean')\n",
    ").reset_index()\n",
    "summary_ndmm_df['mean_pct_scaled'] = summary_ndmm_df.groupby('Neighboring Cell Type')['mean_zscore_nct'].transform(\n",
    "    lambda x: (x - x.mean()) / x.std() if x.std() > 0 else 0\n",
    ")\n",
    "# Draw heatmap of scaled mean pct\n",
    "plt.figure(figsize = (12, 8))\n",
    "heatmap_data = summary_ndmm_df.pivot(index = 'Cell Type', columns = 'Neighboring Cell Type', values = 'mean_zscore_nct')\n",
    "sns.heatmap(heatmap_data, cmap = 'vlag', center = 0, cbar_kws={'label': 'Mean Scaled Neighboring Cell Type Composition'})\n",
    "plt.title(\"NDMM samples\")\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "summary_pt_df = pt_nb_df.groupby(['Cell Type', 'Neighboring Cell Type']).agg(\n",
    "    mean_zscore_nct=('zscore_nct', 'mean')\n",
    ").reset_index()\n",
    "summary_pt_df['mean_pct_scaled'] = summary_pt_df.groupby('Neighboring Cell Type')['mean_zscore_nct'].transform(\n",
    "    lambda x: (x - x.mean()) / x.std() if x.std() > 0 else 0\n",
    ")\n",
    "# Draw heatmap of scaled mean pct\n",
    "plt.figure(figsize = (12, 8))\n",
    "heatmap_data = summary_pt_df.pivot(index = 'Cell Type', columns = 'Neighboring Cell Type', values = 'mean_zscore_nct')\n",
    "sns.heatmap(heatmap_data, cmap = 'vlag', center = 0, cbar_kws={'label': 'Mean Scaled Neighboring Cell Type Composition'})\n",
    "plt.title(\"PT samples\")\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison between timepoints of specific cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_nb_df['pct'] = all_nb_df['counts'] / all_nb_df.groupby(['Sample', 'Cell Type'])['counts'].transform('sum') * 100\n",
    "metadata = ird_cell_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "all_nb_df['Collection'] = all_nb_df['Sample'].map(metadata['Collection'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "B_nb_df = all_nb_df[all_nb_df['Cell Type'] == 'B']\n",
    "\n",
    "plt.figure(figsize = (12, 6))\n",
    "sns.boxplot(B_nb_df, x = 'Neighboring Cell Type', y = 'pct', hue = 'Collection', showfliers = False)\n",
    "sns.swarmplot(B_nb_df, x = 'Neighboring Cell Type', y = 'pct', hue = 'Collection', dodge = True, color = 'k', alpha = .5, size = 3)\n",
    "plt.xticks(rotation = 45)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.title('Nearest neighboring cell type composition of B cells')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cytoT_nb_df = all_nb_df[all_nb_df['Cell Type'] == 'Cytotoxic T']\n",
    "metadata = ird_cell_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "cytoT_nb_df['Collection'] = cytoT_nb_df['Sample'].map(metadata['Collection'])\n",
    "\n",
    "plt.figure(figsize = (12, 6))\n",
    "sns.boxplot(cytoT_nb_df, x = 'Neighboring Cell Type', y = 'pct', hue = 'Collection', showfliers = False)\n",
    "sns.swarmplot(cytoT_nb_df, x = 'Neighboring Cell Type', y = 'pct', hue = 'Collection', dodge = True, color = 'k', alpha = .5, size = 3)\n",
    "plt.xticks(rotation = 45)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.title('Nearest neighboring cell type composition of Cytotoxic T cells')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exhT_nb_df = all_nb_df[all_nb_df['Cell Type'] == 'Exhausted T']\n",
    "metadata = ird_cell_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "exhT_nb_df['Collection'] = exhT_nb_df['Sample'].map(metadata['Collection'])\n",
    "\n",
    "plt.figure(figsize = (12, 6))\n",
    "sns.boxplot(exhT_nb_df, x = 'Neighboring Cell Type', y = 'pct', hue = 'Collection', showfliers = False)\n",
    "sns.swarmplot(exhT_nb_df, x = 'Neighboring Cell Type', y = 'pct', hue = 'Collection', dodge = True, color = 'k', alpha = .5, size = 3)\n",
    "plt.xticks(rotation = 45)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.title('Nearest neighboring cell type composition of Exhausted T cells')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 6: Distance to reference cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_xenium_merge = sc.read_h5ad(\"/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/Xenium/analysis/merged_filtered.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_cell_info = ird_xenium_merge.obs.copy()\n",
    "ird_cell_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = ird_cell_info[ird_cell_info['Sample'] == 'IRD_S18-30740A1U3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test['annot'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial import KDTree\n",
    "bcell_info = test[test['annot'] == 'B']\n",
    "bcell_tree = KDTree(bcell_info[['x_centroid', 'y_centroid']].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "myeloid_info = test[test['annot'] == 'Granulo.']\n",
    "bcell_tree.query(myeloid_info[['x_centroid', 'y_centroid']].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial import KDTree\n",
    "def nearest_dist_to_ref_celltype(df, ct_col_name, ref_ct, x_col='x_centroid', y_col='y_centroid'):\n",
    "    \"\"\"\n",
    "    This function finds the nearest distance for all cells in df to the nearest cell of a reference cell type.\n",
    "\n",
    "    Args:\n",
    "        df (pd.DataFrame): DataFrame containing cell coordinates and cell type annotation.\n",
    "        ct_col_name (str): Name of the column in df that contains cell type annotations.\n",
    "        ref_ct (str): The reference cell type to which distances will be calculated.\n",
    "        x_col (str): Name of the column for x-coordinates. Default is 'x_centroid'.\n",
    "        y_col (str): Name of the column for y-coordinates. Default is 'y_centroid'.\n",
    "    Returns:\n",
    "        res_df (pd.DataFrame): DataFrame with original df columns plus a new column 'nearest_dist_to_{ref_ct}'.\n",
    "    \"\"\"\n",
    "    # Check inputs\n",
    "\n",
    "    # Extract coordinates of reference cell type to build kdtree\n",
    "    ref_cells = df[df[ct_col_name] == ref_ct]\n",
    "    if ref_cells.empty:\n",
    "        raise ValueError(f\"No cells found for reference cell type '{ref_ct}' in column '{ct_col_name}'.\")\n",
    "    ref_tree = KDTree(ref_cells[[x_col, y_col]].values)\n",
    "\n",
    "    # Query kdtree for all cells to find nearest reference cell type\n",
    "    all_cells_coords = df[[x_col, y_col]].values\n",
    "    dists, _ = ref_tree.query(all_cells_coords)\n",
    "\n",
    "    # Create result DataFrame\n",
    "    res_df = df.copy()\n",
    "    res_df[f'nearest_dist_to_{ref_ct}'] = dists\n",
    "\n",
    "    return res_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test1 = nearest_dist_to_ref_celltype(test, 'annot', 'B', x_col='x_centroid', y_col='y_centroid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import gaussian_kde\n",
    "kde1 = gaussian_kde(test1[test1['annot']=='Granulo.']['nearest_dist_to_B'])\n",
    "dist_eval = np.arange(0, 105, 1)\n",
    "kde1_pdf = kde1.pdf(dist_eval)\n",
    "plt.plot(dist_eval, kde1_pdf)\n",
    "print(dist_eval[np.argmax(kde1_pdf)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct_to_plot = test1['annot'].unique().tolist()\n",
    "ct_to_plot.remove('B')\n",
    "\n",
    "nsubplots = len(ct_to_plot)\n",
    "ncols = 3\n",
    "nrows = nsubplots // ncols + 1\n",
    "fig, axes = plt.subplots(nrows=nrows, ncols=ncols, figsize=(3*ncols, 3*nrows))\n",
    "\n",
    "for i in range(len(ct_to_plot)):\n",
    "    sns.histplot(test1[test1['annot']==ct_to_plot[i]], x = 'nearest_dist_to_B', binwidth = 5, stat = 'density', ax = axes[i//ncols, i%ncols])\n",
    "    axes[i//ncols, i%ncols].set_title(ct_to_plot[i])\n",
    "    axes[i//ncols, i%ncols].set_xlim(0, 50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist, bin_edges = np.histogram(test1[test1['annot']=='Granulo.']['nearest_dist_to_B'], bins=range(0, 105, 5), density=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bin_edges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Most likely distance to a reference cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ird_cell_info = ird_xenium_merge.obs.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import gaussian_kde\n",
    "\n",
    "ref_ct = 'B'\n",
    "ct_col = 'annot'\n",
    "dist_eval = np.arange(0, 201, 1)\n",
    "all_other_ct = ird_cell_info[ct_col].unique().tolist()\n",
    "all_other_ct.remove(ref_ct)\n",
    "\n",
    "dist_to_B_df = pd.DataFrame(columns = ['Sample', 'Cell Type', 'Most Likely Distance'])\n",
    "for i in range(len(ird_cell_info['Sample'].unique().tolist())):\n",
    "    sample = ird_cell_info['Sample'].unique().tolist()[i]\n",
    "    sample_df = ird_cell_info[ird_cell_info['Sample'] == sample]\n",
    "    sample_df1 = nearest_dist_to_ref_celltype(sample_df, ct_col, ref_ct, x_col='x_centroid', y_col='y_centroid')\n",
    "    \n",
    "    dist = []\n",
    "    for ct in all_other_ct:\n",
    "        dist_distribution = sample_df1[sample_df1[ct_col]==ct][f'nearest_dist_to_{ref_ct}']\n",
    "        if len(dist_distribution) < 10:\n",
    "            dist.append(np.nan)\n",
    "            continue\n",
    "        else:\n",
    "            kde_ct = gaussian_kde(dist_distribution)\n",
    "            ct_kde_pdf = kde_ct.pdf(dist_eval)\n",
    "            ct_dist = dist_eval[np.argmax(ct_kde_pdf)]\n",
    "            dist.append(ct_dist)\n",
    "\n",
    "    dist_to_B_ct_df = pd.DataFrame({\n",
    "        'Sample': sample,\n",
    "        'Cell Type': all_other_ct,\n",
    "        'Most Likely Distance': dist\n",
    "    })\n",
    "    dist_to_B_df = pd.concat([dist_to_B_df, dist_to_B_ct_df], axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dist_to_B_df = dist_to_B_df.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = ird_cell_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "dist_to_B_df['Collection'] = dist_to_B_df['Sample'].map(metadata['Collection'])\n",
    "plt.figure(figsize = (12, 6))\n",
    "sns.boxplot(dist_to_B_df, x = 'Cell Type', y = 'Most Likely Distance', hue = 'Collection', showfliers = False)\n",
    "sns.swarmplot(dist_to_B_df, x = 'Cell Type', y = 'Most Likely Distance', hue = 'Collection', dodge = True, color = 'k', alpha = .7)\n",
    "plt.title('Most likely distance to B cells by cell type')\n",
    "plt.xticks(rotation = 45)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "#plt.savefig(\"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_to_Bcells_by_celltype_boxplot_v09192025.pdf\", dpi = 300, transparent = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Statistical test within each cell type, compare distance to B cells between NBM/NDMM, NBM/PT, NDMM/PT\n",
    "from scipy.stats import mannwhitneyu\n",
    "from scipy.stats import false_discovery_control\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "stat_results = pd.DataFrame(columns = ['Cell Type', 'Group1', 'Group2', 'U_statistic', 'p_value'])\n",
    "for ct in dist_to_B_df['Cell Type'].unique():\n",
    "    ct_df = dist_to_B_df[dist_to_B_df['Cell Type'] == ct]\n",
    "    nbm_dist = ct_df[ct_df['Collection'] == 'NBM']['Most Likely Distance'].values.tolist()\n",
    "    ndmm_dist = ct_df[ct_df['Collection'] == 'NDMM']['Most Likely Distance'].values.tolist()\n",
    "    pt_dist = ct_df[ct_df['Collection'] == 'PT']['Most Likely Distance'].values.tolist()\n",
    "    \n",
    "    if len(nbm_dist) > 0 and len(ndmm_dist) > 0:\n",
    "        u_stat, p_val = mannwhitneyu(nbm_dist, ndmm_dist, alternative='two-sided')\n",
    "        stat_results = pd.concat([stat_results, pd.DataFrame({\n",
    "            'Cell Type': [ct],\n",
    "            'Group1': ['NBM'],\n",
    "            'Group2': ['NDMM'],\n",
    "            'U_statistic': [u_stat],\n",
    "            'p_value': [p_val]\n",
    "        })], axis = 0)\n",
    "    \n",
    "    if len(nbm_dist) > 0 and len(pt_dist) > 0:\n",
    "        u_stat, p_val = mannwhitneyu(nbm_dist, pt_dist, alternative='two-sided')\n",
    "        stat_results = pd.concat([stat_results, pd.DataFrame({\n",
    "            'Cell Type': [ct],\n",
    "            'Group1': ['NBM'],\n",
    "            'Group2': ['PT'],\n",
    "            'U_statistic': [u_stat],\n",
    "            'p_value': [p_val]\n",
    "        })], axis = 0)\n",
    "    \n",
    "    if len(ndmm_dist) > 0 and len(pt_dist) > 0:\n",
    "        u_stat, p_val = mannwhitneyu(ndmm_dist, pt_dist, alternative='two-sided')\n",
    "        stat_results = pd.concat([stat_results, pd.DataFrame({\n",
    "            'Cell Type': [ct],\n",
    "            'Group1': ['NDMM'],\n",
    "            'Group2': ['PT'],\n",
    "            'U_statistic': [u_stat],\n",
    "            'p_value': [p_val]\n",
    "        })], axis = 0)\n",
    "\n",
    "stat_results['p_adj'] = multipletests(stat_results['p_value'], method='fdr_bh')[1]\n",
    "stat_results.to_csv(\"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_to_Bcells_wilcoxtest_v10242025.csv\")\n",
    "stat_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_ct = 'PC'\n",
    "dist_eval = np.arange(0, 201, 1)\n",
    "all_other_ct = ird_cell_info['annot'].unique().tolist()\n",
    "all_other_ct.remove(ref_ct)\n",
    "\n",
    "dist_to_PC_df = pd.DataFrame(columns = ['Sample', 'Cell Type', 'Most Likely Distance'])\n",
    "for i in range(len(ird_cell_info['Sample'].unique().tolist())):\n",
    "    sample = ird_cell_info['Sample'].unique().tolist()[i]\n",
    "    sample_df = ird_cell_info[ird_cell_info['Sample'] == sample]\n",
    "    sample_df1 = nearest_dist_to_ref_celltype(sample_df, 'annot', ref_ct, x_col='x_centroid', y_col='y_centroid')\n",
    "    \n",
    "    dist = []\n",
    "    for ct in all_other_ct:\n",
    "        kde_ct = gaussian_kde(sample_df1[sample_df1['annot']==ct][f'nearest_dist_to_{ref_ct}'])\n",
    "        ct_kde_pdf = kde_ct.pdf(dist_eval)\n",
    "        ct_dist = dist_eval[np.argmax(ct_kde_pdf)]\n",
    "        dist.append(ct_dist)\n",
    "\n",
    "    dist_to_PC_ct_df = pd.DataFrame({\n",
    "        'Sample': sample,\n",
    "        'Cell Type': all_other_ct,\n",
    "        'Most Likely Distance': dist\n",
    "    })\n",
    "    dist_to_PC_df = pd.concat([dist_to_PC_df, dist_to_PC_ct_df], axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dist_to_PC_df[(dist_to_PC_df['Cell Type'] == 'Erythro.') & (dist_to_PC_df['Collection'] == \"PT\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dist_to_PC_df['Collection'] = dist_to_PC_df['Sample'].map(metadata['Collection'])\n",
    "plt.figure(figsize = (12, 6))\n",
    "sns.boxplot(dist_to_PC_df, x = 'Cell Type', y = 'Most Likely Distance', hue = 'Collection', showfliers = False)\n",
    "sns.swarmplot(dist_to_PC_df, x = 'Cell Type', y = 'Most Likely Distance', hue = 'Collection', dodge = True, color = 'k', alpha = .5)\n",
    "plt.title('Most likely distance to Plasma cells by cell type')\n",
    "plt.xticks(rotation = 45)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "#plt.savefig(\"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_to_PC_by_celltype_boxplot_v09192025.pdf\", dpi = 300, transparent = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Statistical test within each cell type, compare distance to B cells between NBM/NDMM, NBM/PT, NDMM/PT\n",
    "stat_results = pd.DataFrame(columns = ['Cell Type', 'Group1', 'Group2', 'U_statistic', 'p_value'])\n",
    "for ct in dist_to_PC_df['Cell Type'].unique():\n",
    "    ct_df = dist_to_PC_df[dist_to_PC_df['Cell Type'] == ct]\n",
    "    nbm_dist = ct_df[ct_df['Collection'] == 'NBM']['Most Likely Distance'].values.tolist()\n",
    "    ndmm_dist = ct_df[ct_df['Collection'] == 'NDMM']['Most Likely Distance'].values.tolist()\n",
    "    pt_dist = ct_df[ct_df['Collection'] == 'PT']['Most Likely Distance'].values.tolist()\n",
    "    \n",
    "    if len(nbm_dist) > 0 and len(ndmm_dist) > 0:\n",
    "        u_stat, p_val = mannwhitneyu(nbm_dist, ndmm_dist, alternative='two-sided')\n",
    "        stat_results = pd.concat([stat_results, pd.DataFrame({\n",
    "            'Cell Type': [ct],\n",
    "            'Group1': ['NBM'],\n",
    "            'Group2': ['NDMM'],\n",
    "            'U_statistic': [u_stat],\n",
    "            'p_value': [p_val]\n",
    "        })], axis = 0)\n",
    "    \n",
    "    if len(nbm_dist) > 0 and len(pt_dist) > 0:\n",
    "        u_stat, p_val = mannwhitneyu(nbm_dist, pt_dist, alternative='two-sided')\n",
    "        stat_results = pd.concat([stat_results, pd.DataFrame({\n",
    "            'Cell Type': [ct],\n",
    "            'Group1': ['NBM'],\n",
    "            'Group2': ['PT'],\n",
    "            'U_statistic': [u_stat],\n",
    "            'p_value': [p_val]\n",
    "        })], axis = 0)\n",
    "    \n",
    "    if len(ndmm_dist) > 0 and len(pt_dist) > 0:\n",
    "        u_stat, p_val = mannwhitneyu(ndmm_dist, pt_dist, alternative='two-sided')\n",
    "        stat_results = pd.concat([stat_results, pd.DataFrame({\n",
    "            'Cell Type': [ct],\n",
    "            'Group1': ['NDMM'],\n",
    "            'Group2': ['PT'],\n",
    "            'U_statistic': [u_stat],\n",
    "            'p_value': [p_val]\n",
    "        })], axis = 0)\n",
    "\n",
    "stat_results['p_adj'] = multipletests(stat_results['p_value'], method='fdr_bh')[1]\n",
    "stat_results.to_csv(\"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD_JW_Xenium_merge_nearestDist_to_PlasmaCells_wilcoxtest_v10242025.csv\")\n",
    "stat_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import gaussian_kde\n",
    "def nearest_dist_to_ref_celltype_allsamples(df, ref_ct, ct_col, sample_col, dist_eval, x_col='x_centroid', y_col='y_centroid'):\n",
    "    ''' \n",
    "    This function calculates the most likely distance to a reference cell type for all other cell types, across all samples.\n",
    "    Args:\n",
    "        df (pd.DataFrame): DataFrame containing cell coordinates, cell type annotation, and sample information.\n",
    "        ref_ct (str): The reference cell type to which distances will be calculated.\n",
    "        ct_col (str): Name of the column in df that contains cell type annotations.\n",
    "        sample_col (str): Name of the column in df that contains sample identifiers.\n",
    "        dist_eval (np.array): Array of distance values over which to evaluate the KDE.\n",
    "        x_col (str): Name of the column for x-coordinates. Default is 'x_centroid'.\n",
    "        y_col (str): Name of the column for y-coordinates. Default is 'y_centroid'.\n",
    "    Returns:\n",
    "        dist_to_ref_df (pd.DataFrame): DataFrame with columns ['Sample', 'Cell Type', 'Most Likely Distance'].\n",
    "    '''\n",
    "    # Initialize empty dataframe to store results\n",
    "    dist_to_ref_df = pd.DataFrame(columns = ['Sample', 'Cell Type', 'Most Likely Distance'])\n",
    "\n",
    "    # Identify cell types to evaluate, excluding the reference cell type\n",
    "    all_other_ct = df[ct_col].unique().tolist()\n",
    "    all_other_ct.remove(ref_ct)\n",
    "\n",
    "    # Loop through all samples\n",
    "    nSamples = len(df[sample_col].unique().tolist())\n",
    "    for i in range(nSamples):\n",
    "        sample = df[sample_col].unique().tolist()[i]\n",
    "        sample_df = df[df[sample_col] == sample]\n",
    "        if sample_df[sample_df[ct_col] == ref_ct].empty:\n",
    "            print(f\"Skipping sample {sample} as it has no cells of reference cell type '{ref_ct}'\")\n",
    "            continue\n",
    "        sample_df1 = nearest_dist_to_ref_celltype(sample_df, ct_col, ref_ct, x_col='x_centroid', y_col='y_centroid')\n",
    "        \n",
    "        dist = []\n",
    "        for ct in all_other_ct:\n",
    "            dist_distribution = sample_df1[sample_df1[ct_col]==ct][f'nearest_dist_to_{ref_ct}']\n",
    "            if len(dist_distribution) < 10: # Require at least 10 data points to estimate KDE\n",
    "                dist.append(np.nan)\n",
    "                continue\n",
    "            else:\n",
    "                kde_ct = gaussian_kde(dist_distribution)\n",
    "                ct_kde_pdf = kde_ct.pdf(dist_eval)\n",
    "                ct_dist = dist_eval[np.argmax(ct_kde_pdf)]\n",
    "                dist.append(ct_dist)\n",
    "\n",
    "        dist_to_ref_ct_df = pd.DataFrame({\n",
    "            'Sample': sample,\n",
    "            'Cell Type': all_other_ct,\n",
    "            'Most Likely Distance': dist\n",
    "        })\n",
    "        dist_to_ref_df = pd.concat([dist_to_ref_df, dist_to_ref_ct_df], axis = 0)\n",
    "    return dist_to_ref_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fraction of cell types within a distance bin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_ct = 'B'\n",
    "all_other_ct = ird_cell_info['annot'].unique().tolist()\n",
    "all_other_ct.remove(ref_ct)\n",
    "\n",
    "dist_to_B_df = pd.DataFrame(columns = ['Sample', 'Cell Type', 'Distance Bin Start', 'Distance Bin End', 'Fraction'])\n",
    "for i in range(len(ird_cell_info['Sample'].unique().tolist())):\n",
    "    sample = ird_cell_info['Sample'].unique().tolist()[i]\n",
    "    sample_df = ird_cell_info[ird_cell_info['Sample'] == sample]\n",
    "    sample_df1 = nearest_dist_to_ref_celltype(sample_df, 'annot', ref_ct, x_col='x_centroid', y_col='y_centroid')\n",
    "    hist_all, bin_edges = np.histogram(sample_df1[sample_df1['annot']!=ref_ct][f'nearest_dist_to_{ref_ct}'], bins=range(0, 105, 5), density=False)  ## number of cells in each distance bin\n",
    "    \n",
    "    for ct in all_other_ct:\n",
    "        hist, _ = np.histogram(sample_df1[sample_df1['annot']==ct][f'nearest_dist_to_{ref_ct}'], bins=range(0, 105, 5), density=False)  ## number of ct in each distance bin\n",
    "        frac = hist / hist_all  ## fraction of ct in each distance bin\n",
    "        dist_to_B_ct_df = pd.DataFrame({\n",
    "            'Sample': sample,\n",
    "            'Cell Type': ct,\n",
    "            'Distance Bin Start': bin_edges[:-1],\n",
    "            'Distance Bin End': bin_edges[1:],\n",
    "            'Fraction': frac\n",
    "        })\n",
    "        dist_to_B_df = pd.concat([dist_to_B_df, dist_to_B_ct_df], axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = ird_cell_info[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dist_to_B_df['Collection'] = dist_to_B_df['Sample'].map(metadata['Collection'])\n",
    "dist_to_B_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dist_to_B_df['Distance Bin Mid'] = (dist_to_B_df['Distance Bin Start'] + dist_to_B_df['Distance Bin End']) / 2\n",
    "for ct in dist_to_B_df['Cell Type'].unique():\n",
    "    df_plot = dist_to_B_df[dist_to_B_df['Cell Type'] == ct]\n",
    "    plt.figure(figsize = (6, 4))\n",
    "    sns.boxplot(df_plot, x = 'Distance Bin Mid', y = 'Fraction', hue = 'Collection')\n",
    "    plt.title(ct)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Section 7: BANKSY Spatial Neighborhood Analysis\n",
    "\n",
    "Identify and characterize spatially-defined neighborhoods using [BANKSY](https://www.nature.com/articles/s41588-024-01664-3). \n",
    "\n",
    "The BANKSY pipeline is run outside of this notebook. The goals of this section is to perform downstream analysis of the BANKSY results.\n",
    "\n",
    "### Analysis Overview\n",
    "\n",
    "This section performs:\n",
    "1. **Cluster filtering**: Remove small or sample-specific neighborhoods\n",
    "2. **Cell type enrichment**: Identify which cell types are enriched in each neighborhood\n",
    "3. **Neighborhood abundance**: Analyze neighborhood proportions across disease stages\n",
    "4. **Cellular composition**: Characterize cell type composition within neighborhoods across conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy = sc.read_h5ad('/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/Xenium/analysis/banksy/Output/Run10242025_final_banksy_matrix.h5ad')\n",
    "new_annot = pd.read_csv(\"/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/Xenium/analysis/merged_metadata.csv\", index_col = 0)\n",
    "banksy.obs['revised_annot'] = banksy.obs.index.map(new_annot['ct'])\n",
    "banksy.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy.obs['leiden'].value_counts().tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter out clusters that are small or specific to one sample/condition\n",
    "\n",
    "Remove neighborhoods that may represent technical artifacts or sample-specific biology rather than generalizable spatial patterns. Filtering criteria:\n",
    "1. **Size filter**: Remove neighborhoods with < 3000 cells (likely noise or rare spatial artifacts)\n",
    "2. **Sample specificity**: Remove neighborhoods predominantly found in a single sample (sample-specific technical/biological artifacts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First try: filter out clusters with fewer than 3000 cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy_cluster_size = banksy.obs['leiden'].value_counts().tolist()\n",
    "banksy_clusters_to_keep = [i for i, size in enumerate(banksy_cluster_size) if size >= 3000]\n",
    "banksy_clusters_to_keep = [str(c) for c in banksy_clusters_to_keep]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy_clusters_to_keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy1 = banksy[banksy.obs['leiden'].isin(banksy_clusters_to_keep), :]\n",
    "banksy1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot proportion of cells in each bansky cluster per timepoint (Collection)\n",
    "banksy_meta = banksy1.obs.copy()\n",
    "banksy_tp_count = banksy_meta.groupby('Collection').size().reset_index(name='total_cells')\n",
    "banksy_cluster_count = banksy_meta.groupby(['Collection', 'leiden']).size().reset_index(name='cluster_cells')\n",
    "banksy_cluster_prop = banksy_cluster_count.merge(banksy_tp_count, on='Collection')\n",
    "banksy_cluster_prop['proportion'] = banksy_cluster_prop['cluster_cells'] / banksy_cluster_prop['total_cells']\n",
    "plt.figure(figsize=(8,6))\n",
    "sns.barplot(banksy_cluster_prop, x='leiden', y='proportion', hue='Collection')\n",
    "plt.title('Proportion of cells in each Banksy cluster per timepoint')\n",
    "plt.xlabel('Banksy Cluster')\n",
    "plt.ylabel('Proportion of cells')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cluster 8 and 12 seem specific to NDMM - check if they are sample specific"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot percent of cells in banksy cluster 8 per sample\n",
    "banksy_cluster8 = banksy1[banksy1.obs['leiden'] == '8', :]\n",
    "banksy8_meta = banksy_cluster8.obs[['Sample']].copy()\n",
    "banksy8_sample_count = banksy8_meta.groupby('Sample').size().reset_index(name='cluster8_cells')\n",
    "banksy_sample_count = banksy1.obs.groupby('Sample').size().reset_index(name='total_cells')\n",
    "banksy8_sample_prop = banksy8_sample_count.merge(banksy_sample_count, on='Sample')\n",
    "banksy8_sample_prop['proportion'] = banksy8_sample_prop['cluster8_cells'] / banksy8_sample_prop['total_cells']\n",
    "print(\"Over-represented sample in BANKSY cluster 8: \", banksy8_sample_prop.loc[banksy8_sample_prop['proportion']>0.1, 'Sample'].values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot percent of cells in banksy cluster 12 per sample\n",
    "banksy_cluster12 = banksy1[banksy1.obs['leiden'] == '12', :]\n",
    "banksy12_meta = banksy_cluster12.obs[['Sample']].copy()\n",
    "banksy12_sample_count = banksy12_meta.groupby('Sample').size().reset_index(name='cluster12_cells')\n",
    "banksy_sample_count = banksy1.obs.groupby('Sample').size().reset_index(name='total_cells')\n",
    "banksy12_sample_prop = banksy12_sample_count.merge(banksy_sample_count, on='Sample')\n",
    "banksy12_sample_prop['proportion'] = banksy12_sample_prop['cluster12_cells'] / banksy12_sample_prop['total_cells']\n",
    "print(\"Over-represented sample in BANKSY cluster 12: \", banksy12_sample_prop.loc[banksy12_sample_prop['proportion']>0.05, 'Sample'].values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.dotplot(banksy1, var_names=['SDC1', 'SLAMF7', 'TNFRSF17', 'MZB1'], groupby = 'leiden')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nbhd8 and Nbhd12 are both specific to plasma cells in S19-25371 - could be interesting down the road. Discard for now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy_clusters_to_keep = [banksy_cluster for banksy_cluster in banksy_clusters_to_keep if banksy_cluster not in ['8', '12']]\n",
    "banksy_clusters_to_keep\n",
    "banksy1 = banksy[banksy.obs['leiden'].isin(banksy_clusters_to_keep), :]\n",
    "banksy1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cell type enrichment analysis\n",
    "\n",
    "Quantify which cell types are significantly enriched or depleted in each BANKSY neighborhood using statistical tests:\n",
    "\n",
    "**Statistical approach:**\n",
    "1. **Fisher's exact test**: Calculate odds ratios for each cell type in each neighborhood\n",
    "2. **Hypergeometric test**: Calculate p-values for statistical significance of enrichment\n",
    "3. **FDR correction**: Adjust p-values for multiple testing using Benjamini-Hochberg method\n",
    "\n",
    "This analysis reveals the cellular composition defining each spatial neighborhood and identifies which cell types preferentially co-localize."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy1.obs['revised_annot'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_celltype_enrichment(adata, neighborhood_col, celltype_col):\n",
    "    \"\"\"\n",
    "    Calculate cell type enrichment in spatial neighborhoods using Fisher's exact test.\n",
    "    \n",
    "    For each neighborhood-celltype pair, computes odds ratios and p-values to quantify\n",
    "    whether a cell type is significantly enriched or depleted in that neighborhood\n",
    "    compared to the overall population.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    adata : AnnData\n",
    "        Annotated data object with neighborhood and cell type annotations\n",
    "    neighborhood_col : str\n",
    "        Column name in adata.obs containing neighborhood labels (e.g., 'leiden', 'nbhd_annot')\n",
    "    celltype_col : str\n",
    "        Column name in adata.obs containing cell type annotations\n",
    "        \n",
    "    Returns:\n",
    "    --------\n",
    "    tuple: (pd.DataFrame, pd.DataFrame)\n",
    "        - odds_matrix: DataFrame with odds ratios (rows=neighborhoods, cols=cell types)\n",
    "        - pval_matrix: DataFrame with hypergeometric p-values (same shape as odds_matrix)\n",
    "        \n",
    "    Statistical Tests:\n",
    "    ------------------\n",
    "    - **Fisher's exact test**: Computes odds ratio for enrichment\n",
    "    - **Hypergeometric test**: Computes p-value for significance (upper tail test)\n",
    "    \n",
    "    Example:\n",
    "    --------\n",
    "    odds_matrix, pval_matrix = calculate_celltype_enrichment(\n",
    "        banksy_filtered, \n",
    "        neighborhood_col='nbhd_annot',\n",
    "        celltype_col='revised_annot'\n",
    "    )\n",
    "    \n",
    "    # Visualize results\n",
    "    log2_odds = np.log2(odds_matrix)\n",
    "    sns.heatmap(log2_odds, center=0, cmap='Spectral_r')\n",
    "    \"\"\"\n",
    "    from scipy.stats import fisher_exact, hypergeom\n",
    "    \n",
    "    neighborhoods = adata.obs[neighborhood_col].unique()\n",
    "    cell_types = adata.obs[celltype_col].unique()\n",
    "    df = adata.obs[[neighborhood_col, celltype_col]].copy()\n",
    "    \n",
    "    # Initialize output matrices\n",
    "    odds_matrix = pd.DataFrame(\n",
    "        np.nan,\n",
    "        index=neighborhoods.categories if hasattr(neighborhoods, 'categories') else neighborhoods,\n",
    "        columns=cell_types.categories if hasattr(cell_types, 'categories') else cell_types\n",
    "    )\n",
    "    \n",
    "    pval_matrix = pd.DataFrame(\n",
    "        np.nan,\n",
    "        index=neighborhoods.categories if hasattr(neighborhoods, 'categories') else neighborhoods,\n",
    "        columns=cell_types.categories if hasattr(cell_types, 'categories') else cell_types\n",
    "    )\n",
    "    \n",
    "    # Calculate enrichment for each neighborhood-celltype pair\n",
    "    for nb in odds_matrix.index:\n",
    "        for ct in odds_matrix.columns:\n",
    "            # Build 2x2 contingency table\n",
    "            # a = in nb and of type ct\n",
    "            a = ((df[neighborhood_col] == nb) & (df[celltype_col] == ct)).sum()\n",
    "            # b = in nb and not type ct\n",
    "            b = ((df[neighborhood_col] == nb) & (df[celltype_col] != ct)).sum()\n",
    "            # c = not in nb and of type ct\n",
    "            c = ((df[neighborhood_col] != nb) & (df[celltype_col] == ct)).sum()\n",
    "            # d = not in nb and not type ct\n",
    "            d = ((df[neighborhood_col] != nb) & (df[celltype_col] != ct)).sum()\n",
    "            \n",
    "            tbl = np.array([[a, b], [c, d]])\n",
    "            \n",
    "            # Fisher's exact test for odds ratio\n",
    "            oddsratio, _ = fisher_exact(tbl)\n",
    "            odds_matrix.loc[nb, ct] = oddsratio\n",
    "            \n",
    "            # Hypergeometric test (upper tail) for p-value\n",
    "            m = a + c  # total ct cells\n",
    "            n = b + d  # total non-ct cells\n",
    "            k = a + b  # total cells in nb\n",
    "            q = a - 1 if a > 0 else 0\n",
    "            \n",
    "            pval_matrix.loc[nb, ct] = hypergeom.sf(q, m + n, m, k)\n",
    "    \n",
    "    return odds_matrix, pval_matrix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy1.obs['revised_annot'] = pd.Categorical(banksy1.obs['revised_annot'], \n",
    "                                        categories=['HSPC', 'Erythroid', 'Megakaryocyte',\n",
    "                                        'GMP', 'Late Myeloid', 'Neutrophil', 'Ba/Eo/Ma',\n",
    "                                        'Monocyte', 'Macrophage','pDC', 'cDC',\n",
    "                                        'Early B', 'Mature B', 'CD4 T', 'CD8 T', 'NK', 'PC', \n",
    "                                        'Adipocyte', 'MSC', 'Endothelial', 'Fibro/Osteo', 'vSMC/Pericyte', 'Low Confidence'],\n",
    "                                        ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "odds_matrix, pval_matrix = calculate_celltype_enrichment(\n",
    "        banksy1, \n",
    "        neighborhood_col='leiden',\n",
    "        celltype_col='revised_annot'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Draw heatmap to show enriched cell types in each BANKSY neighborhood\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "# Log2 transform odds ratios\n",
    "log2_odds_matrix = np.log2(odds_matrix)\n",
    "\n",
    "# Replace negative infinities with the minimum finite value\n",
    "finite_values = log2_odds_matrix.values[np.isfinite(log2_odds_matrix.values)]\n",
    "max_neg_value = finite_values.min()\n",
    "log2_odds_matrix = log2_odds_matrix.replace(-np.inf, max_neg_value)\n",
    "\n",
    "# Adjust p-values using FDR correction\n",
    "pval_vector = pval_matrix.values.flatten()\n",
    "padj_vector = multipletests(pval_vector, method='fdr_bh')[1]\n",
    "padj_matrix = pd.DataFrame(\n",
    "    padj_vector.reshape(pval_matrix.shape),\n",
    "    index=pval_matrix.index,\n",
    "    columns=pval_matrix.columns\n",
    ")\n",
    "\n",
    "# Create star annotation matrix\n",
    "star_matrix = padj_matrix.map(lambda x: \"*\" if x < 0.05 else \"\")\n",
    "\n",
    "## Define color palette (PiYG-like: purple  white  green)\n",
    "#colors = ['#8e0152', '#f7f7f7', '#276419']  # Similar to RColorBrewer PiYG\n",
    "#cmap = LinearSegmentedColormap.from_list('custom_PiYG', colors)\n",
    "\n",
    "# Create figure and heatmap\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "\n",
    "# Draw heatmap\n",
    "sns.heatmap(\n",
    "    log2_odds_matrix,\n",
    "    cmap='Spectral_r',\n",
    "    center=0,\n",
    "    vmin=-3,\n",
    "    vmax=3,\n",
    "    cbar_kws={'label': 'log2(Odds Ratio)'},\n",
    "    linewidths=0.5,\n",
    "    linecolor='lightgray',\n",
    "    ax=ax\n",
    ")\n",
    "\n",
    "# Add star annotations\n",
    "for i in range(len(log2_odds_matrix.index)):\n",
    "    for j in range(len(log2_odds_matrix.columns)):\n",
    "        if star_matrix.iloc[i, j] == \"*\":\n",
    "            ax.text(j + 0.5, i + 0.5, '*',\n",
    "                   ha='center', va='center',\n",
    "                   fontsize=10, color='black')\n",
    "\n",
    "# Formatting\n",
    "ax.set_xlabel('Cell Type')\n",
    "ax.set_ylabel('BANKSY Neighborhood Clusters')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()\n",
    "#plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/IRD_JW_Xenium_BANKSY_neighborhood_celltype_oddsratio_heatmap_v11062025.pdf', dpi=300, transparent = True)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot percent of cells in banksy cluster 10 per sample\n",
    "banksy_cluster10 = banksy1[banksy1.obs['leiden'] == '10', :]\n",
    "banksy10_meta = banksy_cluster10.obs[['Sample']].copy()\n",
    "banksy10_sample_count = banksy10_meta.groupby('Sample').size().reset_index(name='cluster10_cells')\n",
    "banksy_sample_count = banksy1.obs.groupby('Sample').size().reset_index(name='total_cells')\n",
    "banksy10_sample_prop = banksy10_sample_count.merge(banksy_sample_count, on='Sample')\n",
    "banksy10_sample_prop['proportion'] = banksy10_sample_prop['cluster10_cells'] / banksy10_sample_prop['total_cells']\n",
    "print(\"Over-represented sample in BANKSY cluster 10: \", banksy10_sample_prop.loc[banksy10_sample_prop['proportion']>0.05, 'Sample'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_df = banksy1.obs[banksy1.obs['Sample'] == 'S16-30080A1U1']\n",
    "sns.scatterplot(sample_df, x = 'x_centroid', y = 'y_centroid', hue = 'leiden', s = 1, alpha = 0.8, edgecolor = None)\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "plt.legend(title='BANKSY Neighborhood', bbox_to_anchor=(1.05, 1), loc='upper left', markerscale = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(sample_df[sample_df['leiden'].isin(['5', '10'])], x = 'x_centroid', y = 'y_centroid', hue = 'leiden', s = 1, alpha = 0.8, edgecolor = None)\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "plt.legend(title='BANKSY Neighborhood', bbox_to_anchor=(1.05, 1), loc='upper left', markerscale = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(banksy1, groupby = 'leiden', method = 'wilcoxon', groups = ['10'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.rank_genes_groups(banksy1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot scatterplot of cell centroids colored by BANKSY neighborhood, split by sample, save as multi-page PDF\n",
    "pdf_name = \"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/IRD_JW_Xenium_merge_BANKSY_filtered_scatterplot_110625.pdf\"\n",
    "with PdfPages(pdf_name) as pdf:\n",
    "    for sample in banksy1.obs['Sample'].unique():\n",
    "        sample_data = banksy1[banksy1.obs['Sample'] == sample, :]\n",
    "        plt.figure(figsize=(10, 10))\n",
    "        sns.scatterplot(\n",
    "            x=sample_data.obs['x_centroid'],\n",
    "            y=sample_data.obs['y_centroid'],\n",
    "            hue=sample_data.obs['leiden'],\n",
    "            palette='tab20',\n",
    "            s=2,\n",
    "            alpha=0.8,\n",
    "            edgecolor=None\n",
    "        )\n",
    "        plt.title(f'Cell Centroids Colored by BANKSY Neighborhood - Sample: {sample}')\n",
    "        plt.xlabel('X Centroid')\n",
    "        plt.ylabel('Y Centroid')\n",
    "        plt.gca().set_aspect('equal', adjustable='box')\n",
    "        plt.legend(title='BANKSY Neighborhood', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "        plt.tight_layout()\n",
    "        pdf.savefig()\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export BANKSY neighborhood annotations for each sample to Xenium browser format\n",
    "for sample in banksy1.obs['Sample'].unique():\n",
    "    sample_data = banksy1[banksy1.obs['Sample'] == sample, :]\n",
    "    output_df = sample_data.obs[['Original_Barcode', 'leiden']].copy()\n",
    "    output_df.columns = ['cell_id', 'group']\n",
    "    output_path = f\"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/Annotation/IRD_JW_Xenium_{sample}_BANKSY_neighborhood_annotation_110625.csv\"\n",
    "    output_df.to_csv(output_path, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotated and removed unknowns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Neighborhood 4 is enriched for unknown cells -> remove\n",
    "banksy_filtered = banksy1[banksy1.obs['leiden'].isin(['0', '1', '2', '3', '5', '6', '7', '9', '11']), :]\n",
    "banksy_nbhd_annotation = {\n",
    "    '0': 'Erythroid',\n",
    "    '1': 'Lymphoid',\n",
    "    '2': 'Late Myeloid',\n",
    "    '3': 'Endothelial',\n",
    "    '5': 'Plasma cell',\n",
    "    '6': 'Early Myeloid',\n",
    "    '7': 'Megakaryocyte',\n",
    "    '9': 'Endosteal',\n",
    "    '11': 'Arterial'\n",
    "}\n",
    "banksy_filtered.obs['nbhd_annot'] = banksy_filtered.obs['leiden'].map(banksy_nbhd_annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy_filtered.obs['nbhd_annot'] = pd.Categorical(banksy_filtered.obs['nbhd_annot'], \n",
    "                                        categories=['Erythroid', 'Early Myeloid', 'Late Myeloid',\n",
    "                                                    'Lymphoid', 'Plasma cell', 'Megakaryocyte',\n",
    "                                                    'Endothelial', 'Arterial', 'Endosteal'],\n",
    "                                        ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "odds_matrix, pval_matrix = calculate_celltype_enrichment(\n",
    "        banksy_filtered, \n",
    "        neighborhood_col='nbhd_annot',\n",
    "        celltype_col='revised_annot'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Draw heatmap to show enriched cell types in each BANKSY neighborhood\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "# Log2 transform odds ratios\n",
    "log2_odds_matrix = np.log2(odds_matrix)\n",
    "\n",
    "# Replace negative infinities with the minimum finite value\n",
    "finite_values = log2_odds_matrix.values[np.isfinite(log2_odds_matrix.values)]\n",
    "max_neg_value = finite_values.min()\n",
    "log2_odds_matrix = log2_odds_matrix.replace(-np.inf, max_neg_value)\n",
    "\n",
    "# Adjust p-values using FDR correction\n",
    "pval_vector = pval_matrix.values.flatten()\n",
    "padj_vector = multipletests(pval_vector, method='fdr_bh')[1]\n",
    "padj_matrix = pd.DataFrame(\n",
    "    padj_vector.reshape(pval_matrix.shape),\n",
    "    index=pval_matrix.index,\n",
    "    columns=pval_matrix.columns\n",
    ")\n",
    "\n",
    "# Create star annotation matrix\n",
    "star_matrix = padj_matrix.map(lambda x: \"*\" if x < 0.05 else \"\")\n",
    "\n",
    "## Define color palette (PiYG-like: purple  white  green)\n",
    "#colors = ['#8e0152', '#f7f7f7', '#276419']  # Similar to RColorBrewer PiYG\n",
    "#cmap = LinearSegmentedColormap.from_list('custom_PiYG', colors)\n",
    "\n",
    "# Create figure and heatmap\n",
    "fig, ax = plt.subplots(figsize=(8, 5))\n",
    "\n",
    "# Draw heatmap\n",
    "sns.heatmap(\n",
    "    log2_odds_matrix,\n",
    "    cmap='Spectral_r',\n",
    "    center=0,\n",
    "    vmin=-3,\n",
    "    vmax=3,\n",
    "    cbar_kws={'label': 'log2(Odds Ratio)'},\n",
    "    linewidths=0.5,\n",
    "    linecolor='lightgray',\n",
    "    ax=ax\n",
    ")\n",
    "\n",
    "# Add star annotations\n",
    "for i in range(len(log2_odds_matrix.index)):\n",
    "    for j in range(len(log2_odds_matrix.columns)):\n",
    "        if star_matrix.iloc[i, j] == \"*\":\n",
    "            ax.text(j + 0.5, i + 0.5, '*',\n",
    "                   ha='center', va='center',\n",
    "                   fontsize=10, color='black')\n",
    "\n",
    "# Formatting\n",
    "ax.set_xlabel('Cell Type')\n",
    "ax.set_ylabel('BANKSY Neighborhood Annotation')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.yticks(rotation=0)\n",
    "plt.tight_layout()\n",
    "#plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/IRD_JW_Xenium_BANKSY_annotated_celltype_oddsratio_heatmap_v11062025.pdf', dpi=300, transparent = True)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scatterplot of select samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "neighborhood_colors = {\n",
    "    \"Late Myeloid\":  \"#079450\",  # later granulo\n",
    "    \"Erythroid\":  \"#9e9e9e\", # erythroid\n",
    "    \"Plasma cell\":  \"#ff42ca\", # PC\n",
    "    \"Early Myeloid\": \"#00ff1e\", # early granulo/mye \n",
    "    \"Megakaryocyte\": \"#241717\", # MKC\n",
    "    #\"RN6\":  '#00ba9e', # other myelo (cDC, ba/eo/ma, low confidence)\n",
    "    #\"RN7\":  \"#00f7ff\",  # early B and myelo\n",
    "    \"Lymphoid\":  \"#b50d0d\",  # cytotoxic T NK\n",
    "    \"Endothelial\":  \"#de9835\",  # endothelial\n",
    "    #\"RN10\": \"#c6db02\",  # HSPC\n",
    "    #\"RN11\": \"#7875ff\",  # lymphoid\n",
    "    \"Endosteal\": \"#fabc02\",  # fibro/osteo\n",
    "    \"Arterial\": \"#735b2e\",  # pericyte\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_dim = max(p157_t0.obs['x_centroid'].max(), p157_t0.obs['y_centroid'].max())\n",
    "min_dim = min(p157_t0.obs['x_centroid'].max(), p157_t0.obs['y_centroid'].max())\n",
    "fig_size = max_dim//1000, min_dim//1000\n",
    "fig_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p157_t0 = banksy_filtered[banksy_filtered.obs['DI_Sample'] == 'P157_T0_S1']\n",
    "max_dim = max(p157_t0.obs['x_centroid'].max(), p157_t0.obs['y_centroid'].max())\n",
    "min_dim = min(p157_t0.obs['x_centroid'].max(), p157_t0.obs['y_centroid'].max())\n",
    "fig_size = max_dim//1000, min_dim//1000\n",
    "plt.figure(figsize = fig_size)\n",
    "sns.scatterplot(p157_t0.obs.copy(), x = 'x_centroid', y = 'y_centroid', hue = 'nbhd_annot', palette = neighborhood_colors, s = .5, edgecolor = None, legend = False, rasterized = True)\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/IRD_JW_Xenium_BANKSY_neighborhood_scatterplot_P157T0_v11062025.pdf', dpi=300, transparent = True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p143_t0 = banksy_filtered[banksy_filtered.obs['DI_Sample'] == 'P143_T0_S1']\n",
    "max_dim = max(p143_t0.obs['x_centroid'].max(), p143_t0.obs['y_centroid'].max())\n",
    "min_dim = min(p143_t0.obs['x_centroid'].max(), p143_t0.obs['y_centroid'].max())\n",
    "fig_size = max_dim//1000, min_dim//1000\n",
    "plt.figure(figsize = fig_size)\n",
    "sns.scatterplot(p143_t0.obs.copy(), x = 'y_centroid', y = 'x_centroid', hue = 'nbhd_annot', palette = neighborhood_colors, s = .5, edgecolor = None, legend = False, rasterized = True)\n",
    "plt.gca().set_aspect('equal', adjustable='box')\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/IRD_JW_Xenium_BANKSY_neighborhood_scatterplot_P143T0_v11062025.pdf', dpi=300, transparent = True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Neighborhood abundance analysis across samples and conditions\n",
    "\n",
    "Quantify how the proportion of each spatial neighborhood changes across disease stages. This analysis reveals which neighborhoods increase or decrease in proportion during disease progression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate fraction of cells in each BANKSY neighborhood per sample\n",
    "neighborhood_counts = banksy_filtered.obs.groupby(['Sample', 'nbhd_annot']).size().reset_index(name='count')\n",
    "total_counts = banksy_filtered.obs.groupby('Sample').size().reset_index(name='total_count')\n",
    "neighborhood_frac = neighborhood_counts.merge(total_counts, on='Sample')\n",
    "neighborhood_frac['fraction'] = neighborhood_frac['count'] / neighborhood_frac['total_count']\n",
    "neighborhood_frac.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Match sample ID to collection timepoint\n",
    "metadata = banksy_filtered.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "neighborhood_frac['Collection'] = neighborhood_frac['Sample'].map(metadata['Collection'])\n",
    "neighborhood_frac.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate statistics between timepoints for each neighborhood\n",
    "stats_res = []\n",
    "tps = neighborhood_frac['Collection'].unique().tolist()\n",
    "\n",
    "for nbhd in neighborhood_frac['nbhd_annot'].unique():\n",
    "    nbhd_fracs = neighborhood_frac[neighborhood_frac['nbhd_annot'] == nbhd]\n",
    "\n",
    "    for cond1, cond2 in combinations(tps, 2):\n",
    "        cond1_frac = nbhd_fracs[nbhd_fracs['Collection'] == cond1]['fraction'].values.tolist()\n",
    "        cond2_frac = nbhd_fracs[nbhd_fracs['Collection'] == cond2]['fraction'].values.tolist()\n",
    "        if len(cond1_frac) > 0 and len(cond2_frac) > 0:\n",
    "            u_stat, p_val = mannwhitneyu(cond1_frac, cond2_frac, alternative='two-sided')\n",
    "            stats_res.append({'Condition 1': cond1, 'Condition 2': cond2, 'Neighborhood': nbhd, 'U statistic': u_stat, 'p-value': p_val})\n",
    "stats_df = pd.DataFrame(stats_res)\n",
    "stats_df['p_adj'] = multipletests(stats_df['p-value'], method='fdr_bh')[1]\n",
    "stats_df['Sig'] = stats_df['p_adj'] < 0.05\n",
    "stats_df[stats_df['Sig'] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot boxplot of neighborhood fractions across samples, colored by collection timepoint\n",
    "plt.figure(figsize=(12, 6))\n",
    "timecols = {\"NBM\": \"#0C7515\", \"NDMM\": \"#E619B9\", \"PT\": \"#CF99C3\"} \n",
    "order = neighborhood_frac['nbhd_annot'].unique().tolist()          # set your desired x order\n",
    "hue_order = sorted(neighborhood_frac['Collection'].unique())  # set your desired hue order\n",
    "\n",
    "ax = sns.boxplot(\n",
    "    neighborhood_frac,\n",
    "    x='nbhd_annot', y='fraction', hue='Collection',\n",
    "    palette = timecols,\n",
    "    order=order, hue_order=hue_order,\n",
    "    showfliers=False\n",
    ")\n",
    "sns.swarmplot(\n",
    "    neighborhood_frac,\n",
    "    x='nbhd_annot', y='fraction', hue='Collection',\n",
    "    order=order, hue_order=hue_order, dodge = True,\n",
    "    color = 'k', alpha = .5, size = 3.5\n",
    ")\n",
    "\n",
    "################ These vibe coding snippets to draw significance bars ################\n",
    "# Draw significance bars\n",
    "def dodge_center(i, j, n_hue, width=0.8):\n",
    "    # seaborn's default group width is ~0.8\n",
    "    if n_hue <= 1:\n",
    "        return i\n",
    "    step = width / n_hue\n",
    "    return i - width/2 + step/2 + j*step\n",
    "\n",
    "# Geometry for stacked annotations\n",
    "ymin, ymax = ax.get_ylim()\n",
    "y_span = ymax - ymin if ymax > ymin else 1.0\n",
    "pad = 0.04 * y_span      # space above boxes\n",
    "step = 0.08 * y_span     # vertical spacing between annotations\n",
    "h = 0.015 * y_span       # bracket height\n",
    "\n",
    "top_y_by_nbhd = neighborhood_frac.groupby('nbhd_annot')['fraction'].max().to_dict()\n",
    "top_y_by_nbhd = {k: v + pad for k, v in top_y_by_nbhd.items()}\n",
    "stack_idx = {k: 0 for k in order}\n",
    "\n",
    "# Optional: avoid duplicate pair annotations within the same neighborhood\n",
    "seen_pairs = set()\n",
    "\n",
    "for _, row in stats_df.iterrows():\n",
    "    if not row['Sig']:\n",
    "        continue\n",
    "    nbhd = row['Neighborhood']\n",
    "    cond1, cond2 = row['Condition 1'], row['Condition 2']\n",
    "    if nbhd not in top_y_by_nbhd:\n",
    "        continue\n",
    "    # skip if one of the groups has no data\n",
    "    m1 = ((neighborhood_frac['nbhd_annot'] == nbhd) &\n",
    "          (neighborhood_frac['Collection'] == cond1))\n",
    "    m2 = ((neighborhood_frac['nbhd_annot'] == nbhd) &\n",
    "          (neighborhood_frac['Collection'] == cond2))\n",
    "    if not m1.any() or not m2.any():\n",
    "        continue\n",
    "\n",
    "    # de-duplicate symmetrical pairs per neighborhood\n",
    "    key = (nbhd, tuple(sorted([cond1, cond2])))\n",
    "    if key in seen_pairs:\n",
    "        continue\n",
    "    seen_pairs.add(key)\n",
    "\n",
    "    i = order.index(nbhd)\n",
    "    j1 = hue_order.index(cond1)\n",
    "    j2 = hue_order.index(cond2)\n",
    "    x1 = dodge_center(i, j1, len(hue_order), width=0.8)\n",
    "    x2 = dodge_center(i, j2, len(hue_order), width=0.8)\n",
    "\n",
    "    # stacked y position\n",
    "    y_base = top_y_by_nbhd[nbhd] + stack_idx[nbhd] * step\n",
    "    y0, y1 = y_base - h, y_base\n",
    "\n",
    "    ax.plot([x1, x1, x2, x2], [y0, y1, y1, y0], lw=1.5, c='k', clip_on=False)\n",
    "    ax.text((x1 + x2) / 2, y1 + 0.5*h, f\"p={row['p_adj']:.3e}\",\n",
    "            ha='center', va='bottom', color='k', fontsize=8, clip_on=False)\n",
    "\n",
    "    stack_idx[nbhd] += 1\n",
    "\n",
    "# Ensure enough headroom\n",
    "max_top = max((top_y_by_nbhd[k] + max(0, stack_idx[k]-1) * step + pad) for k in order)\n",
    "if max_top > ymax:\n",
    "    ax.set_ylim(ymin, max_top)\n",
    "########################## End of vibe coding snippets ################\n",
    "\n",
    "ax.set_title('Fraction of Cells in Each BANKSY Neighborhood Across Samples')\n",
    "ax.set_xlabel('BANKSY Neighborhood')\n",
    "ax.set_ylabel('Proportion of Cells')\n",
    "ax.set_ylim(0, 0.65)\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "sns.despine()\n",
    "plt.tight_layout()\n",
    "plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/IRD_JW_Xenium_BANKSY_neighborhood_proportions_v11062025.pdf', dpi=300, transparent = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cell type composition analysis within neighborhood across conditions\n",
    "\n",
    "Analyze how cellular composition within each neighborhood changes across disease stages. This reveals whether specific cell types are recruited to or lost from particular neighborhoods during disease progression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate proportion of cell types within each BANKSY neighborhood per sample\n",
    "celltype_counts = banksy_filtered.obs.groupby(['nbhd_annot', 'revised_annot', 'Sample']).size().reset_index(name='count')\n",
    "total_counts = banksy_filtered.obs.groupby(['nbhd_annot', 'Sample']).size().reset_index(name='total_count')\n",
    "celltype_frac = celltype_counts.merge(total_counts, on=['nbhd_annot', 'Sample'])\n",
    "celltype_frac['fraction'] = celltype_frac['count'] / celltype_frac['total_count']\n",
    "# Map collection timepoint\n",
    "metadata = banksy_filtered.obs[['Sample', 'Collection']].drop_duplicates().set_index('Sample')\n",
    "celltype_frac['Collection'] = celltype_frac['Sample'].map(metadata['Collection'])\n",
    "celltype_frac.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate proportion of cell types per sample\n",
    "celltype_sample_frac = banksy_filtered.obs.groupby(['revised_annot', 'Sample']).size().reset_index(name='count')\n",
    "total_sample_counts = banksy_filtered.obs.groupby('Sample').size().reset_index(name='total_count')\n",
    "celltype_sample_frac = celltype_sample_frac.merge(total_sample_counts, on='Sample')\n",
    "celltype_sample_frac['fraction'] = celltype_sample_frac['count'] / celltype_sample_frac['total_count']\n",
    "celltype_sample_frac['Collection'] = celltype_sample_frac['Sample'].map(metadata['Collection'])\n",
    "celltype_sample_frac.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_name = \"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/IRD_JW_Xenium_merge_BANKSY_filtered_celltype_fraction_by_nbhd_110625.pdf\"\n",
    "\n",
    "with PdfPages(pdf_name) as pdf:\n",
    "    # First plot overall cell type composition per sample across conditions\n",
    "    plt.figure(figsize = (8, 4))\n",
    "    sns.boxplot(celltype_sample_frac, x='revised_annot', y='fraction', hue='Collection')\n",
    "    plt.xlabel('Cell Type')\n",
    "    plt.ylabel('Fraction within Sample')\n",
    "    plt.title('Overall Cell Type Composition Across Samples')\n",
    "    plt.xticks(rotation=45, ha='right')\n",
    "    plt.legend(title='Collection', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    pdf.savefig()\n",
    "    plt.close()\n",
    "    \n",
    "    # Plot boxplot of cell type fractions within each BANKSY neighborhood across samples\n",
    "    nbhd_list = celltype_frac['nbhd_annot'].unique().tolist()\n",
    "    for nbhd in nbhd_list:\n",
    "        nbhd_fracs = celltype_frac[celltype_frac['nbhd_annot'] == nbhd]\n",
    "        plt.figure(figsize=(8, 4))\n",
    "        sns.boxplot(nbhd_fracs, x='revised_annot', y='fraction', hue='Collection')\n",
    "        plt.xlabel('Cell Type')\n",
    "        plt.ylabel(f'Fraction within BANKSY Neighborhood {nbhd}')\n",
    "        plt.title(f'Cell Type Composition in BANKSY Neighborhood {nbhd}')\n",
    "        plt.xticks(rotation=45, ha='right')\n",
    "        plt.legend(title='Collection', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "        plt.tight_layout()\n",
    "        pdf.savefig()\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare cell type proportions within neighborhoods vs overall sample proportions\n",
    "celltype_frac['Sample_fraction'] = celltype_frac.apply(\n",
    "    lambda row: celltype_sample_frac[(celltype_sample_frac['revised_annot'] == row['revised_annot']) & (celltype_sample_frac['Sample'] == row['Sample'])]['fraction'].values[0],\n",
    "    axis=1\n",
    ")\n",
    "celltype_frac['Enrichment'] = (celltype_frac['fraction'] - celltype_frac['Sample_fraction'])/celltype_frac['Sample_fraction']\n",
    "celltype_frac.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf_name = \"/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/IRD_JW_Xenium_merge_BANKSY_filtered_celltype_enrichment_by_nbhd_110625.pdf\"\n",
    "\n",
    "with PdfPages(pdf_name) as pdf:\n",
    "    # Plot boxplot of cell type fractions within each BANKSY neighborhood across samples\n",
    "    nbhd_list = celltype_frac['nbhd_annot'].unique().tolist()\n",
    "    for nbhd in nbhd_list:\n",
    "        nbhd_fracs = celltype_frac[celltype_frac['nbhd_annot'] == nbhd]\n",
    "        plt.figure(figsize=(10, 5))\n",
    "        sns.boxplot(nbhd_fracs, x='revised_annot', y='Enrichment', hue='Collection')\n",
    "        # Plot horizontal red dashed line at y=0\n",
    "        plt.axhline(0, color='red', linestyle='--')\n",
    "        plt.xlabel('Cell Type')\n",
    "        plt.ylabel(f'Enrichment within BANKSY Neighborhood - {nbhd}')\n",
    "        plt.title(f'Cell Type Enrichment in BANKSY Neighborhood - {nbhd}')\n",
    "        plt.xticks(rotation=45, ha='right')\n",
    "        plt.legend(title='Collection', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "        plt.tight_layout()\n",
    "        pdf.savefig()\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plasma cells in neighborhood 6 - Early Myeloid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pc_nbhd6 = celltype_frac[(celltype_frac['nbhd_annot'] == 'Early Myeloid') & (celltype_frac['revised_annot'] == 'PC')]\n",
    "fig, _ = plot_utils.plot_comparison_with_significance(pc_nbhd6, 'Collection', 'Enrichment', palette={\"NBM\": \"#0C7515\", \"NDMM\": \"#E619B9\", \"PT\": \"#CF99C3\"}, \n",
    "                                       xlabel=None, ylabel=None, title='Early Myeloid Neighborhood', figsize = (4, 6))\n",
    "plt.axhline(0, color='red', linestyle='--')\n",
    "ax = fig.get_axes()[0]\n",
    "ax1 = ax.twinx()\n",
    "ax1.set_ylim(ax.get_ylim())\n",
    "plt.tight_layout()\n",
    "plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/IRD_JW_Xenium_BANKSY_neighborhood6_plasma_cell_enrichment_v11062025.pdf', dpi=300, transparent = True)\n",
    "                                       #save_path=None, figsize=(6, 5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### NK cells in endothelial neighborhoods (nbhd 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate statistics between timepoints for NK cells in nbhd3\n",
    "pc_nbhd6 = celltype_frac[(celltype_frac['nbhd_annot'] == 'Endothelial') & (celltype_frac['revised_annot'] == 'NK')]\n",
    "stats_res = []\n",
    "tps = pc_nbhd6['Collection'].unique().tolist()\n",
    "\n",
    "for cond1, cond2 in combinations(tps, 2):\n",
    "    cond1_frac = pc_nbhd6[pc_nbhd6['Collection'] == cond1]['Enrichment'].values.tolist()\n",
    "    cond2_frac = pc_nbhd6[pc_nbhd6['Collection'] == cond2]['Enrichment'].values.tolist()\n",
    "    if len(cond1_frac) > 0 and len(cond2_frac) > 0:\n",
    "        u_stat, p_val = mannwhitneyu(cond1_frac, cond2_frac, alternative='two-sided')\n",
    "        stats_res.append({'Condition 1': cond1, 'Condition 2': cond2, 'Neighborhood': nbhd, 'U statistic': u_stat, 'p-value': p_val})\n",
    "stats_df = pd.DataFrame(stats_res)\n",
    "stats_df['p_adj'] = multipletests(stats_df['p-value'], method='fdr_bh')[1]\n",
    "stats_df['Sig'] = stats_df['p_adj'] < 0.05\n",
    "\n",
    "# Plot plasma cell proportions in nbhd2 across timepoints\n",
    "plt.figure(figsize = (4, 6))\n",
    "timecols = {\"NBM\": \"#0C7515\", \"NDMM\": \"#E619B9\", \"PT\": \"#CF99C3\"} \n",
    "\n",
    "ax = sns.boxplot(pc_nbhd6, x='Collection', y='Enrichment', hue = 'Collection', palette = timecols, showfliers=False)\n",
    "sns.swarmplot(pc_nbhd6, x='Collection', y='Enrichment', color='black', size=5, alpha = .5)\n",
    "plt.axhline(0, color='red', linestyle='--')\n",
    "\n",
    "# Prepare bracket positions (no hue)\n",
    "order = ['NBM', 'NDMM', 'PT']\n",
    "x_pos = {lab: i for i, lab in enumerate(order)}\n",
    "\n",
    "# Geometry\n",
    "ymin, ymax = ax.get_ylim()\n",
    "y_span = max(ymax - ymin, 1.0)\n",
    "pad = 0.05 * y_span\n",
    "step = 0.08 * y_span\n",
    "h = 0.015 * y_span\n",
    "text_pad = 0.5 * h\n",
    "\n",
    "base = pc_nbhd6['Enrichment'].max() + pad\n",
    "levels = [0] * len(order)  # per-x used levels to stack nested brackets\n",
    "\n",
    "# Build pairs as (i, j, p_adj), sort by span so inner brackets go lower\n",
    "pairs = []\n",
    "for _, r in stats_df.iterrows():\n",
    "    i, j = x_pos[r['Condition 1']], x_pos[r['Condition 2']]\n",
    "    if i == j:\n",
    "        continue\n",
    "    if i > j:\n",
    "        i, j = j, i\n",
    "    pairs.append((i, j, r['p_adj']))\n",
    "\n",
    "pairs.sort(key=lambda t: (t[1] - t[0], t[0]))  # shortest span first\n",
    "\n",
    "for i, j, p in pairs:\n",
    "    lvl = max(levels[i:j+1])  # find first free level across the span\n",
    "    y1 = base + lvl * step\n",
    "    y0 = y1 - h\n",
    "    x1, x2 = i, j\n",
    "\n",
    "    ax.plot([x1, x1, x2, x2], [y0, y1, y1, y0], lw=1.5, c='k', clip_on=False)\n",
    "    ax.text((x1 + x2) / 2, y1 + text_pad, f\"p={p:.2g}\", ha='center', va='bottom', fontsize=9, clip_on=False)\n",
    "\n",
    "    # reserve this level across the span so outer brackets go higher\n",
    "    for k in range(i, j + 1):\n",
    "        levels[k] = lvl + 1\n",
    "\n",
    "# Ensure headroom\n",
    "needed_top = base + (max(levels) * step) + pad\n",
    "if needed_top > ymax:\n",
    "    ax.set_ylim(ymin, needed_top)\n",
    "\n",
    "plt.xlabel('Timepoint')\n",
    "plt.ylabel('Fold enrichment of NK cells in neighborhood')\n",
    "plt.title('Endothelial Neighborhood (nbhd3)')\n",
    "plt.tight_layout()\n",
    "plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/IRD_JW_Xenium_BANKSY_neighborhood3_NK_cell_enrichment_v11062025.pdf', dpi=300, transparent = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison with radial neighborhoods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rn_obj = sc.read_h5ad(\"/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/Xenium/analysis/radial_neighborhoods/Output/merged_RN.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rn_obj.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy_filtered.obs['RN_nbhd'] = rn_obj.obs.loc[banksy_filtered.obs_names, 'rn']\n",
    "banksy_filtered = banksy_filtered[banksy_filtered.obs['RN_nbhd'] != 'Unassigned']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nbhd_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nbhd_counts = banksy_filtered.obs.groupby(['nbhd_annot', 'RN_nbhd']).size().reset_index(name='count')\n",
    "rn_nbhd_counts = nbhd_counts.groupby('RN_nbhd')['count'].sum()\n",
    "rn_nbhd_counts = rn_nbhd_counts.to_frame().reset_index()\n",
    "nbhd_counts = nbhd_counts.merge(rn_nbhd_counts, on = 'RN_nbhd')\n",
    "nbhd_counts['fraction'] = nbhd_counts['count_x'] / nbhd_counts['count_y']\n",
    "nbhd_counts.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(nbhd_counts, y = 'RN_nbhd', x = 'nbhd_annot', size = 'fraction', hue = 'fraction', palette = 'viridis')\n",
    "plt.legend(title = 'BANKSY Neighborhood', bbox_to_anchor = (1.05, 1), loc = 'upper left')\n",
    "plt.xticks(rotation = 90, ha = 'center')\n",
    "plt.xlabel('Radial Neighborhood')\n",
    "plt.ylabel('BANKSY Neighborhood')\n",
    "plt.title('Fraction of BANKSY Neighborhoods in each Radial Neighborhood')\n",
    "plt.tight_layout()\n",
    "#plt.savefig('/diskmnt/Users2/chouw/Projects/BM_spatial/IRD/BANKSY/IRD_JW_Xenium_BANKSY_neighborhood_fractions_by_RN_v11062025.pdf', dpi=300, transparent = True)\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
