{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8dca7db9-605f-4f0b-8a29-1753fc4bdc04",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, pack, unpack\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:reshape2’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ────────\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.4     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.1.0     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.3.0\n",
      "── \u001b[1mConflicts\u001b[22m ─────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m    masks \u001b[34mMatrix\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m    masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m       masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mpack()\u001b[39m      masks \u001b[34mMatrix\u001b[39m::pack()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32mstamp()\u001b[39m masks \u001b[34mcowplot\u001b[39m::stamp()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32munpack()\u001b[39m    masks \u001b[34mMatrix\u001b[39m::unpack()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.25.2\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "Welcome to enrichR\n",
      "Checking connections ... \n",
      "\n",
      "Enrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "FlyEnrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "WormEnrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "YeastEnrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "FishEnrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "OxEnrichr ... \n",
      "Connection is Live!\n",
      "\n",
      "\n",
      "Attaching package: ‘ggpubr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:cowplot’:\n",
      "\n",
      "    get_legend\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:lubridate’:\n",
      "\n",
      "    second, second<-\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:lubridate’:\n",
      "\n",
      "    %within%\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    reduce\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘rstatix’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    desc\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    filter\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "########################################################\n",
    "# run using cellchat environment\n",
    "########################################################\n",
    "#library(harmony)\n",
    "library(RColorBrewer)\n",
    "library(Seurat)\n",
    "library(ggplot2)\n",
    "library(reshape2)\n",
    "library(viridis)\n",
    "library(dplyr)\n",
    "library(Matrix)\n",
    "library(cowplot)\n",
    "library(getopt)\n",
    "library(tidyr)\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "library(ComplexHeatmap)\n",
    "library(scProportionTest)\n",
    "library(enrichR)\n",
    "library(ggpubr)\n",
    "library(DESeq2)\n",
    "library(rstatix)\n",
    "options(future.globals.maxSize = 500 * 1024^3) \n",
    "\n",
    "#---------Gene Annotations---------------------------------------\n",
    "gs <- readRDS('/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/analysis/resources/gene_sets.rds')\n",
    "annotations <- gs$annotations\n",
    "s.genes <- gs$s.genes\n",
    "g2m.genes <- gs$g2m.genes\n",
    "ig_genes <- gs$ig_genes #length=436\n",
    "hemogenes <- gs$hemo_genes\n",
    "protein_coding_genes <- gs$protein_coding_genes\n",
    "#--- palettes\n",
    "colors <- readRDS('/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/analysis/resources/colors.rds')\n",
    "#----------------------------------------------\n",
    "# function for filtering out mitochondrial, hemoglobin, Ig, ribosomal genes from a list of genes\n",
    "#----------------------------------------------\n",
    "filter_genes <- function(genes){\n",
    "  de_fi1 <- genes[!genes%in% ig_genes & !genes %in% hemogenes]\n",
    "  de_fim1 <- de_fi1[!str_starts(de_fi1, 'MT-')]\n",
    "  de_fimr1 <- de_fim1[!str_starts(de_fim1, 'RPL') & !str_starts(de_fim1, 'RPS')]\n",
    "  genesf <- de_fimr1\n",
    "  return(genesf)\n",
    "}\n",
    "# return only protein-coding genes (excl. mito, hemo, Ig, ribo)\n",
    "filter_genes_p <- function(genes){\n",
    "de_fi1 <- genes[!genes%in% ig_genes & !genes %in% hemogenes & genes %in% protein_coding_genes]\n",
    "  de_fim1 <- de_fi1[!str_starts(de_fi1, 'MT-')]\n",
    "  de_fimr1 <- de_fim1[!str_starts(de_fim1, 'RPL') & !str_starts(de_fim1, 'RPS')]\n",
    "  genesf <- de_fimr1\n",
    "  return(genesf)\n",
    "}\n",
    "\n",
    "#-----------------------\n",
    "setwd(\"/diskmnt/Projects/myeloma_scRNA_analysis/MMY_IRD/analysis/MMRF_longitudinal\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a969572f-fedd-4dbb-946b-342186661564",
   "metadata": {},
   "outputs": [],
   "source": [
    "# longitudinal object\n",
    "obj <- readRDS('mmrf_longitudinal_noDoublets.rds')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "886a21fe-1f5f-41a6-ac88-a358c6a0f15a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 36</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>sample_id</th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>site</th><th scope=col>sample_orig_id</th><th scope=col>batch_num</th><th scope=col>siteXbatch</th><th scope=col>RNA_snn_res.0.5</th><th scope=col>seurat_clusters</th><th scope=col>⋯</th><th scope=col>G2M.Score</th><th scope=col>Phase</th><th scope=col>old.ident</th><th scope=col>CC.Difference</th><th scope=col>nCount_SCT</th><th scope=col>nFeature_SCT</th><th scope=col>IRD_subset_score</th><th scope=col>IRD_subset</th><th scope=col>IRD_subset_f</th><th scope=col>celltype_f</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GGGAGTATCCAGTTCC-1_1284_010761_SM</th><td>1284_010761_SM</td><td>MMRF</td><td>9903</td><td> 791</td><td>EMORY</td><td>1284_010761</td><td>B2</td><td>EMORY_B2</td><td>13</td><td>13</td><td>⋯</td><td> 0.01252350</td><td>G2M</td><td>13</td><td>-0.05217056</td><td>5302</td><td> 569</td><td>1.0000000</td><td>LateEry   </td><td>LateEry   </td><td>Erythrocyte_IFN   </td></tr>\n",
       "\t<tr><th scope=row>CCACGAGTCACCCTCA-1_1284_010761_SM</th><td>1284_010761_SM</td><td>MMRF</td><td>9978</td><td>2909</td><td>EMORY</td><td>1284_010761</td><td>B2</td><td>EMORY_B2</td><td>7 </td><td>7 </td><td>⋯</td><td> 0.69719738</td><td>G2M</td><td>7 </td><td>-0.25171374</td><td>5312</td><td>2428</td><td>0.5431845</td><td>EarlyEry  </td><td>EarlyEry  </td><td>EB_MKi67_1        </td></tr>\n",
       "\t<tr><th scope=row>TAAGCCAGTCCACTTC-1_1284_010761_SM</th><td>1284_010761_SM</td><td>MMRF</td><td>9899</td><td>2650</td><td>EMORY</td><td>1284_010761</td><td>B2</td><td>EMORY_B2</td><td>7 </td><td>7 </td><td>⋯</td><td> 0.69741525</td><td>G2M</td><td>7 </td><td>-0.27604777</td><td>5263</td><td>2170</td><td>0.8678911</td><td>LateEry   </td><td>LateEry   </td><td>EB_MKi67_1        </td></tr>\n",
       "\t<tr><th scope=row>ATGACCACACATATGC-1_1284_010761_SM</th><td>1284_010761_SM</td><td>MMRF</td><td>9950</td><td>2658</td><td>EMORY</td><td>1284_010761</td><td>B2</td><td>EMORY_B2</td><td>7 </td><td>7 </td><td>⋯</td><td> 0.58559358</td><td>G2M</td><td>7 </td><td>-0.32687868</td><td>5294</td><td>2241</td><td>0.6894119</td><td>LateEry   </td><td>LateEry   </td><td>EB                </td></tr>\n",
       "\t<tr><th scope=row>GGTTAACTCCTTCGAC-1_1284_010761_SM</th><td>1284_010761_SM</td><td>MMRF</td><td>9959</td><td>2301</td><td>EMORY</td><td>1284_010761</td><td>B2</td><td>EMORY_B2</td><td>9 </td><td>9 </td><td>⋯</td><td>-0.05608178</td><td>S  </td><td>9 </td><td> 0.12818399</td><td>4985</td><td>1801</td><td>1.0000000</td><td>Neutrophil</td><td>Neutrophil</td><td>Neutrophil_RPS/RPL</td></tr>\n",
       "\t<tr><th scope=row>TTGCGTCTCCGTGGTG-1_1284_010761_SM</th><td>1284_010761_SM</td><td>MMRF</td><td>9892</td><td>1576</td><td>EMORY</td><td>1284_010761</td><td>B2</td><td>EMORY_B2</td><td>14</td><td>14</td><td>⋯</td><td> 0.61417626</td><td>G2M</td><td>14</td><td>-0.50606499</td><td>5343</td><td>1224</td><td>1.0000000</td><td>LateEry   </td><td>LateEry   </td><td>EB_MKi67_1        </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 36\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & sample\\_id & orig.ident & nCount\\_RNA & nFeature\\_RNA & site & sample\\_orig\\_id & batch\\_num & siteXbatch & RNA\\_snn\\_res.0.5 & seurat\\_clusters & ⋯ & G2M.Score & Phase & old.ident & CC.Difference & nCount\\_SCT & nFeature\\_SCT & IRD\\_subset\\_score & IRD\\_subset & IRD\\_subset\\_f & celltype\\_f\\\\\n",
       "  & <chr> & <fct> & <dbl> & <int> & <chr> & <chr> & <chr> & <chr> & <fct> & <fct> & ⋯ & <dbl> & <chr> & <fct> & <dbl> & <dbl> & <int> & <dbl> & <fct> & <fct> & <fct>\\\\\n",
       "\\hline\n",
       "\tGGGAGTATCCAGTTCC-1\\_1284\\_010761\\_SM & 1284\\_010761\\_SM & MMRF & 9903 &  791 & EMORY & 1284\\_010761 & B2 & EMORY\\_B2 & 13 & 13 & ⋯ &  0.01252350 & G2M & 13 & -0.05217056 & 5302 &  569 & 1.0000000 & LateEry    & LateEry    & Erythrocyte\\_IFN   \\\\\n",
       "\tCCACGAGTCACCCTCA-1\\_1284\\_010761\\_SM & 1284\\_010761\\_SM & MMRF & 9978 & 2909 & EMORY & 1284\\_010761 & B2 & EMORY\\_B2 & 7  & 7  & ⋯ &  0.69719738 & G2M & 7  & -0.25171374 & 5312 & 2428 & 0.5431845 & EarlyEry   & EarlyEry   & EB\\_MKi67\\_1        \\\\\n",
       "\tTAAGCCAGTCCACTTC-1\\_1284\\_010761\\_SM & 1284\\_010761\\_SM & MMRF & 9899 & 2650 & EMORY & 1284\\_010761 & B2 & EMORY\\_B2 & 7  & 7  & ⋯ &  0.69741525 & G2M & 7  & -0.27604777 & 5263 & 2170 & 0.8678911 & LateEry    & LateEry    & EB\\_MKi67\\_1        \\\\\n",
       "\tATGACCACACATATGC-1\\_1284\\_010761\\_SM & 1284\\_010761\\_SM & MMRF & 9950 & 2658 & EMORY & 1284\\_010761 & B2 & EMORY\\_B2 & 7  & 7  & ⋯ &  0.58559358 & G2M & 7  & -0.32687868 & 5294 & 2241 & 0.6894119 & LateEry    & LateEry    & EB                \\\\\n",
       "\tGGTTAACTCCTTCGAC-1\\_1284\\_010761\\_SM & 1284\\_010761\\_SM & MMRF & 9959 & 2301 & EMORY & 1284\\_010761 & B2 & EMORY\\_B2 & 9  & 9  & ⋯ & -0.05608178 & S   & 9  &  0.12818399 & 4985 & 1801 & 1.0000000 & Neutrophil & Neutrophil & Neutrophil\\_RPS/RPL\\\\\n",
       "\tTTGCGTCTCCGTGGTG-1\\_1284\\_010761\\_SM & 1284\\_010761\\_SM & MMRF & 9892 & 1576 & EMORY & 1284\\_010761 & B2 & EMORY\\_B2 & 14 & 14 & ⋯ &  0.61417626 & G2M & 14 & -0.50606499 & 5343 & 1224 & 1.0000000 & LateEry    & LateEry    & EB\\_MKi67\\_1        \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 36\n",
       "\n",
       "| <!--/--> | sample_id &lt;chr&gt; | orig.ident &lt;fct&gt; | nCount_RNA &lt;dbl&gt; | nFeature_RNA &lt;int&gt; | site &lt;chr&gt; | sample_orig_id &lt;chr&gt; | batch_num &lt;chr&gt; | siteXbatch &lt;chr&gt; | RNA_snn_res.0.5 &lt;fct&gt; | seurat_clusters &lt;fct&gt; | ⋯ ⋯ | G2M.Score &lt;dbl&gt; | Phase &lt;chr&gt; | old.ident &lt;fct&gt; | CC.Difference &lt;dbl&gt; | nCount_SCT &lt;dbl&gt; | nFeature_SCT &lt;int&gt; | IRD_subset_score &lt;dbl&gt; | IRD_subset &lt;fct&gt; | IRD_subset_f &lt;fct&gt; | celltype_f &lt;fct&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| GGGAGTATCCAGTTCC-1_1284_010761_SM | 1284_010761_SM | MMRF | 9903 |  791 | EMORY | 1284_010761 | B2 | EMORY_B2 | 13 | 13 | ⋯ |  0.01252350 | G2M | 13 | -0.05217056 | 5302 |  569 | 1.0000000 | LateEry    | LateEry    | Erythrocyte_IFN    |\n",
       "| CCACGAGTCACCCTCA-1_1284_010761_SM | 1284_010761_SM | MMRF | 9978 | 2909 | EMORY | 1284_010761 | B2 | EMORY_B2 | 7  | 7  | ⋯ |  0.69719738 | G2M | 7  | -0.25171374 | 5312 | 2428 | 0.5431845 | EarlyEry   | EarlyEry   | EB_MKi67_1         |\n",
       "| TAAGCCAGTCCACTTC-1_1284_010761_SM | 1284_010761_SM | MMRF | 9899 | 2650 | EMORY | 1284_010761 | B2 | EMORY_B2 | 7  | 7  | ⋯ |  0.69741525 | G2M | 7  | -0.27604777 | 5263 | 2170 | 0.8678911 | LateEry    | LateEry    | EB_MKi67_1         |\n",
       "| ATGACCACACATATGC-1_1284_010761_SM | 1284_010761_SM | MMRF | 9950 | 2658 | EMORY | 1284_010761 | B2 | EMORY_B2 | 7  | 7  | ⋯ |  0.58559358 | G2M | 7  | -0.32687868 | 5294 | 2241 | 0.6894119 | LateEry    | LateEry    | EB                 |\n",
       "| GGTTAACTCCTTCGAC-1_1284_010761_SM | 1284_010761_SM | MMRF | 9959 | 2301 | EMORY | 1284_010761 | B2 | EMORY_B2 | 9  | 9  | ⋯ | -0.05608178 | S   | 9  |  0.12818399 | 4985 | 1801 | 1.0000000 | Neutrophil | Neutrophil | Neutrophil_RPS/RPL |\n",
       "| TTGCGTCTCCGTGGTG-1_1284_010761_SM | 1284_010761_SM | MMRF | 9892 | 1576 | EMORY | 1284_010761 | B2 | EMORY_B2 | 14 | 14 | ⋯ |  0.61417626 | G2M | 14 | -0.50606499 | 5343 | 1224 | 1.0000000 | LateEry    | LateEry    | EB_MKi67_1         |\n",
       "\n"
      ],
      "text/plain": [
       "                                  sample_id      orig.ident nCount_RNA\n",
       "GGGAGTATCCAGTTCC-1_1284_010761_SM 1284_010761_SM MMRF       9903      \n",
       "CCACGAGTCACCCTCA-1_1284_010761_SM 1284_010761_SM MMRF       9978      \n",
       "TAAGCCAGTCCACTTC-1_1284_010761_SM 1284_010761_SM MMRF       9899      \n",
       "ATGACCACACATATGC-1_1284_010761_SM 1284_010761_SM MMRF       9950      \n",
       "GGTTAACTCCTTCGAC-1_1284_010761_SM 1284_010761_SM MMRF       9959      \n",
       "TTGCGTCTCCGTGGTG-1_1284_010761_SM 1284_010761_SM MMRF       9892      \n",
       "                                  nFeature_RNA site  sample_orig_id batch_num\n",
       "GGGAGTATCCAGTTCC-1_1284_010761_SM  791         EMORY 1284_010761    B2       \n",
       "CCACGAGTCACCCTCA-1_1284_010761_SM 2909         EMORY 1284_010761    B2       \n",
       "TAAGCCAGTCCACTTC-1_1284_010761_SM 2650         EMORY 1284_010761    B2       \n",
       "ATGACCACACATATGC-1_1284_010761_SM 2658         EMORY 1284_010761    B2       \n",
       "GGTTAACTCCTTCGAC-1_1284_010761_SM 2301         EMORY 1284_010761    B2       \n",
       "TTGCGTCTCCGTGGTG-1_1284_010761_SM 1576         EMORY 1284_010761    B2       \n",
       "                                  siteXbatch RNA_snn_res.0.5 seurat_clusters ⋯\n",
       "GGGAGTATCCAGTTCC-1_1284_010761_SM EMORY_B2   13              13              ⋯\n",
       "CCACGAGTCACCCTCA-1_1284_010761_SM EMORY_B2   7               7               ⋯\n",
       "TAAGCCAGTCCACTTC-1_1284_010761_SM EMORY_B2   7               7               ⋯\n",
       "ATGACCACACATATGC-1_1284_010761_SM EMORY_B2   7               7               ⋯\n",
       "GGTTAACTCCTTCGAC-1_1284_010761_SM EMORY_B2   9               9               ⋯\n",
       "TTGCGTCTCCGTGGTG-1_1284_010761_SM EMORY_B2   14              14              ⋯\n",
       "                                  G2M.Score   Phase old.ident CC.Difference\n",
       "GGGAGTATCCAGTTCC-1_1284_010761_SM  0.01252350 G2M   13        -0.05217056  \n",
       "CCACGAGTCACCCTCA-1_1284_010761_SM  0.69719738 G2M   7         -0.25171374  \n",
       "TAAGCCAGTCCACTTC-1_1284_010761_SM  0.69741525 G2M   7         -0.27604777  \n",
       "ATGACCACACATATGC-1_1284_010761_SM  0.58559358 G2M   7         -0.32687868  \n",
       "GGTTAACTCCTTCGAC-1_1284_010761_SM -0.05608178 S     9          0.12818399  \n",
       "TTGCGTCTCCGTGGTG-1_1284_010761_SM  0.61417626 G2M   14        -0.50606499  \n",
       "                                  nCount_SCT nFeature_SCT IRD_subset_score\n",
       "GGGAGTATCCAGTTCC-1_1284_010761_SM 5302        569         1.0000000       \n",
       "CCACGAGTCACCCTCA-1_1284_010761_SM 5312       2428         0.5431845       \n",
       "TAAGCCAGTCCACTTC-1_1284_010761_SM 5263       2170         0.8678911       \n",
       "ATGACCACACATATGC-1_1284_010761_SM 5294       2241         0.6894119       \n",
       "GGTTAACTCCTTCGAC-1_1284_010761_SM 4985       1801         1.0000000       \n",
       "TTGCGTCTCCGTGGTG-1_1284_010761_SM 5343       1224         1.0000000       \n",
       "                                  IRD_subset IRD_subset_f celltype_f        \n",
       "GGGAGTATCCAGTTCC-1_1284_010761_SM LateEry    LateEry      Erythrocyte_IFN   \n",
       "CCACGAGTCACCCTCA-1_1284_010761_SM EarlyEry   EarlyEry     EB_MKi67_1        \n",
       "TAAGCCAGTCCACTTC-1_1284_010761_SM LateEry    LateEry      EB_MKi67_1        \n",
       "ATGACCACACATATGC-1_1284_010761_SM LateEry    LateEry      EB                \n",
       "GGTTAACTCCTTCGAC-1_1284_010761_SM Neutrophil Neutrophil   Neutrophil_RPS/RPL\n",
       "TTGCGTCTCCGTGGTG-1_1284_010761_SM LateEry    LateEry      EB_MKi67_1        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(obj@meta.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e52d3f46-12de-44c6-bc04-91e4f0921c8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'sample_id'</li><li>'orig.ident'</li><li>'nCount_RNA'</li><li>'nFeature_RNA'</li><li>'site'</li><li>'sample_orig_id'</li><li>'batch_num'</li><li>'siteXbatch'</li><li>'RNA_snn_res.0.5'</li><li>'seurat_clusters'</li><li>'barcode_full'</li><li>'percent.mt'</li><li>'ambient_rho'</li><li>'doublet_score'</li><li>'predicted_doublet'</li><li>'cohort'</li><li>'seurat_subclusters_label_transferring_Yizhe_v1'</li><li>'celltype_subclusters_label_transferring_Yizhe_v1'</li><li>'seurat_subclusters_integration_Yizhe_v1'</li><li>'celltype_subclusters_integration_Yizhe_v1'</li><li>'compartment'</li><li>'UMAP_1'</li><li>'UMAP_2'</li><li>'public_id'</li><li>'collection_event'</li><li>'S.Score'</li><li>'G2M.Score'</li><li>'Phase'</li><li>'old.ident'</li><li>'CC.Difference'</li><li>'nCount_SCT'</li><li>'nFeature_SCT'</li><li>'IRD_subset_score'</li><li>'IRD_subset'</li><li>'IRD_subset_f'</li><li>'celltype_f'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'sample\\_id'\n",
       "\\item 'orig.ident'\n",
       "\\item 'nCount\\_RNA'\n",
       "\\item 'nFeature\\_RNA'\n",
       "\\item 'site'\n",
       "\\item 'sample\\_orig\\_id'\n",
       "\\item 'batch\\_num'\n",
       "\\item 'siteXbatch'\n",
       "\\item 'RNA\\_snn\\_res.0.5'\n",
       "\\item 'seurat\\_clusters'\n",
       "\\item 'barcode\\_full'\n",
       "\\item 'percent.mt'\n",
       "\\item 'ambient\\_rho'\n",
       "\\item 'doublet\\_score'\n",
       "\\item 'predicted\\_doublet'\n",
       "\\item 'cohort'\n",
       "\\item 'seurat\\_subclusters\\_label\\_transferring\\_Yizhe\\_v1'\n",
       "\\item 'celltype\\_subclusters\\_label\\_transferring\\_Yizhe\\_v1'\n",
       "\\item 'seurat\\_subclusters\\_integration\\_Yizhe\\_v1'\n",
       "\\item 'celltype\\_subclusters\\_integration\\_Yizhe\\_v1'\n",
       "\\item 'compartment'\n",
       "\\item 'UMAP\\_1'\n",
       "\\item 'UMAP\\_2'\n",
       "\\item 'public\\_id'\n",
       "\\item 'collection\\_event'\n",
       "\\item 'S.Score'\n",
       "\\item 'G2M.Score'\n",
       "\\item 'Phase'\n",
       "\\item 'old.ident'\n",
       "\\item 'CC.Difference'\n",
       "\\item 'nCount\\_SCT'\n",
       "\\item 'nFeature\\_SCT'\n",
       "\\item 'IRD\\_subset\\_score'\n",
       "\\item 'IRD\\_subset'\n",
       "\\item 'IRD\\_subset\\_f'\n",
       "\\item 'celltype\\_f'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'sample_id'\n",
       "2. 'orig.ident'\n",
       "3. 'nCount_RNA'\n",
       "4. 'nFeature_RNA'\n",
       "5. 'site'\n",
       "6. 'sample_orig_id'\n",
       "7. 'batch_num'\n",
       "8. 'siteXbatch'\n",
       "9. 'RNA_snn_res.0.5'\n",
       "10. 'seurat_clusters'\n",
       "11. 'barcode_full'\n",
       "12. 'percent.mt'\n",
       "13. 'ambient_rho'\n",
       "14. 'doublet_score'\n",
       "15. 'predicted_doublet'\n",
       "16. 'cohort'\n",
       "17. 'seurat_subclusters_label_transferring_Yizhe_v1'\n",
       "18. 'celltype_subclusters_label_transferring_Yizhe_v1'\n",
       "19. 'seurat_subclusters_integration_Yizhe_v1'\n",
       "20. 'celltype_subclusters_integration_Yizhe_v1'\n",
       "21. 'compartment'\n",
       "22. 'UMAP_1'\n",
       "23. 'UMAP_2'\n",
       "24. 'public_id'\n",
       "25. 'collection_event'\n",
       "26. 'S.Score'\n",
       "27. 'G2M.Score'\n",
       "28. 'Phase'\n",
       "29. 'old.ident'\n",
       "30. 'CC.Difference'\n",
       "31. 'nCount_SCT'\n",
       "32. 'nFeature_SCT'\n",
       "33. 'IRD_subset_score'\n",
       "34. 'IRD_subset'\n",
       "35. 'IRD_subset_f'\n",
       "36. 'celltype_f'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"sample_id\"                                       \n",
       " [2] \"orig.ident\"                                      \n",
       " [3] \"nCount_RNA\"                                      \n",
       " [4] \"nFeature_RNA\"                                    \n",
       " [5] \"site\"                                            \n",
       " [6] \"sample_orig_id\"                                  \n",
       " [7] \"batch_num\"                                       \n",
       " [8] \"siteXbatch\"                                      \n",
       " [9] \"RNA_snn_res.0.5\"                                 \n",
       "[10] \"seurat_clusters\"                                 \n",
       "[11] \"barcode_full\"                                    \n",
       "[12] \"percent.mt\"                                      \n",
       "[13] \"ambient_rho\"                                     \n",
       "[14] \"doublet_score\"                                   \n",
       "[15] \"predicted_doublet\"                               \n",
       "[16] \"cohort\"                                          \n",
       "[17] \"seurat_subclusters_label_transferring_Yizhe_v1\"  \n",
       "[18] \"celltype_subclusters_label_transferring_Yizhe_v1\"\n",
       "[19] \"seurat_subclusters_integration_Yizhe_v1\"         \n",
       "[20] \"celltype_subclusters_integration_Yizhe_v1\"       \n",
       "[21] \"compartment\"                                     \n",
       "[22] \"UMAP_1\"                                          \n",
       "[23] \"UMAP_2\"                                          \n",
       "[24] \"public_id\"                                       \n",
       "[25] \"collection_event\"                                \n",
       "[26] \"S.Score\"                                         \n",
       "[27] \"G2M.Score\"                                       \n",
       "[28] \"Phase\"                                           \n",
       "[29] \"old.ident\"                                       \n",
       "[30] \"CC.Difference\"                                   \n",
       "[31] \"nCount_SCT\"                                      \n",
       "[32] \"nFeature_SCT\"                                    \n",
       "[33] \"IRD_subset_score\"                                \n",
       "[34] \"IRD_subset\"                                      \n",
       "[35] \"IRD_subset_f\"                                    \n",
       "[36] \"celltype_f\"                                      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "colnames(obj@meta.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "49ee08a3-441a-4c64-8c09-3c496634d193",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>Erythrocyte_IFN</li><li>EB_MKi67_1</li><li>EB</li><li>Neutrophil_RPS/RPL</li><li>GMP</li><li>pDCs</li><li>Pc_MALAT1</li><li>Erythrocyte</li><li>HSC</li><li>cDC2</li><li>MAIT</li><li>CD4_Teff</li><li>CD14+Mono_S100</li><li>CD14+Mono_IFN</li><li>Granulocyte</li><li>CD16+Mono</li><li>CD4_Tcm_NFKBIA</li><li>Erythrocyte_mito</li><li>CD8_Tn</li><li>CD4_Teff_TNF</li><li>CD4_Tcm_KLRB1</li><li>CD14+Mono_CTSS</li><li>CD4_Tem_IFN</li><li>B_preLg</li><li>Pc_MT-CO2</li><li>CD14+Mono pro-inflam</li><li>Treg</li><li>CD8_Tcm_GZMK</li><li>CD4_Tn</li><li>CD4_Tcm_IFN</li><li>CD4_Th</li><li>B_pro</li><li>Fibroblasts</li><li>CD8_Teff_TNF</li><li>B_mem</li><li>CD8_Teff_HLA</li><li>CD8_T_adp</li><li>NK_CD56bright</li><li>CD8_Tem_NFKB</li><li>NK_CD56dim</li><li>CD8_Teff</li><li>Pc_JUN</li><li>CD8_Tem_IFN</li><li>CD8_Tem</li><li>NK_resident</li><li>B_imm</li><li>NK_adp</li><li>Pc_IGKC</li><li>MegaK</li><li>B_naive</li><li>CD4_Th_LEF1</li><li>B_preSm</li><li>B_trns</li><li>CD8_T_Apoptotic</li><li>Erythrocyte_prolif</li><li>cDC1</li><li>Pc_HBB_S100A8</li><li>CD8_Teff_b</li><li>Macro/Mono</li><li>Pc_HBB_HMGB1</li><li>CD4_CTL</li><li>CD8_Tcm</li><li>CD8_Teff_c</li><li>M2_Macro</li><li>EB_MKi67_2</li><li>Pc_ATF5</li><li>Pc_IFI27</li><li>Pc_EIF1</li><li>CD14+Mono_hypo</li><li>Pc_CD74</li><li>B_Unswitched_mem</li><li>Pc_IGHV3</li><li>Pc_IGHG3</li><li>Neutrohpil_ARG1</li><li>Pc_PPM1K</li><li>MastC</li><li>Pc_NOL4</li><li>CD8_T_adp_b</li><li>Pc_CCDC88A</li><li>Pc_IGHA1</li><li>Pc_BCMA</li><li>Pc_PUS3</li><li>Pc_MPO</li><li>Pc_IGKV4</li></ol>\n",
       "\n",
       "<details>\n",
       "\t<summary style=display:list-item;cursor:pointer>\n",
       "\t\t<strong>Levels</strong>:\n",
       "\t</summary>\n",
       "\t<style>\n",
       "\t.list-inline {list-style: none; margin:0; padding: 0}\n",
       "\t.list-inline>li {display: inline-block}\n",
       "\t.list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "\t</style>\n",
       "\t<ol class=list-inline><li>'Pc_ATF5'</li><li>'Pc_BCMA'</li><li>'Pc_CCDC88A'</li><li>'Pc_CD74'</li><li>'Pc_EIF1'</li><li>'Pc_HBB_HMGB1'</li><li>'Pc_HBB_S100A8'</li><li>'Pc_IFI27'</li><li>'Pc_IGHA1'</li><li>'Pc_IGHG3'</li><li>'Pc_IGHV3'</li><li>'Pc_IGKV4'</li><li>'Pc_IGKC'</li><li>'Pc_JUN'</li><li>'Pc_MALAT1'</li><li>'Pc_MPO'</li><li>'Pc_MT-CO2'</li><li>'Pc_NOL4'</li><li>'Pc_PPM1K'</li><li>'Pc_PUS3'</li><li>'B_pro'</li><li>'B_preLg'</li><li>'B_preSm'</li><li>'B_imm'</li><li>'B_naive'</li><li>'B_trns'</li><li>'B_mem'</li><li>'B_Unswitched_mem'</li><li>'CD4_CTL'</li><li>'CD4_Tcm_IFN'</li><li>'CD4_Tcm_KLRB1'</li><li>'CD4_Tcm_NFKBIA'</li><li>'CD4_Teff'</li><li>'CD4_Teff_TNF'</li><li>'CD4_Tem_IFN'</li><li>'CD4_Th'</li><li>'CD4_Th_LEF1'</li><li>'CD4_Tn'</li><li>'Treg'</li><li>'CD8_T_adp'</li><li>'CD8_T_adp_b'</li><li>'CD8_T_Apoptotic'</li><li>'CD8_Tcm'</li><li>'CD8_Tcm_GZMK'</li><li>'CD8_Teff'</li><li>'CD8_Teff_b'</li><li>'CD8_Teff_c'</li><li>'CD8_Teff_HLA'</li><li>'CD8_Teff_TNF'</li><li>'CD8_Tem'</li><li>'CD8_Tem_IFN'</li><li>'CD8_Tem_NFKB'</li><li>'CD8_Tn'</li><li>'MAIT'</li><li>'NK_adp'</li><li>'NK_CD56bright'</li><li>'NK_CD56dim'</li><li>'NK_resident'</li><li>'CD14+Mono pro-inflam'</li><li>'CD14+Mono_CTSS'</li><li>'CD14+Mono_hypo'</li><li>'CD14+Mono_IFN'</li><li>'CD14+Mono_S100'</li><li>'CD16+Mono'</li><li>'M2_Macro'</li><li>'Macro/Mono'</li><li>'cDC1'</li><li>'cDC2'</li><li>'Granulocyte'</li><li>'MastC'</li><li>'Neutrohpil_ARG1'</li><li>'Neutrophil_RPS/RPL'</li><li>'pDCs'</li><li>'EB'</li><li>'EB_MKi67_1'</li><li>'EB_MKi67_2'</li><li>'Erythrocyte'</li><li>'Erythrocyte_IFN'</li><li>'Erythrocyte_mito'</li><li>'Erythrocyte_prolif'</li><li>'MegaK'</li><li>'HSC'</li><li>'GMP'</li><li>'Fibroblasts'</li></ol>\n",
       "</details>"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item Erythrocyte\\_IFN\n",
       "\\item EB\\_MKi67\\_1\n",
       "\\item EB\n",
       "\\item Neutrophil\\_RPS/RPL\n",
       "\\item GMP\n",
       "\\item pDCs\n",
       "\\item Pc\\_MALAT1\n",
       "\\item Erythrocyte\n",
       "\\item HSC\n",
       "\\item cDC2\n",
       "\\item MAIT\n",
       "\\item CD4\\_Teff\n",
       "\\item CD14+Mono\\_S100\n",
       "\\item CD14+Mono\\_IFN\n",
       "\\item Granulocyte\n",
       "\\item CD16+Mono\n",
       "\\item CD4\\_Tcm\\_NFKBIA\n",
       "\\item Erythrocyte\\_mito\n",
       "\\item CD8\\_Tn\n",
       "\\item CD4\\_Teff\\_TNF\n",
       "\\item CD4\\_Tcm\\_KLRB1\n",
       "\\item CD14+Mono\\_CTSS\n",
       "\\item CD4\\_Tem\\_IFN\n",
       "\\item B\\_preLg\n",
       "\\item Pc\\_MT-CO2\n",
       "\\item CD14+Mono pro-inflam\n",
       "\\item Treg\n",
       "\\item CD8\\_Tcm\\_GZMK\n",
       "\\item CD4\\_Tn\n",
       "\\item CD4\\_Tcm\\_IFN\n",
       "\\item CD4\\_Th\n",
       "\\item B\\_pro\n",
       "\\item Fibroblasts\n",
       "\\item CD8\\_Teff\\_TNF\n",
       "\\item B\\_mem\n",
       "\\item CD8\\_Teff\\_HLA\n",
       "\\item CD8\\_T\\_adp\n",
       "\\item NK\\_CD56bright\n",
       "\\item CD8\\_Tem\\_NFKB\n",
       "\\item NK\\_CD56dim\n",
       "\\item CD8\\_Teff\n",
       "\\item Pc\\_JUN\n",
       "\\item CD8\\_Tem\\_IFN\n",
       "\\item CD8\\_Tem\n",
       "\\item NK\\_resident\n",
       "\\item B\\_imm\n",
       "\\item NK\\_adp\n",
       "\\item Pc\\_IGKC\n",
       "\\item MegaK\n",
       "\\item B\\_naive\n",
       "\\item CD4\\_Th\\_LEF1\n",
       "\\item B\\_preSm\n",
       "\\item B\\_trns\n",
       "\\item CD8\\_T\\_Apoptotic\n",
       "\\item Erythrocyte\\_prolif\n",
       "\\item cDC1\n",
       "\\item Pc\\_HBB\\_S100A8\n",
       "\\item CD8\\_Teff\\_b\n",
       "\\item Macro/Mono\n",
       "\\item Pc\\_HBB\\_HMGB1\n",
       "\\item CD4\\_CTL\n",
       "\\item CD8\\_Tcm\n",
       "\\item CD8\\_Teff\\_c\n",
       "\\item M2\\_Macro\n",
       "\\item EB\\_MKi67\\_2\n",
       "\\item Pc\\_ATF5\n",
       "\\item Pc\\_IFI27\n",
       "\\item Pc\\_EIF1\n",
       "\\item CD14+Mono\\_hypo\n",
       "\\item Pc\\_CD74\n",
       "\\item B\\_Unswitched\\_mem\n",
       "\\item Pc\\_IGHV3\n",
       "\\item Pc\\_IGHG3\n",
       "\\item Neutrohpil\\_ARG1\n",
       "\\item Pc\\_PPM1K\n",
       "\\item MastC\n",
       "\\item Pc\\_NOL4\n",
       "\\item CD8\\_T\\_adp\\_b\n",
       "\\item Pc\\_CCDC88A\n",
       "\\item Pc\\_IGHA1\n",
       "\\item Pc\\_BCMA\n",
       "\\item Pc\\_PUS3\n",
       "\\item Pc\\_MPO\n",
       "\\item Pc\\_IGKV4\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\emph{Levels}: \\begin{enumerate*}\n",
       "\\item 'Pc\\_ATF5'\n",
       "\\item 'Pc\\_BCMA'\n",
       "\\item 'Pc\\_CCDC88A'\n",
       "\\item 'Pc\\_CD74'\n",
       "\\item 'Pc\\_EIF1'\n",
       "\\item 'Pc\\_HBB\\_HMGB1'\n",
       "\\item 'Pc\\_HBB\\_S100A8'\n",
       "\\item 'Pc\\_IFI27'\n",
       "\\item 'Pc\\_IGHA1'\n",
       "\\item 'Pc\\_IGHG3'\n",
       "\\item 'Pc\\_IGHV3'\n",
       "\\item 'Pc\\_IGKV4'\n",
       "\\item 'Pc\\_IGKC'\n",
       "\\item 'Pc\\_JUN'\n",
       "\\item 'Pc\\_MALAT1'\n",
       "\\item 'Pc\\_MPO'\n",
       "\\item 'Pc\\_MT-CO2'\n",
       "\\item 'Pc\\_NOL4'\n",
       "\\item 'Pc\\_PPM1K'\n",
       "\\item 'Pc\\_PUS3'\n",
       "\\item 'B\\_pro'\n",
       "\\item 'B\\_preLg'\n",
       "\\item 'B\\_preSm'\n",
       "\\item 'B\\_imm'\n",
       "\\item 'B\\_naive'\n",
       "\\item 'B\\_trns'\n",
       "\\item 'B\\_mem'\n",
       "\\item 'B\\_Unswitched\\_mem'\n",
       "\\item 'CD4\\_CTL'\n",
       "\\item 'CD4\\_Tcm\\_IFN'\n",
       "\\item 'CD4\\_Tcm\\_KLRB1'\n",
       "\\item 'CD4\\_Tcm\\_NFKBIA'\n",
       "\\item 'CD4\\_Teff'\n",
       "\\item 'CD4\\_Teff\\_TNF'\n",
       "\\item 'CD4\\_Tem\\_IFN'\n",
       "\\item 'CD4\\_Th'\n",
       "\\item 'CD4\\_Th\\_LEF1'\n",
       "\\item 'CD4\\_Tn'\n",
       "\\item 'Treg'\n",
       "\\item 'CD8\\_T\\_adp'\n",
       "\\item 'CD8\\_T\\_adp\\_b'\n",
       "\\item 'CD8\\_T\\_Apoptotic'\n",
       "\\item 'CD8\\_Tcm'\n",
       "\\item 'CD8\\_Tcm\\_GZMK'\n",
       "\\item 'CD8\\_Teff'\n",
       "\\item 'CD8\\_Teff\\_b'\n",
       "\\item 'CD8\\_Teff\\_c'\n",
       "\\item 'CD8\\_Teff\\_HLA'\n",
       "\\item 'CD8\\_Teff\\_TNF'\n",
       "\\item 'CD8\\_Tem'\n",
       "\\item 'CD8\\_Tem\\_IFN'\n",
       "\\item 'CD8\\_Tem\\_NFKB'\n",
       "\\item 'CD8\\_Tn'\n",
       "\\item 'MAIT'\n",
       "\\item 'NK\\_adp'\n",
       "\\item 'NK\\_CD56bright'\n",
       "\\item 'NK\\_CD56dim'\n",
       "\\item 'NK\\_resident'\n",
       "\\item 'CD14+Mono pro-inflam'\n",
       "\\item 'CD14+Mono\\_CTSS'\n",
       "\\item 'CD14+Mono\\_hypo'\n",
       "\\item 'CD14+Mono\\_IFN'\n",
       "\\item 'CD14+Mono\\_S100'\n",
       "\\item 'CD16+Mono'\n",
       "\\item 'M2\\_Macro'\n",
       "\\item 'Macro/Mono'\n",
       "\\item 'cDC1'\n",
       "\\item 'cDC2'\n",
       "\\item 'Granulocyte'\n",
       "\\item 'MastC'\n",
       "\\item 'Neutrohpil\\_ARG1'\n",
       "\\item 'Neutrophil\\_RPS/RPL'\n",
       "\\item 'pDCs'\n",
       "\\item 'EB'\n",
       "\\item 'EB\\_MKi67\\_1'\n",
       "\\item 'EB\\_MKi67\\_2'\n",
       "\\item 'Erythrocyte'\n",
       "\\item 'Erythrocyte\\_IFN'\n",
       "\\item 'Erythrocyte\\_mito'\n",
       "\\item 'Erythrocyte\\_prolif'\n",
       "\\item 'MegaK'\n",
       "\\item 'HSC'\n",
       "\\item 'GMP'\n",
       "\\item 'Fibroblasts'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. Erythrocyte_IFN\n",
       "2. EB_MKi67_1\n",
       "3. EB\n",
       "4. Neutrophil_RPS/RPL\n",
       "5. GMP\n",
       "6. pDCs\n",
       "7. Pc_MALAT1\n",
       "8. Erythrocyte\n",
       "9. HSC\n",
       "10. cDC2\n",
       "11. MAIT\n",
       "12. CD4_Teff\n",
       "13. CD14+Mono_S100\n",
       "14. CD14+Mono_IFN\n",
       "15. Granulocyte\n",
       "16. CD16+Mono\n",
       "17. CD4_Tcm_NFKBIA\n",
       "18. Erythrocyte_mito\n",
       "19. CD8_Tn\n",
       "20. CD4_Teff_TNF\n",
       "21. CD4_Tcm_KLRB1\n",
       "22. CD14+Mono_CTSS\n",
       "23. CD4_Tem_IFN\n",
       "24. B_preLg\n",
       "25. Pc_MT-CO2\n",
       "26. CD14+Mono pro-inflam\n",
       "27. Treg\n",
       "28. CD8_Tcm_GZMK\n",
       "29. CD4_Tn\n",
       "30. CD4_Tcm_IFN\n",
       "31. CD4_Th\n",
       "32. B_pro\n",
       "33. Fibroblasts\n",
       "34. CD8_Teff_TNF\n",
       "35. B_mem\n",
       "36. CD8_Teff_HLA\n",
       "37. CD8_T_adp\n",
       "38. NK_CD56bright\n",
       "39. CD8_Tem_NFKB\n",
       "40. NK_CD56dim\n",
       "41. CD8_Teff\n",
       "42. Pc_JUN\n",
       "43. CD8_Tem_IFN\n",
       "44. CD8_Tem\n",
       "45. NK_resident\n",
       "46. B_imm\n",
       "47. NK_adp\n",
       "48. Pc_IGKC\n",
       "49. MegaK\n",
       "50. B_naive\n",
       "51. CD4_Th_LEF1\n",
       "52. B_preSm\n",
       "53. B_trns\n",
       "54. CD8_T_Apoptotic\n",
       "55. Erythrocyte_prolif\n",
       "56. cDC1\n",
       "57. Pc_HBB_S100A8\n",
       "58. CD8_Teff_b\n",
       "59. Macro/Mono\n",
       "60. Pc_HBB_HMGB1\n",
       "61. CD4_CTL\n",
       "62. CD8_Tcm\n",
       "63. CD8_Teff_c\n",
       "64. M2_Macro\n",
       "65. EB_MKi67_2\n",
       "66. Pc_ATF5\n",
       "67. Pc_IFI27\n",
       "68. Pc_EIF1\n",
       "69. CD14+Mono_hypo\n",
       "70. Pc_CD74\n",
       "71. B_Unswitched_mem\n",
       "72. Pc_IGHV3\n",
       "73. Pc_IGHG3\n",
       "74. Neutrohpil_ARG1\n",
       "75. Pc_PPM1K\n",
       "76. MastC\n",
       "77. Pc_NOL4\n",
       "78. CD8_T_adp_b\n",
       "79. Pc_CCDC88A\n",
       "80. Pc_IGHA1\n",
       "81. Pc_BCMA\n",
       "82. Pc_PUS3\n",
       "83. Pc_MPO\n",
       "84. Pc_IGKV4\n",
       "\n",
       "\n",
       "\n",
       "**Levels**: 1. 'Pc_ATF5'\n",
       "2. 'Pc_BCMA'\n",
       "3. 'Pc_CCDC88A'\n",
       "4. 'Pc_CD74'\n",
       "5. 'Pc_EIF1'\n",
       "6. 'Pc_HBB_HMGB1'\n",
       "7. 'Pc_HBB_S100A8'\n",
       "8. 'Pc_IFI27'\n",
       "9. 'Pc_IGHA1'\n",
       "10. 'Pc_IGHG3'\n",
       "11. 'Pc_IGHV3'\n",
       "12. 'Pc_IGKV4'\n",
       "13. 'Pc_IGKC'\n",
       "14. 'Pc_JUN'\n",
       "15. 'Pc_MALAT1'\n",
       "16. 'Pc_MPO'\n",
       "17. 'Pc_MT-CO2'\n",
       "18. 'Pc_NOL4'\n",
       "19. 'Pc_PPM1K'\n",
       "20. 'Pc_PUS3'\n",
       "21. 'B_pro'\n",
       "22. 'B_preLg'\n",
       "23. 'B_preSm'\n",
       "24. 'B_imm'\n",
       "25. 'B_naive'\n",
       "26. 'B_trns'\n",
       "27. 'B_mem'\n",
       "28. 'B_Unswitched_mem'\n",
       "29. 'CD4_CTL'\n",
       "30. 'CD4_Tcm_IFN'\n",
       "31. 'CD4_Tcm_KLRB1'\n",
       "32. 'CD4_Tcm_NFKBIA'\n",
       "33. 'CD4_Teff'\n",
       "34. 'CD4_Teff_TNF'\n",
       "35. 'CD4_Tem_IFN'\n",
       "36. 'CD4_Th'\n",
       "37. 'CD4_Th_LEF1'\n",
       "38. 'CD4_Tn'\n",
       "39. 'Treg'\n",
       "40. 'CD8_T_adp'\n",
       "41. 'CD8_T_adp_b'\n",
       "42. 'CD8_T_Apoptotic'\n",
       "43. 'CD8_Tcm'\n",
       "44. 'CD8_Tcm_GZMK'\n",
       "45. 'CD8_Teff'\n",
       "46. 'CD8_Teff_b'\n",
       "47. 'CD8_Teff_c'\n",
       "48. 'CD8_Teff_HLA'\n",
       "49. 'CD8_Teff_TNF'\n",
       "50. 'CD8_Tem'\n",
       "51. 'CD8_Tem_IFN'\n",
       "52. 'CD8_Tem_NFKB'\n",
       "53. 'CD8_Tn'\n",
       "54. 'MAIT'\n",
       "55. 'NK_adp'\n",
       "56. 'NK_CD56bright'\n",
       "57. 'NK_CD56dim'\n",
       "58. 'NK_resident'\n",
       "59. 'CD14+Mono pro-inflam'\n",
       "60. 'CD14+Mono_CTSS'\n",
       "61. 'CD14+Mono_hypo'\n",
       "62. 'CD14+Mono_IFN'\n",
       "63. 'CD14+Mono_S100'\n",
       "64. 'CD16+Mono'\n",
       "65. 'M2_Macro'\n",
       "66. 'Macro/Mono'\n",
       "67. 'cDC1'\n",
       "68. 'cDC2'\n",
       "69. 'Granulocyte'\n",
       "70. 'MastC'\n",
       "71. 'Neutrohpil_ARG1'\n",
       "72. 'Neutrophil_RPS/RPL'\n",
       "73. 'pDCs'\n",
       "74. 'EB'\n",
       "75. 'EB_MKi67_1'\n",
       "76. 'EB_MKi67_2'\n",
       "77. 'Erythrocyte'\n",
       "78. 'Erythrocyte_IFN'\n",
       "79. 'Erythrocyte_mito'\n",
       "80. 'Erythrocyte_prolif'\n",
       "81. 'MegaK'\n",
       "82. 'HSC'\n",
       "83. 'GMP'\n",
       "84. 'Fibroblasts'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] Erythrocyte_IFN      EB_MKi67_1           EB                  \n",
       " [4] Neutrophil_RPS/RPL   GMP                  pDCs                \n",
       " [7] Pc_MALAT1            Erythrocyte          HSC                 \n",
       "[10] cDC2                 MAIT                 CD4_Teff            \n",
       "[13] CD14+Mono_S100       CD14+Mono_IFN        Granulocyte         \n",
       "[16] CD16+Mono            CD4_Tcm_NFKBIA       Erythrocyte_mito    \n",
       "[19] CD8_Tn               CD4_Teff_TNF         CD4_Tcm_KLRB1       \n",
       "[22] CD14+Mono_CTSS       CD4_Tem_IFN          B_preLg             \n",
       "[25] Pc_MT-CO2            CD14+Mono pro-inflam Treg                \n",
       "[28] CD8_Tcm_GZMK         CD4_Tn               CD4_Tcm_IFN         \n",
       "[31] CD4_Th               B_pro                Fibroblasts         \n",
       "[34] CD8_Teff_TNF         B_mem                CD8_Teff_HLA        \n",
       "[37] CD8_T_adp            NK_CD56bright        CD8_Tem_NFKB        \n",
       "[40] NK_CD56dim           CD8_Teff             Pc_JUN              \n",
       "[43] CD8_Tem_IFN          CD8_Tem              NK_resident         \n",
       "[46] B_imm                NK_adp               Pc_IGKC             \n",
       "[49] MegaK                B_naive              CD4_Th_LEF1         \n",
       "[52] B_preSm              B_trns               CD8_T_Apoptotic     \n",
       "[55] Erythrocyte_prolif   cDC1                 Pc_HBB_S100A8       \n",
       "[58] CD8_Teff_b           Macro/Mono           Pc_HBB_HMGB1        \n",
       "[61] CD4_CTL              CD8_Tcm              CD8_Teff_c          \n",
       "[64] M2_Macro             EB_MKi67_2           Pc_ATF5             \n",
       "[67] Pc_IFI27             Pc_EIF1              CD14+Mono_hypo      \n",
       "[70] Pc_CD74              B_Unswitched_mem     Pc_IGHV3            \n",
       "[73] Pc_IGHG3             Neutrohpil_ARG1      Pc_PPM1K            \n",
       "[76] MastC                Pc_NOL4              CD8_T_adp_b         \n",
       "[79] Pc_CCDC88A           Pc_IGHA1             Pc_BCMA             \n",
       "[82] Pc_PUS3              Pc_MPO               Pc_IGKV4            \n",
       "84 Levels: Pc_ATF5 Pc_BCMA Pc_CCDC88A Pc_CD74 Pc_EIF1 ... Fibroblasts"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(obj$celltype_f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "877c4db6-d64b-4947-9041-d01a440ce9c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'B_preLg'</li><li>'B_pro'</li><li>'B_mem'</li><li>'B_imm'</li><li>'B_naive'</li><li>'B_preSm'</li><li>'B_trns'</li><li>'B_Unswitched_mem'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'B\\_preLg'\n",
       "\\item 'B\\_pro'\n",
       "\\item 'B\\_mem'\n",
       "\\item 'B\\_imm'\n",
       "\\item 'B\\_naive'\n",
       "\\item 'B\\_preSm'\n",
       "\\item 'B\\_trns'\n",
       "\\item 'B\\_Unswitched\\_mem'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'B_preLg'\n",
       "2. 'B_pro'\n",
       "3. 'B_mem'\n",
       "4. 'B_imm'\n",
       "5. 'B_naive'\n",
       "6. 'B_preSm'\n",
       "7. 'B_trns'\n",
       "8. 'B_Unswitched_mem'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"B_preLg\"          \"B_pro\"            \"B_mem\"            \"B_imm\"           \n",
       "[5] \"B_naive\"          \"B_preSm\"          \"B_trns\"           \"B_Unswitched_mem\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bcells <- grep(\"^B_\", unique(obj$celltype_f), value = TRUE)\n",
    "bcells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3c76e345-ecd0-4380-9770-ebec7a5fae3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "b <- subset(obj, subset= celltype_f %in% bcells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f4dcab64-255f-41af-abd6-40a91fa3e069",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Baseline'</li><li>'Remission/Response'</li><li>'Relapse/Progression'</li><li>'Post-transplant'</li><li>'100 Post ASCT'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Baseline'\n",
       "\\item 'Remission/Response'\n",
       "\\item 'Relapse/Progression'\n",
       "\\item 'Post-transplant'\n",
       "\\item '100 Post ASCT'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Baseline'\n",
       "2. 'Remission/Response'\n",
       "3. 'Relapse/Progression'\n",
       "4. 'Post-transplant'\n",
       "5. '100 Post ASCT'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"Baseline\"            \"Remission/Response\"  \"Relapse/Progression\"\n",
       "[4] \"Post-transplant\"     \"100 Post ASCT\"      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(b$collection_event)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a793d4f5-8413-4d5f-aa82-5e012a277b67",
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(b) <- 'collection_event'\n",
    "b <- RenameIdents(b, \n",
    "             '100 Post ASCT'='Post-transplant'\n",
    "                )\n",
    "b$collection <- Idents(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "09c175d1-c43e-4dfd-bf1f-2eb5431772e1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Scaling data with a low number of groups may produce misleading results”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>agg_record_2070518142:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_2070518142:} 2"
      ],
      "text/markdown": [
       "**agg_record_2070518142:** 2"
      ],
      "text/plain": [
       "agg_record_2070518142 \n",
       "                    2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "p <- DotPlot(\n",
    "  b,\n",
    "  features = 'SOX4',\n",
    "  assay = 'RNA',\n",
    "  cols = c(\"white\", \"black\"),\n",
    "  group.by='collection'\n",
    ")\n",
    "pdf('sox4expr_mmrf_longitudinal_bcells.pdf', width=5, height=2.5)\n",
    "print(p)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b2c61a7-f903-4360-a03d-3e0e9ab72ba4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
